{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Biopython"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## General"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1.79\n",
      "Sequence: GGCAT\n",
      " - GC content: 60.0\n",
      " - Complement: CCGTA\n",
      " - Reverse complement: ATGCC\n",
      " - Change T>U: GGCAU\n",
      " - Change U>T: GGCAT\n",
      " - Translate: G\n",
      " - Melting temperature (Wallace): Tm = 16.0\n",
      " - Melting temperature (GC): Tm = -35.745107734584835\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "4"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "####################################################################################\n",
    "###   General   ####################################################################\n",
    "####################################################################################\n",
    "\n",
    "\n",
    "import Bio; from Bio.Seq import Seq; from Bio.SeqUtils import GC; from Bio.SeqUtils import MeltingTemp as mt\n",
    "\n",
    "# Methods and attributes:\n",
    "dir(Bio)\n",
    "dir(Seq)\n",
    "\n",
    "print(Bio.__version__)\n",
    "\n",
    "# Make a special 'sequence' object\n",
    "s1 = Seq(\"GGCAT\")\n",
    "print( f\"Sequence: { s1 }\" )\n",
    "# Information about sequences\n",
    "print( f\" - GC content: { GC(s1) }\" )\n",
    "print( f\" - Complement: { s1.complement() }\" )\n",
    "print( f\" - Reverse complement: { s1.reverse_complement() }\" )\n",
    "print( f\" - Change T>U: { s1.transcribe() }\" )\n",
    "print( f\" - Change U>T: { s1.back_transcribe() }\" )\n",
    "print( f\"\"\" - Translate: { s1.translate(\n",
    "\tstop_symbol= '@'\n",
    ") }\"\"\" )\n",
    "print( f\" - Melting temperature (Wallace): Tm = { mt.Tm_Wallace(s1) }\" )\n",
    "print( f\" - Melting temperature (GC): Tm = { mt.Tm_GC(s1) }\" )\n",
    "\n",
    "\n",
    "# 1- and 3-letter codes transitioning\n",
    "from Bio.SeqUtils import seq1, seq3\n",
    "seq3('MDE') # 1 -> 3 letter code\n",
    "seq1('MetAspGlu') # 3 -> 1 letter code\n",
    "\n",
    "# view codon table\n",
    "from Bio.Data import CodonTable\n",
    "# dir(CodonTable)\n",
    "# print(CodonTable.unambiguous_dna_by_name['Standard'])\n",
    "# print(CodonTable.unambiguous_rna_by_name['Standard'])\n",
    "\n",
    "# Levenshtein distance\n",
    "from Levenshtein import distance\n",
    "s2 = Seq(\"GATCATAC\")\n",
    "distance(s1, s2) \n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Alignments"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[32mGlobal alignment:\u001b[0m\n",
      "CCC---DADBADGADGSDG\n",
      "      ||||   |   | \n",
      "---XXXDADB---A---D-\n",
      "  Score=6\n",
      "\n",
      "Max similarity\n",
      "CCC---DADBADGADGSDG\n",
      "      ||||||       \n",
      "---XXXDADBAD-------\n",
      "  Score=9.5\n",
      "\n",
      "\n",
      "\u001b[32mLocal alignment:\u001b[0m\n",
      "Alignment(seqA='CCCDADBADGADGSDG', seqB='XXXDADB---A---D-', score=6.0, start=3, end=15)\n",
      "4 DADBADGADGSD\n",
      "  ||||   |   |\n",
      "4 DADB---A---D\n",
      "  Score=6\n",
      "\n",
      "Max similarity\n",
      "4 DADBAD\n",
      "  ||||||\n",
      "4 DADBAD\n",
      "  Score=12\n",
      "\n"
     ]
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAPYAAAD4CAYAAADfGBu4AAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjQuMywgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy/MnkTPAAAACXBIWXMAAAsTAAALEwEAmpwYAAALgUlEQVR4nO3dQYhcVRbG8e+bjCZkxEViSBgMERdBUJNgmrgxSBLBpQkTJL0WwojMbsRtz8pNVooQsguC9CwkrkQXyUKFYEyLPdCbgIYgo4FMu5DMYDqEM4uudpqmM/Wq3r2p28f/D0L3e1W5Obdffbmvq16dckQIQC6/m3QBAMoj2EBCBBtIiGADCRFsIKHf1xr4YW+OLfpD73H27vtPgWrKufaPrcXGam1uGK7k8e/rF/1bS3HH693mWi93Pept8byP9R7n0x++6V9MQS//8UCxsVqbG4Yrefz7+jIu6uf4ad1gcyoOJESwgYQINpAQwQYS6hxs28dth+2nahYEoL9RVuxpSV8MvgJoWKdg235E0guSXpN0qmpFAHrrumK/IumTiLgmadH2wfXuZPu07au2r97VnWJFAhhN12BPS5odfD+r+5yOR8S5iJiKiKmHtLlEfQDGMPSSUtvbJB2V9KztkLRJUth+M+jSADSpy4p9UtL7EbEnIp6IiN2Srks6XLc0AOPqEuxpSRfW7PtQPDsONGvoqXhEHFln3zt1ygFQAleeAQkRbCAhgg0kVK3RwtT+LXHl0929xyn1xvYWmxq0NrfW6ikla3OMQy9/r6vzv9BoAfitINhAQgQbSIhgAwkRbCChTn3FbW+XdHGwuUvSPUm3BtuHImKpQm0AxtQp2BGxKOmAJNmekXQ7Is7UKwtAH5yKAwkRbCChosFe3Rrp1uK9kkMDGEHRYK9ujbRj+6aSQwMYAafiQEIEG0ho5M/HjoiZCnUAKIgVG0iIYAMJEWwgoZF/x37QMncHaakbh9RePS0es42CFRtIiGADCRFsICGCDSREsIGEOgXb9k7bH9j+zvac7cu2T9QuDsB4hgbbtiV9JOmziHgyIg5KOiXp8cq1ARhTl9exj0paioizKzsi4oakd6tVBaCXLqfiT0v6unYhAMoZ+ckz2+/Znrf91Tq30UEFaECXYC9Iem5lIyLekHRM0o61d6SDCtCGLsG+JGmL7ddX7dtaqR4ABQwNdix/zu5xSS/avm77iqTzkt6qXBuAMXX9wIAftfwSF4ANgCvPgIQINpAQwQYS8vJzY+U96m3xvI/1Hqe17helunpI7c0Nw5U8/n19GRf1c/zk9W5jxQYSIthAQgQbSIhgAwkRbCAhgg0k1DnYto/bDttP1SwIQH+jrNjTkr4YfAXQsK7NDB+R9IKk18SbQYDmdV2xX5H0SURck7Ro++B6d1rdQeWu7hQrEsBougZ7WtLs4PtZ3ed0fHUHlYe0uUR9AMYw9P3YtrdpuVPps7ZD0iZJYfvNqHWhOYBeuqzYJyW9HxF7IuKJiNgt6bqkw3VLAzCuLsGelnRhzb4PxbPjQLOGnopHxJF19r1TpxwAJXDlGZAQwQYSIthAQtVaI03t3xJXPt3de5xSrWhabEPU2txaq6eUrO2sDr38va7O/0JrJOC3gmADCRFsICGCDSREsIGEOn0on+3tki4ONndJuifp1mD7UEQsVagNwJi6ftrmoqQDkmR7RtLtiDhTrywAfXAqDiRUNNirO6jcWrxXcmgAIyga7NUdVHZs31RyaAAj4FQcSIhgAwkRbCChTi93rRYRMxXqAFAQKzaQEMEGEiLYQEIj/479oLXUsaK0rJ1PWqsn82PoflixgYQINpAQwQYSIthAQgQbSKhzsG3vsj1r+1vbc7Y/tr23ZnEAxtO1NZK1/Imb5yPi1GDffkk7JV2rVx6AcXR9HfuIpLsRcXZlR0TM1ykJQF9dT8WfkTQ37E50UAHaQAcVIKGuwV6QdLBmIQDK6RrsS5I22z69ssP2PtuH65QFoI9OwY7lz9o9IemlwctdC5LelnSzZnEAxtP53V0R8YOkVyvWAqAQrjwDEiLYQEIEG0io+Q4qrXXjKKm1ubVWTyml5iW1N7f7YcUGEiLYQEIEG0iIYAMJEWwgoa6NFrZLujjY3CXpnqRbg+1DEbFUoTYAY+oU7IhYlHRAkmzPSLodEWfqlQWgD07FgYSKBpsOKkAb6KACJMSpOJAQwQYSIthAQiO/uysiZirUAaAgVmwgIYINJESwgYSa76CyUTpWjCNr55PW6sn8GLofVmwgIYINJESwgYQINpAQwQYS6hxs27tszw4+lG/O9se299YsDsB4urZGsqQLks5HxKnBvv2Sdkq6Vq88AOPo+jr2EUl3I+Lsyo6ImK9TEoC+up6KPyNpbtid6KACtIEOKkBCXYO9IOlgzUIAlNM12JckbbZ9emWH7X22D9cpC0AfnYIdESHphKSXBi93LUh6W9LNmsUBGE/nd3dFxA+SXq1YC4BCuPIMSIhgAwkRbCCh5juolNJaV4+SWqspayeWjYQVG0iIYAMJEWwgIYINJESwgYQ6Bdv2Ttsf2P5u0D3lsu0TtYsDMJ6hwR50T/lI0mcR8WREHJR0StLjlWsDMKYur2MflbS0pnvKDUnvVqsKQC9dTsWflvR1l8HooAK0YeQnz2y/Z3ve9ldrb6ODCtCGLsFekPTcykZEvCHpmKQdtYoC0E+XYF+StMX266v2ba1UD4AChgZ70D3luKQXbV+3fUXSeUlvVa4NwJg6vbsrIn7U8ktcADYArjwDEiLYQEIEG0io+Q4qmbtotDa31uppbV5Sm4+j9bBiAwkRbCAhgg0kRLCBhAg2kFCnZ8Vtb5d0cbC5S9I9SbcG24ciYqlCbQDG1PWS0kVJByTJ9oyk2xFxpl5ZAPrgVBxIiGADCRUNNq2RgDYUDTatkYA2cCoOJESwgYRGfndXRMxUqANAQazYQEIEG0iIYAMJEWwgIS+3DS/vUW+L532s9zittaL5LbbZwf+UPP59fRkX9XP85PVuY8UGEiLYQEIEG0iIYAMJdQ627eO2w/ZTNQsC0N8oK/a0pC8GXwE0rFOwbT8i6QVJr4lP3QSa13XFfkXSJxFxTdKi7YMVawLQU9dgT0uaHXw/q/ucjq/uoHJXd0rUB2AMQ9+2aXubpKOSnrUdkjZJCttvxprL1iLinKRz0vKVZxXqBdBBlxX7pKT3I2JPRDwREbslXZd0uG5pAMbVJdjTki6s2feheHYcaNbQU/GIOLLOvnfqlAOgBK48AxIi2EBCBBtIiGADCRFsICGCDSREsIGECDaQEMEGEiLYQEK0RgISojUSkBCtkYCEirZGooMK0IairZEi4lxETEXE1EPaXKI+AGMo2hoJQBtojQQkRGskICFaIwEJceUZkBDBBhIi2EBCrvWKle1bkm4Mudtjkv5V4J9jnAc3FuO0M86eiNix3g3Vgt2F7asRMcU4dcdpsSbGqTsOp+JAQgQbSGjSwT7HOA9knJJjMc4GGGeiv2MDqGPSKzaACgg2kNDEgt23h5rt7ba/Gfy5afufq7YfHnGsnbY/sP2d7Tnbl22fGLOuZuY1GG+X7Vnb3w7m9rHtvZOqqUQ9g3F6H7MKP+sifQGLjBMRE/kj6e+SPpf0twJjzUj665h/15IuS/rzqn17JP1lI8/r/8xtv6TDDf2sR66n9DEr8bMueexLjDORFbuxHmpHJS1FxNmVHRFxIyLeHXWgxuYlSUck3V0zt/mI+HyD11PsmJVS6tiXGmdSp+Kdeqg9IE9L+rrQWC3NS5KekTQ34RpWK1VPyWNWSqljX2ScSQW7Uw+1SbD9nu1521+N8debnVdmPY9ZKaWOfZFxhjZaKK3BHmoLkv60shERb9h+TNLVUQZpcF7S8txOTujfXk+peoocs1JKHfuSj6FJrNit9VC7JGmL7ddX7ds6xjitzUtanttm26dXdtjeZ3uSP+sS9ZQ6ZqWUOvbFHkOTCHZTPdQG/xMel/Si7eu2r0g6L+mtEYdqal7Sr3M7IemlwctLC5LelnRzI9dT8JiVUurYF3sMcUkpkBBXngEJEWwgIYINJESwgYQINpAQwQYSIthAQv8FULG+y1D2SbYAAAAASUVORK5CYII=",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "####################################################################################\n",
    "###   Alignments   #################################################################\n",
    "####################################################################################\n",
    "\n",
    "from Bio import pairwise2\n",
    "from Bio.pairwise2 import format_alignment\n",
    "from Bio.Seq import Seq\n",
    "from colorama import Fore, Style\n",
    "\n",
    "sequence1 = Seq(\"CCCDADBADGADGSDG\")\n",
    "sequence2 = Seq(\"XXXDADBAD\")\n",
    "\n",
    "# Global Alignment\n",
    "print(f\"{Fore.GREEN}Global alignment:{Style.RESET_ALL}\")\n",
    "alignments = pairwise2.align.globalxx(sequence1, sequence2)\n",
    "print(format_alignment(*alignments[0]))\n",
    "\n",
    "## to view all alignments:\n",
    "# for a in alignments:\n",
    "# \tprint(format_alignment(*a))\n",
    "\n",
    "## Get alignment by the score only:\n",
    "# alignment2 = pairwise2.align.globalxx(sequence1, sequence2, one_alignment_only=True, score_only=True)\n",
    "# alignment2 / len(seq1) * 100\n",
    "\n",
    "# Max similarity\n",
    "print(\"Max similarity\")\n",
    "glb_alignment_ms = pairwise2.align.globalms(sequence1, sequence2, 2, -1, -0.5, -0.1)\n",
    "for a in glb_alignment_ms:\n",
    "\tprint(format_alignment(*a))\n",
    "\n",
    "# Local Alignment\n",
    "print(f\"\\n{Fore.GREEN}Local alignment:{Style.RESET_ALL}\")\n",
    "loc_alignments = pairwise2.align.localxx(sequence1, sequence2)\n",
    "print(loc_alignments[0])\n",
    "print(format_alignment(*loc_alignments[0]))\n",
    "# Max similarity\n",
    "print(\"Max similarity\")\n",
    "local_alignment_ms = pairwise2.align.localms(sequence1, sequence2, 2, -1, -0.5, -0.1)\n",
    "for a in local_alignment_ms:\n",
    "\tprint(format_alignment(*a))\n",
    "\n",
    "###############################################################################\n",
    "###   Using dotplot to check two sequences for similarity   ###################\n",
    "###############################################################################\n",
    "\n",
    "from Bio.Seq import Seq\n",
    "import matplotlib.pyplot as plt\n",
    "import numpy as np\n",
    "\n",
    "\n",
    "def delta(x,y):\n",
    "    return 0 if x == y else 1\n",
    "\n",
    "def M(seq1,seq2,i,j,k):\n",
    "    return sum(delta(x,y) for x,y in zip(seq1[i:i+k],seq2[j:j+k]))\n",
    "\n",
    "def makeMatrix(seq1,seq2,k):\n",
    "    n = len(seq1)\n",
    "    m = len(seq2)\n",
    "    return [[M(seq1,seq2,i,j,k) for j in range(m-k+1)] for i in range(n-k+1)]\n",
    "\n",
    "def plotMatrix(M,t, seq1, seq2, nonblank = chr(0x25A0), blank = ' '):\n",
    "    print(' |' + seq2)\n",
    "    print('-'*(2 + len(seq2)))\n",
    "    for label,row in zip(seq1,M):\n",
    "        line = ''.join(nonblank if s < t else blank for s in row)\n",
    "        print(label + '|' + line)\n",
    "\n",
    "\n",
    "\n",
    "def dotplot(seq1,seq2,k = 1,t = 1):\n",
    "    M = makeMatrix(seq1,seq2,k)\n",
    "    plotMatrix(M, t, seq1,seq2) #experiment with character choice\n",
    "\n",
    "seq1 = Seq(\"ACTTAG\")\n",
    "seq2 = Seq(\"ACTT\")\n",
    "# dotplot(seq1, seq2)\n",
    "\n",
    "def dotplotx(seq1, seq2):\n",
    "    plt.imshow(np.array(makeMatrix(seq1, seq2, 1)))\n",
    "    xt = plt.xticks(np.arange(len(list(seq1))), list(seq1))\n",
    "    yt = plt.yticks(np.arange(len(list(seq1))), list(seq1))\n",
    "    plt.show()\n",
    "# dotplotx(seq1, seq2)\n",
    "\n",
    "dna1 = Seq(\"ATGATCTCGTAA\")\n",
    "dna2 = Seq(\"ATTATGTCGTAA\")\n",
    "dotplotx(dna1, dna1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ID: MN908947.3\n",
      "Name: MN908947.3\n",
      "Description: MN908947.3 Severe acute respiratory syndrome coronavirus 2 isolate Wuhan-Hu-1, complete genome\n",
      "Number of features: 0\n",
      "Seq('ATTAAAGGTTTATACCTTCCCAGGTAACAAACCAACCAACTTTCGATCTCTTGT...AAA')\n",
      "ID: MN908947.3\n",
      "Name: MN908947\n",
      "Description: Severe acute respiratory syndrome coronavirus 2 isolate Wuhan-Hu-1, complete genome\n",
      "Number of features: 23\n",
      "/molecule_type=ss-RNA\n",
      "/topology=linear\n",
      "/data_file_division=VRL\n",
      "/date=18-MAR-2020\n",
      "/accessions=['MN908947']\n",
      "/sequence_version=3\n",
      "/keywords=['']\n",
      "/source=Severe acute respiratory syndrome coronavirus 2 (SARS-CoV-2)\n",
      "/organism=Severe acute respiratory syndrome coronavirus 2\n",
      "/taxonomy=['Viruses', 'Riboviria', 'Orthornavirae', 'Pisuviricota', 'Pisoniviricetes', 'Nidovirales', 'Cornidovirineae', 'Coronaviridae', 'Orthocoronavirinae', 'Betacoronavirus', 'Sarbecovirus']\n",
      "/references=[Reference(title='A new coronavirus associated with human respiratory disease in China', ...), Reference(title='Direct Submission', ...)]\n",
      "/comment=On Jan 17, 2020 this sequence version replaced MN908947.2.\n",
      "/structured_comment=OrderedDict([('Assembly-Data', OrderedDict([('Assembly Method', 'Megahit v. V1.1.3'), ('Sequencing Technology', 'Illumina')]))])\n",
      "Seq('ATTAAAGGTTTATACCTTCCCAGGTAACAAACCAACCAACTTTCGATCTCTTGT...AAA')\n"
     ]
    }
   ],
   "source": [
    "####################################################################################\n",
    "###   Read biological formats (.fasta, .gb)   ######################################\n",
    "####################################################################################\n",
    "\n",
    "from Bio import SeqIO\n",
    "\n",
    "# loading FASTA\n",
    "for record in SeqIO.parse(\"Examples/sequence.fasta\", \"fasta\"):\n",
    "\t# print(record.id)\n",
    "\t# print(record.description)\n",
    "\tprint(record)\n",
    "\tpass\n",
    "\n",
    "# reading sequences in FASTA\n",
    "# dna_record = SeqIO.read(\"Examples/sequence.fasta\", \"fasta\")\n",
    "# dna_seq = dna_record.seq\n",
    "\n",
    "\n",
    "# Read GenBank\n",
    "for record in SeqIO.parse(\"Examples/sequence.gb\", \"genbank\"):\n",
    "\tprint(record)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Visualise 3D structures"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\Users\\evgen\\AppData\\Local\\Programs\\Python\\Python39\\lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:89: PDBConstructionWarning: WARNING: Chain A is discontinuous at line 2811.\n",
      "  warnings.warn(\n"
     ]
    },
    {
     "data": {
      "application/3dmoljs_load.v0": "<div id=\"3dmolviewer_1645195829195651\"  style=\"position: relative; width: 640px; height: 480px\">\n        <p id=\"3dmolwarning_1645195829195651\" style=\"background-color:#ffcccc;color:black\">You appear to be running in JupyterLab (or JavaScript failed to load for some other reason).  You need to install the 3dmol extension: <br>\n        <tt>jupyter labextension install jupyterlab_3dmol</tt></p>\n        </div>\n<script>\n\nvar loadScriptAsync = function(uri){\n  return new Promise((resolve, reject) => {\n    var tag = document.createElement('script');\n    tag.src = uri;\n    tag.async = true;\n    tag.onload = () => {\n      resolve();\n    };\n  var firstScriptTag = document.getElementsByTagName('script')[0];\n  firstScriptTag.parentNode.insertBefore(tag, firstScriptTag);\n});\n};\n\nif(typeof $3Dmolpromise === 'undefined') {\n$3Dmolpromise = null;\n  $3Dmolpromise = loadScriptAsync('https://3dmol.org/build/3Dmol.js');\n}\n\nvar viewer_1645195829195651 = null;\nvar warn = document.getElementById(\"3dmolwarning_1645195829195651\");\nif(warn) {\n    warn.parentNode.removeChild(warn);\n}\n$3Dmolpromise.then(function() {\nviewer_1645195829195651 = $3Dmol.createViewer($(\"#3dmolviewer_1645195829195651\"),{backgroundColor:\"white\"});\n$3Dmol.download(\"pdb:6LU7\", viewer_1645195829195651, {}, function() {\nviewer_1645195829195651.zoomTo();\n\tviewer_1645195829195651.setStyle({\"cartoon\": {\"color\": \"spectrum\"}});\nviewer_1645195829195651.render();\n})\n});\n</script>",
      "text/html": [
       "<div id=\"3dmolviewer_1645195829195651\"  style=\"position: relative; width: 640px; height: 480px\">\n",
       "        <p id=\"3dmolwarning_1645195829195651\" style=\"background-color:#ffcccc;color:black\">You appear to be running in JupyterLab (or JavaScript failed to load for some other reason).  You need to install the 3dmol extension: <br>\n",
       "        <tt>jupyter labextension install jupyterlab_3dmol</tt></p>\n",
       "        </div>\n",
       "<script>\n",
       "\n",
       "var loadScriptAsync = function(uri){\n",
       "  return new Promise((resolve, reject) => {\n",
       "    var tag = document.createElement('script');\n",
       "    tag.src = uri;\n",
       "    tag.async = true;\n",
       "    tag.onload = () => {\n",
       "      resolve();\n",
       "    };\n",
       "  var firstScriptTag = document.getElementsByTagName('script')[0];\n",
       "  firstScriptTag.parentNode.insertBefore(tag, firstScriptTag);\n",
       "});\n",
       "};\n",
       "\n",
       "if(typeof $3Dmolpromise === 'undefined') {\n",
       "$3Dmolpromise = null;\n",
       "  $3Dmolpromise = loadScriptAsync('https://3dmol.org/build/3Dmol.js');\n",
       "}\n",
       "\n",
       "var viewer_1645195829195651 = null;\n",
       "var warn = document.getElementById(\"3dmolwarning_1645195829195651\");\n",
       "if(warn) {\n",
       "    warn.parentNode.removeChild(warn);\n",
       "}\n",
       "$3Dmolpromise.then(function() {\n",
       "viewer_1645195829195651 = $3Dmol.createViewer($(\"#3dmolviewer_1645195829195651\"),{backgroundColor:\"white\"});\n",
       "$3Dmol.download(\"pdb:6LU7\", viewer_1645195829195651, {}, function() {\n",
       "viewer_1645195829195651.zoomTo();\n",
       "\tviewer_1645195829195651.setStyle({\"cartoon\": {\"color\": \"spectrum\"}});\n",
       "viewer_1645195829195651.render();\n",
       "})\n",
       "});\n",
       "</script>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "<py3Dmol.view at 0x1deb000d4f0>"
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "####################################################################################\n",
    "###   Visualise 3D structures   ####################################################\n",
    "####################################################################################\n",
    "\"\"\"\n",
    "Packages to visualise 3D structures:\n",
    "- nglview\n",
    "- py3Dmol\n",
    "- pytraj\n",
    "\"\"\"\n",
    "\n",
    "\n",
    "\n",
    "# Reading pdb file for 3d structure\n",
    "from Bio.PDB import PDBParser, MMCIFParser\n",
    "\n",
    "# creating a parser\n",
    "parser = PDBParser()\n",
    "structure = parser.get_structure(\"6LU7\", \"Examples/6lu7.pdb\") \n",
    "# can use the same command as the above one but for format MMCIF\n",
    "\n",
    "structure\n",
    "\n",
    "# models in the structure\n",
    "model = structure[0]\n",
    "\n",
    "# structure => model => chain => residue => atom => residue\n",
    "\n",
    "# check for chains\n",
    "# for chain in model:\n",
    "# \tprint(f\"Chain {chain}, Chain_ID {chain.id}\")\n",
    "\n",
    "# Check for residue\n",
    "# for chain in model:\n",
    "# \tfor residue in chain:\n",
    "# \t\tprint(residue)\n",
    "\n",
    "# check for atoms\n",
    "# for chain in model:\n",
    "# \tprint(f\"Chain {chain}, Chain_ID{chain.id}\")\n",
    "# \tfor residue in chain:\n",
    "# \t\tfor atom in residue:\n",
    "# \t\t\tprint(atom)\n",
    "\n",
    "##############################################################################\n",
    "###   Method1. py3Dmol   #####################################################\n",
    "##############################################################################\n",
    "\n",
    "import py3Dmol \n",
    "\n",
    "view1 = py3Dmol.view(query='pdb:6LU7')\n",
    "view1.setStyle({'cartoon':{'color':'spectrum'}})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "a5ec6d22688c4340b0e0d222c99d7c7d",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": []
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\Users\\evgen\\AppData\\Local\\Programs\\Python\\Python39\\lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:89: PDBConstructionWarning: WARNING: Chain A is discontinuous at line 2811.\n",
      "  warnings.warn(\n"
     ]
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "d3435242d65340e69307cdd801117eca",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "NGLWidget()"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "####################################################################################\n",
    "###   Method2. nglview   #################################################################\n",
    "####################################################################################\n",
    "\n",
    "import nglview as nv\n",
    "\n",
    "# nv.demo()\n",
    "\n",
    "from Bio.PDB import PDBParser, MMCIFParser\n",
    "\n",
    "# creating a parser\n",
    "parser = PDBParser()\n",
    "structure = parser.get_structure(\"6LU7\", \"Examples/6lu7.pdb\") \n",
    "\n",
    "view = nv.show_biopython(structure)\n",
    "view\n",
    "\n",
    "# option if you just want an image:\n",
    "# view.render_image()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Read .pdb file, make an enantiomer (just a mirror image) by multiplying each of X, Y, and Z coordinates by -1, and then save the pdb file\n",
    "\n",
    "\"\"\"Original file - view\"\"\"\n",
    "from Bio.PDB import *\n",
    "import nglview as nv\n",
    "\n",
    "parser = PDBParser(PERMISSIVE=1)\n",
    "structure = parser.get_structure(\"Helix\", \"helix (2).pdb\")\n",
    "view = nv.show_biopython(structure)\n",
    "view\n",
    "\n",
    "\n",
    "\n",
    "\"\"\"Change coordinates and then save the file\"\"\"\n",
    "from Bio.PDB import *\n",
    "import nglview as nv\n",
    "\n",
    "parser = PDBParser(PERMISSIVE=1)\n",
    "structure = parser.get_structure(\"Helix\", \"helix (2).pdb\")\n",
    "view = nv.show_biopython(structure)\n",
    "# view\n",
    "for model in structure:\n",
    "\tfor chain in model:\n",
    "\t\tfor residue in chain:\n",
    "\t\t\tfor atom in residue:\n",
    "\t\t\t\t# print(atom.get_coord())\n",
    "\t\t\t\t# print(atom.get_vector())\n",
    "\t\t\t\tatom.set_coord(atom.get_coord() * -1)\n",
    "\t\t\t\t# print(atom.get_coord())\n",
    "view = nv.show_biopython(structure); view\n",
    "\n",
    "io = PDBIO()\n",
    "io.set_structure(structure)\n",
    "io.save('new_structure.pdb')\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## BLAST"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "####################################################################################\n",
    "###   BLAST   ######################################################################\n",
    "####################################################################################\n",
    "\n",
    "from Bio.Blast import NCBIWWW\n",
    "\n",
    "result_handle = NCBIWWW.qblast(\"blastn\", \"nt\", 'MT747438')\n",
    "\n",
    "\n",
    "# save the result in a file\n",
    "with open(\"blast_BE_isolate.xml\", \"w+\") as save_to:\n",
    "\tsave_to.write(result_handle.read())\n",
    "\tresult_handle.close()\n",
    "\n",
    "# look at the BLAST results\n",
    "from Bio.Blast import NCBIXML\n",
    "\n",
    "result_handle = open(\"blast_BE_isolate.xml\", \"r\")\n",
    "blast_record = NCBIXML.read(result_handle)\n",
    "\n",
    "E_VALUE_THRESH = 0.0001\n",
    "ct = 0\n",
    "for alignment in blast_record.alignments:\n",
    "\tfor hsp in alignment.hsps:\n",
    "\t\tct += 1\n",
    "\t\tif hsp.expect < E_VALUE_THRESH:\n",
    "\t\t\tprint(\"\\n\")\n",
    "\t\t\tprint(\"****Alignment****\")\n",
    "\t\t\tprint(\"sequence:\", alignment.title)\n",
    "\t\t\tprint(\"lenght:\", alignment.length)\n",
    "\t\t\tprint(\"E value:\", hsp.expect)\n",
    "\t\t\tprint(hsp.query[0:75] + '...')\n",
    "\t\t\tprint(hsp.match[0:75] + '...')\n",
    "\t\t\tprint(hsp.sbjct[0:75] + '...')\n",
    "\n",
    "print(\"\\n\" + \"There are \",ct,\"sequences in the BLAST output!\")\n",
    "\n",
    "\"\"\"\n",
    "with NCBIWWW.qblast(\"blastn\", \"nt\", 'rcsb_pdb_7ACT.fasta') as result_handle:\n",
    "\twith open(\"result_blast_covid.xml\", \"w\") as xml_file:\n",
    "\t\txml_file.write(result_handle)\n",
    "\n",
    "with open(\"result_covid_blast.xml\", \"w\") as f:\n",
    "\tfor i in result_handle:\n",
    "\t\tf.write(i)\n",
    "\"\"\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# ANALYSING COVID 19 SEQUENCE\n",
    "\n",
    "# ID MT385497.1\n",
    "# https://www.ncbi.nlm.nih.gov/nuccore/MT385497\n",
    "\n",
    "### Outline\n",
    "# Reading\n",
    "# Sequence analysis\n",
    "# Freq N, AA, GC and AT content\n",
    "# Protein synthesis\n",
    "\n",
    "from Bio import SeqIO\n",
    "\n",
    "# Read our FASTA file\n",
    "covid_record = SeqIO.read(\"Examples/covid_sequence_MT385497.fasta\", \"fasta\")\n",
    "\n",
    "covid_record\n",
    "\n",
    "covid_dna = covid_record.seq\n",
    "covid_dna\n",
    "\n",
    "# Length of DNA\n",
    "len(covid_dna)\n",
    "\n",
    "# Transcription = DNA -> mRNA\n",
    "covid_mrna = covid_dna.transcribe()\n",
    "covid_mrna\n",
    "\n",
    "# Translation = mRNA -> polypeptide\n",
    "covid_protein = covid_mrna.translate()\n",
    "covid_protein\n",
    "\n",
    "# Sequence analysis\n",
    "# Freq of N/Bases and Amino Acids\n",
    "\n",
    "# check for the N or bases\n",
    "## using counter module\n",
    "from collections import Counter\n",
    "covid_ncount = Counter(covid_dna)\n",
    "covid_ncount\n",
    "\n",
    "# plot\n",
    "import matplotlib.pyplot as plt\n",
    "# plt.bar(covid_ncount.keys(), covid_ncount.values())\n",
    "\n",
    "# Molecular weight\n",
    "from Bio.SeqUtils import molecular_weight\n",
    "# to determine molecular weight, first need to strip the sequence of the 'N' characters\n",
    "\n",
    "molecular_weight(str(covid_dna).replace(\"N\", \"\"))\n",
    "\n",
    "### DNA composition\n",
    "# GC content\n",
    "# AT \n",
    "\n",
    "# GC content\n",
    "from Bio.SeqUtils import GC, GC123, GC_skew\n",
    "\n",
    "GC(covid_dna) \n",
    "# not very stable, as GC content is only 33%\n",
    "\n",
    "# AT content\n",
    "def at_content(seq):\n",
    "\treturn float(seq.count(\"A\") + seq.count(\"T\"))/len(seq) * 100\n",
    "\n",
    "at_content(covid_dna)\n",
    "\n",
    "# GC skew\n",
    "# lagging or leading strand\n",
    "\n",
    "\n",
    "\n",
    "GC_skew(covid_dna)\n",
    "\n",
    "\n",
    "from Bio.SeqUtils import MeltingTemp as mt\n",
    "\n",
    "mt.Tm_GC(covid_dna, strict=False)\n",
    "\n",
    "# Check the number of amino acids\n",
    "\n",
    "covid_protein\n",
    "\n",
    "covid_pr_freq = Counter(covid_protein)\n",
    "plt.bar(covid_pr_freq.keys(), covid_pr_freq.values())\n",
    "\n",
    "# freq\n",
    "Counter(covid_protein).most_common(5)\n",
    "## IMPORTANT FOR COUNTER METHOD !!!!!\n",
    "\n",
    "# longest seq AA before a stop codon\n",
    "\n",
    "covid_clean = covid_protein.split(\"*\")\n",
    "covid_clean = [str(i) for i in covid_clean]\n",
    "covid_clean\n",
    "\n",
    "import pandas as pd\n",
    "\n",
    "df = pd.DataFrame({\"amino_acids\":covid_clean})\n",
    "df['count'] = df['amino_acids'].str.len()\n",
    "df.head()\n",
    "\n",
    "df.nlargest(10, \"count\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Biotite"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['A', 'C', 'G', 'T']\n",
      "TAGACGTA\n",
      "M MET\n",
      "I ILE\n",
      "T THR\n"
     ]
    }
   ],
   "source": [
    "import biotite.sequence as seq\n",
    "\n",
    "# show methods\n",
    "dir(seq)\n",
    "\n",
    "# general sequence\n",
    "dna = seq.NucleotideSequence(\"ATCTGCAT\")\n",
    "# get all the letters in my sequence\n",
    "print(dna.alphabet)\n",
    "# dna.get_alphabet()\n",
    "\n",
    "# frequencies of nucleotides in dna sequence\n",
    "dna.get_symbol_frequency()\n",
    "\n",
    "# DNA complement\n",
    "var = dna.complement(); print(var)\n",
    "\n",
    "# reverse complement\n",
    "dna.reverse().complement()\n",
    "\n",
    "# create protein sequence\n",
    "protein_seq = seq.ProteinSequence(\"MIT\")\n",
    "\n",
    "# AA frequency count\n",
    "protein_seq.get_symbol_frequency()\n",
    "\n",
    "# convert AA (1 --> 3 codons)\n",
    "def convert_protein_seq_1to3 (sequence):\n",
    "\tempty_list = []\n",
    "\tfor aa in sequence:\n",
    "\t\tprint(aa, seq.ProteinSequence.convert_letter_1to3(aa))\n",
    "\n",
    "convert_protein_seq_1to3(protein_seq)\n",
    "\n",
    "# create a protein sequence\n",
    "protein_seq2 = seq.ProteinSequence(\"MetIleThr\")\n",
    "\n",
    "# convert 3--> 1 codons\n",
    "# for aa in protein_seq2:\n",
    "# \tprint(aa, seq.ProteinSequence.convert_letter_1to3(aa))\n",
    "# \tDOESN'T WORK!!!\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[2]\n",
      "[2 8]\n",
      "2\n",
      "8\n",
      "MIS*\n"
     ]
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAXQAAAD4CAYAAAD8Zh1EAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjQuMywgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy/MnkTPAAAACXBIWXMAAAsTAAALEwEAmpwYAAAOv0lEQVR4nO3df4gnd33H8eerlzNaFENzi4b7kbU1pajVaJao2NI0Vog/yClGuEDViHIgBhWkbeIfEfNP9R9TNGI4TPBMrYkkKquelUAiGqhn9q6X6OW0bP1B7ohkvdOLQY09efePncj22937znf3u/e9/fB8wLAz83nvzPsG7rWzszPfSVUhSdr4/mjSDUiSxsNAl6RGGOiS1AgDXZIaYaBLUiPOmdSOt2zZUtPT05PavSRtSAcOHPh5VU0tNzaxQJ+enmZubm5Su5ekDSnJT1ca85KLJDXCQJekRhjoktQIA12SGmGgS1IjDHRJakTvQE+yKcl/JvnqMmPnJrkzyXyS/Ummx9qlJGmoUc7Q3wccWWHsncAvqur5wE3AR9famCRpNL0CPck24PXAp1co2Qns7ebvAl6dJGtvT5LUV98nRf8F+EfgWSuMbwUeAaiqU0lOAucDP19alGQ3sBtgx44dq2h30fR1X1v197bgJx95/aRbkNbE/8Pr83946Bl6kjcAj1XVgbXurKr2VNVMVc1MTS37UQSSpFXqc8nlVcCVSX4C3AFcnuRfB2qOAdsBkpwDPBs4PsY+JUlDDA30qrq+qrZV1TSwC7i3qv5+oGwWeHs3f1VX48tKJekMWvWnLSa5EZirqlngVuD2JPPACRaDX5J0Bo0U6FX1TeCb3fwNS9b/FnjLOBuTJI3GJ0UlqREGuiQ1wkCXpEYY6JLUCANdkhphoEtSIwx0SWqEgS5JjTDQJakRBrokNcJAl6RGGOiS1AgDXZIaYaBLUiMMdElqhIEuSY3o85Lopyf5bpIHkxxO8uFlaq5JspDkUDe9a33alSStpM8bi54ELq+qJ5JsBu5P8vWq+s5A3Z1Vde34W5Qk9TE00LuXPT/RLW7uJl8ALUlnmV7X0JNsSnIIeAy4p6r2L1P25iQPJbkryfZxNilJGq5XoFfV76vqYmAbcGmSFw2UfAWYrqoXA/cAe5fbTpLdSeaSzC0sLKyhbUnSoJHucqmqXwL3AVcMrD9eVU92i58GLlnh+/dU1UxVzUxNTa2iXUnSSvrc5TKV5Lxu/hnAa4AfDNRcsGTxSuDIGHuUJPXQ5y6XC4C9STax+APgC1X11SQ3AnNVNQu8N8mVwCngBHDNejUsSVpen7tcHgJeusz6G5bMXw9cP97WJEmj8ElRSWqEgS5JjTDQJakRBrokNcJAl6RGGOiS1AgDXZIaYaBLUiMMdElqhIEuSY0w0CWpEQa6JDXCQJekRhjoktQIA12SGmGgS1IjDHRJakSfd4o+Pcl3kzyY5HCSDy9Tc26SO5PMJ9mfZHpdupUkrajPGfqTwOVV9RLgYuCKJK8YqHkn8Iuqej5wE/DRsXYpSRpqaKDXoie6xc3dVANlO4G93fxdwKuTZGxdSpKG6nUNPcmmJIeAx4B7qmr/QMlW4BGAqjoFnATOX2Y7u5PMJZlbWFhYU+OSpP+rV6BX1e+r6mJgG3BpkhetZmdVtaeqZqpqZmpqajWbkCStYKS7XKrql8B9wBUDQ8eA7QBJzgGeDRwfQ3+SpJ763OUyleS8bv4ZwGuAHwyUzQJv7+avAu6tqsHr7JKkdXROj5oLgL1JNrH4A+ALVfXVJDcCc1U1C9wK3J5kHjgB7Fq3jiVJyxoa6FX1EPDSZdbfsGT+t8BbxtuaJGkUPikqSY0w0CWpEQa6JDXCQJekRhjoktQIA12SGmGgS1IjDHRJaoSBLkmNMNAlqREGuiQ1wkCXpEYY6JLUCANdkhphoEtSIwx0SWqEgS5JjejzTtHtSe5L8nCSw0net0zNZUlOJjnUTTcsty1J0vrp807RU8AHqupgkmcBB5LcU1UPD9R9u6reMP4WJUl9DD1Dr6pHq+pgN/8r4Aiwdb0bkySNZqRr6EmmWXxh9P5lhl+Z5MEkX0/ywhW+f3eSuSRzCwsLo3crSVpR70BP8kzgbuD9VfX4wPBB4MKqegnwCeDLy22jqvZU1UxVzUxNTa2yZUnScnoFepLNLIb556rqi4PjVfV4VT3Rze8DNifZMtZOJUmn1eculwC3Akeq6mMr1Dy3qyPJpd12j4+zUUnS6fW5y+VVwFuB7yU51K37ILADoKpuAa4C3p3kFPAbYFdV1fjblSStZGigV9X9QIbU3AzcPK6mJEmj80lRSWqEgS5JjTDQJakRBrokNcJAl6RGGOiS1AgDXZIaYaBLUiMMdElqhIEuSY0w0CWpEQa6JDXCQJekRhjoktQIA12SGmGgS1IjDHRJakSfd4puT3JfkoeTHE7yvmVqkuTjSeaTPJTkZevTriRpJX3eKXoK+EBVHUzyLOBAknuq6uElNa8FLuqmlwOf6r5Kks6QoWfoVfVoVR3s5n8FHAG2DpTtBD5bi74DnJfkgrF3K0laUZ8z9D9IMg28FNg/MLQVeGTJ8tFu3aMD378b2A2wY8eOEVvVOE1f97VJtzBRP/nI69f0/R6/tR0/rY/efxRN8kzgbuD9VfX4anZWVXuqaqaqZqamplazCUnSCnoFepLNLIb556rqi8uUHAO2L1ne1q2TJJ0hfe5yCXArcKSqPrZC2Szwtu5ul1cAJ6vq0RVqJUnroM819FcBbwW+l+RQt+6DwA6AqroF2Ae8DpgHfg28Y+ydSpJOa2igV9X9QIbUFPCecTUlSRqdT4pKUiMMdElqhIEuSY0w0CWpEQa6JDXCQJekRhjoktQIA12SGmGgS1IjDHRJaoSBLkmNMNAlqREGuiQ1wkCXpEYY6JLUCANdkhphoEtSI/q8U/S2JI8l+f4K45clOZnkUDfdMP42JUnD9Hmn6GeAm4HPnqbm21X1hrF0JElalaFn6FX1LeDEGehFkrQG47qG/sokDyb5epIXrlSUZHeSuSRzCwsLY9q1JAnGE+gHgQur6iXAJ4Avr1RYVXuqaqaqZqampsawa0nSU9Yc6FX1eFU90c3vAzYn2bLmziRJI1lzoCd5bpJ085d22zy+1u1KkkYz9C6XJJ8HLgO2JDkKfAjYDFBVtwBXAe9Ocgr4DbCrqmrdOpYkLWtooFfV1UPGb2bxtkZJ0gT5pKgkNcJAl6RGGOiS1AgDXZIaYaBLUiMMdElqhIEuSY0w0CWpEQa6JDXCQJekRhjoktQIA12SGmGgS1IjDHRJaoSBLkmNMNAlqREGuiQ1YmigJ7ktyWNJvr/CeJJ8PMl8koeSvGz8bUqShulzhv4Z4IrTjL8WuKibdgOfWntbkqRRDQ30qvoWcOI0JTuBz9ai7wDnJblgXA1KkvoZxzX0rcAjS5aPduv+nyS7k8wlmVtYWBjDriVJTzmjfxStqj1VNVNVM1NTU2dy15LUvHEE+jFg+5Llbd06SdIZNI5AnwXe1t3t8grgZFU9OobtSpJGcM6wgiSfBy4DtiQ5CnwI2AxQVbcA+4DXAfPAr4F3rFezkqSVDQ30qrp6yHgB7xlbR5KkVfFJUUlqhIEuSY0w0CWpEQa6JDXCQJekRhjoktQIA12SGmGgS1IjDHRJaoSBLkmNMNAlqREGuiQ1wkCXpEYY6JLUCANdkhphoEtSIwx0SWpEr0BPckWSHyaZT3LdMuPXJFlIcqib3jX+ViVJp9PnnaKbgE8CrwGOAg8kma2qhwdK76yqa9ehR0lSD33O0C8F5qvqR1X1O+AOYOf6tiVJGlWfQN8KPLJk+Wi3btCbkzyU5K4k25fbUJLdSeaSzC0sLKyiXUnSSsb1R9GvANNV9WLgHmDvckVVtaeqZqpqZmpqaky7liRBv0A/Biw9497WrfuDqjpeVU92i58GLhlPe5KkvvoE+gPARUmel+RpwC5gdmlBkguWLF4JHBlfi5KkPobe5VJVp5JcC3wD2ATcVlWHk9wIzFXVLPDeJFcCp4ATwDXr2LMkaRlDAx2gqvYB+wbW3bBk/nrg+vG2JkkahU+KSlIjDHRJaoSBLkmNMNAlqREGuiQ1wkCXpEYY6JLUCANdkhphoEtSIwx0SWqEgS5JjTDQJakRBrokNcJAl6RGGOiS1AgDXZIaYaBLUiN6BXqSK5L8MMl8kuuWGT83yZ3d+P4k02PvVJJ0WkMDPckm4JPAa4EXAFcnecFA2TuBX1TV84GbgI+Ou1FJ0un1OUO/FJivqh9V1e+AO4CdAzU7gb3d/F3Aq5NkfG1Kkobp85LorcAjS5aPAi9fqaaqTiU5CZwP/HxpUZLdwO5u8YkkP1xN02eBLQz8286ktPH7j8dwbTx+a7ORj9+FKw30CfSxqao9wJ4zuc/1kGSuqmYm3cdG5jFcG4/f2rR6/PpccjkGbF+yvK1bt2xNknOAZwPHx9GgJKmfPoH+AHBRkucleRqwC5gdqJkF3t7NXwXcW1U1vjYlScMMveTSXRO/FvgGsAm4raoOJ7kRmKuqWeBW4PYk88AJFkO/ZRv+stFZwGO4Nh6/tWny+MUTaUlqg0+KSlIjDHRJaoSBPqIkb0xSSf5i0r1sREmem+SOJP+d5ECSfUn+fNJ9bQRJnpPk35L8qDt2/5HkTZPua6NIcn6SQ930syTHliw/bdL9jYOBPrqrgfu7rxpB9/Twl4BvVtWfVdUlwPXAcybb2dmvO3ZfBr5VVX/aHbtdLN5GrB6q6nhVXVxVFwO3ADc9tdw9Bb/hGegjSPJM4K9Y/Oya1u/kWQ9/C/xPVd3y1IqqerCqvj3BnjaKy4HfDRy7n1bVJybYk84yBvpodgL/XlX/BRxPcsmkG9pgXgQcmHQTG9QLgYOTbkJnNwN9NFez+OFkdF+97KKJSPLJJA8meWDSvejscUY/y2UjS/InLP7a+5dJisWHrCrJP/hUbG+HWXySWKM7DLz5qYWqek+SLcDc5FrS2cYz9P6uAm6vqgurarqqtgM/Bv56wn1tJPcC53afuglAkhcn8RgOdy/w9CTvXrLujyfVjM5OBnp/V7N4h8ZSd+Nll96632TeBPxdd9viYeCfgZ9NtrOzX3fs3gj8TZIfJ/kui+8g+KeJNqazio/+S1IjPEOXpEYY6JLUCANdkhphoEtSIwx0SWqEgS5JjTDQJakR/wsBAtturWMK4QAAAABJRU5ErkJggg==",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "# sequence manipulation\n",
    "# convert to string first\n",
    "\n",
    "import biotite.sequence as seq\n",
    "\n",
    "dna = seq.NucleotideSequence(\"ATGATCTCGTAA\")\n",
    "\n",
    "# find position of base or nucleotides\n",
    "str(dna).find(\"A\")\n",
    "\n",
    "# count freq of nucleotide\n",
    "str(dna).count(\"G\")\n",
    "\n",
    "dna_freq = dna.get_symbol_frequency()\n",
    "\n",
    "# searching for subseq\n",
    "sub_seq = seq.NucleotideSequence(\"GAT\")\n",
    "print(seq.find_subsequence(dna, sub_seq))\n",
    "\n",
    "# find the occurrence of a nucleotide in sequence\n",
    "print(seq.find_symbol(dna, \"G\"))\n",
    "print(seq.find_symbol_first(dna, \"G\"))\n",
    "print(seq.find_symbol_last(dna, \"G\"))\n",
    "\n",
    "# plot of nucleotide frequency / count \n",
    "import matplotlib.pyplot as plt\n",
    "plt.bar(dna_freq.keys(), dna_freq.values())\n",
    "\n",
    "\n",
    "# transcription \n",
    "# dna to mRNA\n",
    "\n",
    "\n",
    "# translate\n",
    "protein = dna.translate()\n",
    "protein = dna.translate(complete=True)\n",
    "print(protein)\n",
    "\n",
    "# amino acids in codontable\n",
    "# official NCBI\n",
    "# print(seq.CodonTable.load(1))\n",
    "\n",
    "# NCBI biotite\n",
    "# print(seq.CodonTable.default_table())\n",
    "\n",
    "\n",
    "# search for the amino acid combination\n",
    "table = seq.CodonTable.default_table()\n",
    "type(table)\n",
    "table['GAA']\n",
    "table['E']\n",
    "\n",
    "\n",
    "# check for codon table of other species\n",
    "# for bacterial genome\n",
    "bacterial_tb = seq.CodonTable.load(11)\n",
    "# print(bacterial_tb)\n",
    "\n",
    "# for yeast mitochondrial \n",
    "# using name\n",
    "yeast_tb = seq.CodonTable.load(\"Yeast Mitochondrial\")\n",
    "# print(yeast_tb)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ITTITE\n",
      "ITTITE\n",
      "MITTITE\n",
      "-ITTITE\n",
      "Score 28\n",
      "Recalc Score 28\n",
      "Sequence Identity 1.0\n",
      "ATCTCGTA\n",
      "ATCTCGTA\n",
      "ATGATCTCGTAA\n",
      "AT---CTCGTAG\n",
      "ATGATCTCGTAA\n",
      "A---TCTCGTAG\n",
      "ATGATCTCGTAA\n",
      "---ATCTCGTAG\n",
      "Score 6\n",
      "Recalc Score 6\n",
      "Sequence Identity 0.6666666666666666\n"
     ]
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAIgAAAAyCAYAAACH65NBAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjQuMywgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy/MnkTPAAAACXBIWXMAAAsTAAALEwEAmpwYAAABoklEQVR4nO3aP0oDQRiG8W93TSBESBDBoIVu4Q0EsfMPpLIWPYB4A09iY+kdrJTYWnkEUQshlWIbSNYiEETiu2Ezuxl3n19lInw7gYeZiBMkSWLAX8JFLwB+IxBIBAKJQCARCCQCgbSkftntxX0zW3P8zJG5DdP3eXnMdL7Gu6OXYNr7aQ9xHccszyzbvDxmFrbzc8RAIhBI8jvIrB6O3+zwdnPuOYOvoT1d9sc/fwwtCAOrtccN716tW1ibekwWNu8nXz+zmdl999WW49rkdWe/afFZO9P6nATiSr0V2d71hpmZPd98WtQIbeuk5c28POSxxqgeTGbOiyMGklc7CNwYDhJ7vHifvI5P29Y5aGaaRSAlxBGDwhAIJI6YEvr9HWR1p2Hb5yuZZgXqymG3F3MfsSKy/i8GFUcgkAgEEoFAIhBIaYGMClkFvJUWCDtMxREAJAKBxI2yjHz9zGbcKFvYvDxwowz/mlc7CNzgRhkkjhgUhkAgccSUEDfK4Bw3ypAJgUAiEEgEAolAIMm/YgB2EEgEAolAIBEIJAKBRCCQvgHyFZpewzhusgAAAABJRU5ErkJggg==",
      "text/plain": [
       "<Figure size 144x57.6 with 2 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAIgAAAAyCAYAAACH65NBAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjQuMywgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy/MnkTPAAAACXBIWXMAAAsTAAALEwEAmpwYAAAChklEQVR4nO3azasSURgG8Gc0L4jBHSLomkK5aKnM4sLQrg+4iyHcFYo7IVoMuJL+kly0DITxD9BNUg1tWrURt5EtEjcWdxHKzY8WokToGWbOuc1cfX4g+AHvvMLDOfPxasvlEkS7xMJugKKNASEhBoSEGBASYkBIiAEhoWsev48A3JI9yNOPp4vz32PZMC6gJtAq6uxdL28ff9W2fe9VWDocAKAgHIC61Y69RKU4XX0MCAl5nYNspWkaKpUKms0mAGA2myGdTsM0TbTbbd/13j/5hkftO0FawcX5HJ9fjlbvf8yhxTQk9FXuzVe3EUts3VqV1/hb2P9nrXs2wPVcYvP55EEKubLuq59AAUmlUuj3+5hMJkgmk+h2u8hkMkFKSTs6juP+69Wxv7z5iXgyhrvPjv97DVVU9hI/0ja1ggq8xViWhU6nAwBwHAflclmqEYqmwAEplUpotVqYTqfo9XowTVNlX6TA/GKJTy++b16jD7981wi0xQBAoVDAYDCA4ziwLCtoGbpEKraYwAEBgGKxiHq9Dtd1MR6PpRqhaJIKSLVaha7ryOfzcF1XUUsUJVIByWazqNVqqnohxdbnIGs3T5O49/yGrxqax8ihknnEs3c5FWXoEgV9FkMHjgEhIQaEhBgQEmJASMgrIAsVB9Hg72koRYdXQJSsMEs1V8sUAm4xJMSAkJDUrXZVwp7A4kTZbpEIiAxOlO0W6kQZHYYrv4LQbv8+zc2VdJw8TPmqwYDssdC3mEajAcMwYBgGhsOhVCMUTVIriG3bsG1bVS8UQdxi9hgnykgZTpRRIAwICTEgJMSAkBADQkJeVzF04LiCkBADQkIMCAkxICTEgJAQA0JCfwCWIuRJf1D8kQAAAABJRU5ErkJggg==",
      "text/plain": [
       "<Figure size 144x57.6 with 2 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAIgAAAAyCAYAAACH65NBAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjQuMywgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy/MnkTPAAAACXBIWXMAAAsTAAALEwEAmpwYAAAGS0lEQVR4nO2cX0xTZxjGn69/jrCDaEFAmIiNk+mYUh24jI0lwkK8WWJjNkg20ostxkg0ISHcbJmQXQnZzbJKlswluxh1RhMkuzKhTGHOP4FVGSM6nRCUtALFIq2l0J5dkBZKy9fvnHbW2O93Rc95+7497/twvvOlDxBJksDhrIUq2R+A82LDBcKhwgXCocIFwqHCBcKhwgXCoaKhnazp0dsB5DHkCYBRbOcqbwY2CptYYllzMsV9dKUs4FqYTmRdObFJiZPRawAg0Q7GejOLOFjyhJDxgRMaxygOOXXlxCYlTkav4yvESV24QDhUqM8gq3n8uxu3Tk6i4scCiFuFiPM+lx8Dzfaln51+EBWBduOSBt/+rgAq7fIy53A40NjYiGvXrkGn00EQBDQ3N8NoNEat3dXVBaPRiJGREezcuTPi/PT0NKqrqwEAdrsdarUaOTk5AIAbN26Exc47F3HntBOzd3zQZKgg6NR4/VgWxC1aRdfCmhMA5mf8uHvaCdfIPDTrVVBpCLbVZiL3PVFx7VhzYe1hNAjtu5iaHn3YydtfP8b8tB9Ze9Ow3aSjJr7/0wzU6Sps+3hD2PFL1Q8gSRIqKipgMplw9OhRAMDY2Bi6u7tx/PjxqPlqa2sxMTGBqqoqtLa2Umu3tLQgIyMDTU1NK68FACBJEm6esCO/RkThh5kAgKf3fVj0BKDbnSbrWoKw5owW98yxiMmrHmw1ZiqqDaw9l0vVD8LiYvRQ0UNqiMVnATz5ax5vNG2CvdfN+raoWK1WCIIQEgcAFBUVrSmOubk59Pf348yZMzh79mxctWdsXhANQgMCgPXbhTXFkciczj+9INrwuPQ8zZriYIF1Lkp7yCyQyaseZJenQ9yihTZTjdm788xFVjM8PIx9+/Yxx1+8eBEHDx5EcXExsrOzMTAwoLj23OgCMnesU/z+eHK6xxaQ+Vpia7PORWkPmQVit7qx+cDSOrn5gBj3XWQlDQ0NKC0tRXl5edTzFosFdXV1AIC6ujpYLJaE1U4mI99O448jj3D92ITiHKxzUdpDpofUhVk/nDYv5h74AAJIgaUFa8cRHQiJunRRKSkpwYULF0KvzWYzpqamUFZWFhHrdDphtVoxNDQEQgj8fj8IIWhvb1dUWyzSwnElceKWk1Ms0sLRtxy360Q2fC6/YoGwziWeHjLdQRxXPMj/QERlZyEqfy7E+5ZCpOVr8GRI2TJTVVUFr9eLjo6O0DGPxxM19vz586ivr8fY2BhGR0cxPj4OvV6Pvr4+RbWz9qYhsCDh4a9PQ8ee/uvDzJBXUT45ObP2piHgkzDePRs65vcqN2yxziWeHjIJxN7rRu67r4Qdy6sUYbcq+00khKCrqwuXL1+GXq/H/v37YTKZcOrUqYhYi8USsfU9fPiw4mWGEILSllw4B5+hv/4hrn72CPd+mIGgUyvKJycnIQSG1lzM3Pai79OHuN4wgeG2Kez4PEtRXda5xNNDWdvcRLB66/W8CG5zUwmZvY5vm8tJTbhAOFS4QDhUuEA4VLhAOFRiCSSQ6IIBKeEpmSDRH9JfahLR61gCSfgdRkWSc9OSkHp/QZiIXvMlhkOFC4RDJWmOsmSSio4yuQ6+ILIEYre6sfHNdbD3urHdFPlBhA1qvPP9qwDYnFDJQJIk3Do5ifwaEXu+zAWw5P7yzfjDhinnWlhzSpKEW189Rn6NiN1fLNkhg46ylcjtY6y5SJKEQ4cOwWQyobOzE8Cygy8WSXGUJZNUdJTJdfCtJCmOsmSSio4yuQ6+lbwQjrJU5Xk6ylYSy8G3EiaBBJ1Lf38zhb5PxjF6zgXHb27E+9+JzGYzDAYDDAYDJiboTWKNjRUnFmkx+09i736sOcUiLWbvLcftOpGNt9o3w+fyK6rLOpeSkhIMDg6GXpvNZvT09GBycjJmjaQ4yoI0NDTAZrPBZrOhoKAgIbGx4lLRUSbHwbcapl2MvdeNbbXhD1JB55Juj/KHu2QQdH/dPe3E6C8uqASC9DwNio8pc3XJyRl0lN3pcGL03CyEDSqo01RxOcpY5hJ08DU2NqKtrQ05OTkQRTGqgy/i2rij7OWFO8o4/ztcIBwqXCAcKlwgHCpcIBwq1F0Mh8PvIBwqXCAcKlwgHCpcIBwqXCAcKlwgHCr/AX5L7IN9oNn0AAAAAElFTkSuQmCC",
      "text/plain": [
       "<Figure size 144x57.6 with 2 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "# global alignment - optimisation that forces the alignment to span the entire length of all query sequences\n",
    "\n",
    "# local alignment: \n",
    "# local alignments identify regions of similarity within long sequences that are often widely divergent overall\n",
    "# local alignments are often preferable, but can be more difficult to calculate because of the additional challenge of identifying the regions of similarity\n",
    "\n",
    "# biotite align\n",
    "# function align_optimal()\n",
    "# performs an optimal global alignment using Needleman-Wunsch algorithm\n",
    "# performs an optimal local alignment using Smith-Waterman algorithm\n",
    "\n",
    "import biotite.sequence.align as align\n",
    "\n",
    "# method\n",
    "# dir(align)\n",
    "\n",
    "# create our sub matrix\n",
    "# standard submatrix\n",
    "n_matrix = align.SubstitutionMatrix.std_nucleotide_matrix()\n",
    "# print(n_matrix)\n",
    "\n",
    "p_matrix = align.SubstitutionMatrix.std_protein_matrix()\n",
    "# print(p_matrix)\n",
    "\n",
    "# load matrix from internal database\n",
    "alph = seq.ProteinSequence.alphabet\n",
    "matrix_bl50 = align.SubstitutionMatrix(alph, alph, \"BLOSUM50\")\n",
    "# print(matrix_bl50)\n",
    "\n",
    "pseq1 = seq.ProteinSequence(\"MITTITE\")\n",
    "pseq2 = seq.ProteinSequence(\"ITTITE\")\n",
    "\n",
    "\n",
    "# local alignment\n",
    "alignments = align.align_optimal(pseq1, pseq2, p_matrix, local=True)\n",
    "\n",
    "for a in alignments:\n",
    "\tprint(a)\n",
    "\n",
    "\n",
    "gl_alignments = align.align_optimal(pseq1, pseq2, p_matrix, local=False)\n",
    "\n",
    "for a in gl_alignments:\n",
    "\tprint(a)\n",
    "\n",
    "\n",
    "import biotite.sequence.graphics as graphics\n",
    "\n",
    "# local alignment visualisation\n",
    "fig, ax = plt.subplots(figsize=(2.0, 0.8))\n",
    "graphics.plot_alignment_similarity_based(\n",
    "\tax, alignments[0], matrix=p_matrix, symbols_per_line=len(alignments[0])\n",
    ")\n",
    "fig.tight_layout()\n",
    "\n",
    "# global alignment visualisation\n",
    "fig, ax = plt.subplots(figsize=(2.0, 0.8))\n",
    "graphics.plot_alignment_similarity_based(\n",
    "\tax, gl_alignments[0], matrix=p_matrix, symbols_per_line=len(gl_alignments[0])\n",
    ")\n",
    "fig.tight_layout()\n",
    "\n",
    "# alignment score\n",
    "loc_alignments = alignments[0]\n",
    "print(\"Score\", loc_alignments.score)\n",
    "print(\"Recalc Score\", align.score(loc_alignments, matrix=p_matrix))\n",
    "print(\"Sequence Identity\", align.get_sequence_identity(loc_alignments))\n",
    "\n",
    "# Nucleotide sequence alignment\n",
    "dna_seq1 = seq.NucleotideSequence(\"ATGATCTCGTAA\")\n",
    "dna_seq2 = seq.NucleotideSequence(\"ATCTCGTAG\")\n",
    "\n",
    "# Local alignment of nucleotides\n",
    "dna_alignments = align.align_optimal(dna_seq1, dna_seq2, n_matrix, local=True)\n",
    "for a in dna_alignments:\n",
    "\tprint(a)\n",
    "\n",
    "# global alignment of nucleotides\n",
    "dna_glb_alignments = align.align_optimal(dna_seq1, dna_seq2, n_matrix, local=False)\n",
    "for a in dna_glb_alignments:\n",
    "\tprint(a)\n",
    "\n",
    "\n",
    "# global alignment of nucleotides\n",
    "fig, ax = plt.subplots(figsize=(2.0, 0.8))\n",
    "graphics.plot_alignment_similarity_based(\n",
    "\tax, dna_glb_alignments[0], matrix=n_matrix, symbols_per_line=len(dna_glb_alignments[0])\n",
    ")\n",
    "fig.tight_layout()\n",
    "\n",
    "# alignment score for nucleotides global alignment\n",
    "glob_alignments = dna_glb_alignments[0]\n",
    "print(\"Score\", glob_alignments.score)\n",
    "print(\"Recalc Score\", align.score(glob_alignments, matrix=n_matrix))\n",
    "print(\"Sequence Identity\", align.get_sequence_identity(glob_alignments))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "LOCUS       KU182909               18959 bp    cRNA    linear   VRL 22-DEC-2015\n",
      "DEFINITION  Ebola virus isolate Ebola virus/H.\n",
      "            sapiens-tc/COD/1995/Kikwit-9510622, complete genome.\n",
      "ACCESSION   KU182909\n",
      "VERSION     KU182909.1\n",
      "DBLINK      BioProject: PRJNA242964\n",
      "KEYWORDS    .\n",
      "SOURCE      Ebola virus (EBOV)\n",
      "  ORGANISM  Ebola virus\n",
      "            Viruses; Riboviria; Orthornavirae; Negarnaviricota;\n",
      "            Haploviricotina; Monjiviricetes; Mononegavirales; Filoviridae;\n",
      "            Ebolavirus.\n",
      "REFERENCE   1  (bases 1 to 18959)\n",
      "  AUTHORS   Das,S.R., Shabman,R., Halpin,R.A., Akopov,A., Fedorova,N., Puri,V.,\n",
      "            Stockwell,T., Amedeo,P., Bishop,B., Katzel,D., Schobel,S.,\n",
      "            Shrivastava,S., Brasel,T., Yun,N., Paessler,S., Dowling,W. and\n",
      "            Barrett,A.\n",
      "  TITLE     Direct Submission\n",
      "  JOURNAL   Submitted (24-NOV-2015) J. Craig Venter Institute, 9704 Medical\n",
      "            Center Drive, Rockville, MD 20850, USA\n",
      "COMMENT     This work was supported by the National Institute of Allergy and\n",
      "            Infectious Diseases (NIAID), Genome Center for Infectious Diseases\n",
      "            (GCID) program and Department of Homeland Security (DHS).\n",
      "            Sequence data was generated using next-generation sequencing\n",
      "            platforms. Primers and barcodes were trimmed before assembly unless\n",
      "            otherwise stated in the feature section.\n",
      "            \n",
      "            ##Genome-Assembly-Data-START##\n",
      "            Current Finishing Status :: Finished\n",
      "            Assembly Method          :: clc_ref_assemble_long v. 3.22.55705\n",
      "            Genome Coverage          :: 2573.4x\n",
      "            Sequencing Technology    :: Sanger;Illumina;Ion Torrent\n",
      "            ##Genome-Assembly-Data-END##\n",
      "FEATURES             Location/Qualifiers\n",
      "     source          1..18959\n",
      "                     /organism=\"Ebola virus\"\n",
      "                     /mol_type=\"viral cRNA\"\n",
      "                     /isolate=\"Ebola virus/H.\n",
      "                     sapiens-tc/COD/1995/Kikwit-9510622\"\n",
      "                     /host=\"Homo sapiens; female; 65 years\"\n",
      "                     /db_xref=\"taxon:1570291\"\n",
      "                     /lab_host=\"Vero E6 cells\"\n",
      "                     /country=\"Zaire\"\n",
      "                     /collection_date=\"04-May-1995\"\n",
      "                     /PCR_primers=\"fwd_name: 3RACE_Forward, fwd_seq:\n",
      "                     ccaaaacgccatttccaccttctcttc, rev_name: 3RACE_Reverse,\n",
      "                     rev_seq: ccgcactctcttgaaaatcaacacc\"\n",
      "                     /PCR_primers=\"fwd_name: 5RACE_Forward, fwd_seq:\n",
      "                     ccttctttacaatatagcagactagataataat, rev_name:\n",
      "                     5RACE_Reverse, rev_seq: ccaaaacgccatttccaccttctcttc\"\n",
      "                     /PCR_primers=\"fwd_name: 3_5_self-ligation_Forward,\n",
      "                     fwd_seq: gtgttaatgataattaagrcattgacc, rev_name:\n",
      "                     3_5_self-ligation_Reverse, rev_seq: aagggattttcaactgagca\"\n",
      "                     /PCR_primers=\"fwd_name: Zaire_1a_Forward, fwd_seq:\n",
      "                     ttgtgtgcgaataactatgagg, rev_name: Zaire_1a_Reverse,\n",
      "                     rev_seq: caagtgttttaccatcttgaagc\"\n",
      "                     /PCR_primers=\"fwd_name: Zaire_1b_Forward, fwd_seq:\n",
      "                     tcctctcattgaaatttatatcgg, rev_name: Zaire_1b_Reverse,\n",
      "                     rev_seq: ttcagcagttgaagcctctgc\"\n",
      "                     /PCR_primers=\"fwd_name: Zaire_2a_Forward, fwd_seq:\n",
      "                     gagttccaatcttccaagatgc, rev_name: Zaire_2a_Reverse,\n",
      "                     rev_seq: catttgtcaagtaatcttagattcaag\"\n",
      "                     /PCR_primers=\"fwd_name: Zaire_2b_Forward, fwd_seq:\n",
      "                     gacattccccgagcttgcc, rev_name: Zaire_2b_Reverse, rev_seq:\n",
      "                     ttaaargagttagaatcacattggc\"\n",
      "                     /PCR_primers=\"fwd_name: Zaire_3a_Forward, fwd_seq:\n",
      "                     tggataccggcaggtattgg, rev_name: Zaire_3a_Reverse, rev_seq:\n",
      "                     gtaacattgaaatygggatcttcc\"\n",
      "                     /PCR_primers=\"fwd_name: Zaire_3b_Forward, fwd_seq:\n",
      "                     agttatygctttattctgtatatgc, rev_name: Zaire_3b_Reverse,\n",
      "                     rev_seq: ttgaatactgatgcctgatacc\"\n",
      "                     /PCR_primers=\"fwd_name: Zaire_4a_Forward, fwd_seq:\n",
      "                     tagattaaatcgaggaaactctag, rev_name: Zaire_4a_Reverse,\n",
      "                     rev_seq: caactgcataatttataacattctgg\"\n",
      "                     /PCR_primers=\"fwd_name: Zaire_4b_Forward, fwd_seq:\n",
      "                     catgatgatttaatagacatcttagg, rev_name: Zaire_4b_Reverse,\n",
      "                     rev_seq: ttagttagatgaaggtgagcacg\"\n",
      "                     /PCR_primers=\"fwd_name: Zaire_5a_Forward, fwd_seq:\n",
      "                     gaaaccattcgtgtcagtggc, rev_name: Zaire_5a_Reverse,\n",
      "                     rev_seq: atttttaaatttttgtgtgcgacc\"\n",
      "                     /PCR_primers=\"fwd_name: Zaire_5b_Forward, fwd_seq:\n",
      "                     gataggacaacctgctattaaacc, rev_name: Zaire_5b_Reverse,\n",
      "                     rev_seq: tccaggaatcctttttgcaacg\"\n",
      "                     /PCR_primers=\"fwd_name: Zaire_2aV2_Forward, fwd_seq:\n",
      "                     actgctttccaccarttagtaca, rev_name: Zaire_2aV2_Reverse,\n",
      "                     rev_seq: argagttagaatcacattggct\"\n",
      "                     /PCR_primers=\"fwd_name: Zaire_2bV2_Forward, fwd_seq:\n",
      "                     atyattcatgctgagttccagg, rev_name: Zaire_2bV2_Reverse,\n",
      "                     rev_seq: caatgctttrtgtgtgttaggtt\"\n",
      "                     /PCR_primers=\"fwd_name: Zaire_AdapLig_II_Forward, fwd_seq:\n",
      "                     gtgttaatgataattaagrcattgacc, rev_name:\n",
      "                     Zaire_AdapLig_II_Reverse, rev_seq:\n",
      "                     gaagagaaggtggaaatggcgttttgg\"\n",
      "                     /PCR_primers=\"fwd_name: Zaire_AdapLig_III_Forward,\n",
      "                     fwd_seq: ccaaaacgccatttccaccttctcttc, rev_name:\n",
      "                     Zaire_AdapLig_III_Reverse, rev_seq:\n",
      "                     tgattacagtaacaatttcaatttaaattc\"\n",
      "                     /note=\"Replicate 2: Virus propagated in confluent Vero E6\n",
      "                     cell monolayer at 0.01 MOI using P2 material in 850 cm2\n",
      "                     roller bottle (n=1) without antibiotics. Supernatant\n",
      "                     collected upon reaching 70% CPE. Harvested supernatant was\n",
      "                     clarified via centrifugation (10,000 x g for 10 min).\n",
      "                     Clarified supernatant was supplemented with 16.5% HI-FBS\n",
      "                     and 0.01M Tris pH 8.5.\"\n",
      "     misc_feature    1..18959\n",
      "                     /db_xref=\"ViPR:FLVSP_ABEBO2_00005.MAIN\"\n",
      "     gene            470..2689\n",
      "                     /gene=\"NP\"\n",
      "                     /locus_tag=\"DF49_53416gpNP\"\n",
      "     CDS             470..2689\n",
      "                     /gene=\"NP\"\n",
      "                     /locus_tag=\"DF49_53416gpNP\"\n",
      "                     /codon_start=1\n",
      "                     /product=\"nucleoprotein\"\n",
      "                     /protein_id=\"ALT19749.1\"\n",
      "                     /translation=\"MDSRPQKVWMTPSLTESDMDYHKILTAGLSVQQGIVRQRVIPVY\n",
      "                     QVNNLEEICQLIIQAFEAGVDFQESADSFLLMLCLHHAYQGDYKLFLESGAVKYLEGH\n",
      "                     GFRFEVKKRDGVKRLEELLPAVSSGKNIKRTLAAMPEEETTEANAGQFLSFASLFLPK\n",
      "                     LVVGEKACLEKVQRQIQVHAEQGLIQYPTAWQSVGHMMVIFRLMRTNFLIKFLLIHQG\n",
      "                     MHMVAGHDANDAVISNSVAQARFSGLLIVKTVLDHILQKTERGVRLHPLARTAKVKNE\n",
      "                     VNSFKAALSSLAKHGEYAPFARLLNLSGVNNLEHGLFPQLSAIALGVATAHGSTLAGV\n",
      "                     NVGEQYQQLREAATEAEKQLQQYAESRELDHLGLDDQEKKILMNFHQKKNEISFQQTN\n",
      "                     AMVTLRKERLAKLTEAITAASLPKTSGHYDDDDDIPFPGPINDDDNPGHQDDDPTDSQ\n",
      "                     DTTIPDVVVDPDDGSYGEYQSYSENGMNAPDDLVLFDLDEDDEDTKPVPNRSTKGGQQ\n",
      "                     KNSQKGQHTEGRQTQSRPTQNVPGPHRTIHHASAPLTDNDRRNEPSGSTSPRMLTPIN\n",
      "                     EEADPLDDADDETSSLPPLESDDEEQDRDGTSNRTPTVAPPAPVYRDHSEKRELPQDE\n",
      "                     QQDQDHTQEARNQDSDNTQPEHSFEEMYRHILRSQGPFDAVLYYHMMKDEPVVFSTSD\n",
      "                     GKEYTYPDSLEEEYPPWLTEKEAMNEENRFVTLDGQQFYWPVMNHKNKFMAILQHHQ\"\n",
      "     gene            3129..4151\n",
      "                     /gene=\"VP35\"\n",
      "                     /locus_tag=\"DF49_53416gpVP35\"\n",
      "     CDS             3129..4151\n",
      "                     /gene=\"VP35\"\n",
      "                     /locus_tag=\"DF49_53416gpVP35\"\n",
      "                     /codon_start=1\n",
      "                     /product=\"polymerase cofactor VP35\"\n",
      "                     /protein_id=\"ALT19750.1\"\n",
      "                     /translation=\"MTTRTKGRGHTAATTQNDRMPGPELSGWISEQLMTGRIPVSDIF\n",
      "                     CDIENNPGLCYASQMQQTKPNPKTRNSQTQTDPICNHSFEEVVQTLASLATVVQQQTI\n",
      "                     ASESLEQRITSLENGLKPVYDMAKTISSLNRVCAEMVAKYDLLVMTTGRATATAAATE\n",
      "                     AYWAEHGQPPPGPSLYEESAIRGKIESRDETVPQSVREAFNNLDSTTSLTEENFGKPD\n",
      "                     ISAKDLRNIMYDHLPGFGTAFHQLVQVICKLGKDSNSLDIIHAEFQASLAEGDSPQCA\n",
      "                     LIQITKRVPIFQDAAPPVIHIRSRGDIPRACQKSLRPVPPSPKIDRGWVCVFQLQDGK\n",
      "                     TLGLKI\"\n",
      "     gene            4479..5459\n",
      "                     /gene=\"VP40\"\n",
      "                     /locus_tag=\"DF49_53416gpVP40\"\n",
      "     CDS             4479..5459\n",
      "                     /gene=\"VP40\"\n",
      "                     /locus_tag=\"DF49_53416gpVP40\"\n",
      "                     /codon_start=1\n",
      "                     /product=\"matrix protein VP40\"\n",
      "                     /protein_id=\"ALT19751.1\"\n",
      "                     /translation=\"MRRVILPTAPPEYMEAIYPVRSNSTIARGGNSNTGFLTPESVNG\n",
      "                     DTPSNPLRPIADDTIDHASHTPGSVSSAFILEAMVNVISGPKVLMKQIPIWLPLGVAD\n",
      "                     QKTYSFDSTTAAIMLASYTITHFGKATNPLVRVNRLGPGIPDHPLRLLRIGNQAFLQE\n",
      "                     FVLPPVQLPQYFTFDLTALKLITQPLPAATWTDDTPTGSNGALRPGISFHPKLRPILL\n",
      "                     PNKSGKKGNSADLTSPEKIQAIMTSLQDFKIVPIDPTKNIMGIEVPETLVHKLTGKKV\n",
      "                     TSKNGQPIIPVLLPKYIGLDPVAPGDLTMVITQDCDTCHSPASLPAVIEK\"\n",
      "     gene            6039..8068\n",
      "                     /gene=\"GP\"\n",
      "                     /locus_tag=\"DF49_53416gpGP\"\n",
      "     CDS             6039..8068\n",
      "                     /gene=\"GP\"\n",
      "                     /locus_tag=\"DF49_53416gpGP\"\n",
      "                     /exception=\"RNA editing\"\n",
      "                     /note=\"synonym: preGP; non-templated A inserted during\n",
      "                     transcription due to RNA editing\"\n",
      "                     /codon_start=1\n",
      "                     /product=\"envelope glycoprotein\"\n",
      "                     /protein_id=\"ALT19754.1\"\n",
      "                     /translation=\"MGVTGILQLPRDRFKRTSFFLWVIILFQRTFSIPLGVIHNSTLQ\n",
      "                     VSDVDKLVCRDKLSSTNQLRSVGLNLEGNGVATDVPSATKRWGFRSGVPPKVVNYEAG\n",
      "                     EWAENCYNLEIKKPDGSECLPAAPDGIRGFPRCRYVHKVSGTGPCAGDFAFHKEGAFF\n",
      "                     LYDRLASTVIYRGTTFAEGVVAFLILPQAKKDFFSSHPLREPVNATEDPSSGYYSTTI\n",
      "                     RYQATGFGTNETEYLFEVDNLTYVQLESRFTPQFLLQLNETIYTSGKRSNTTGKLIWK\n",
      "                     VNPEIDTTIGEWAFWETKKNLTRKIRSEELSFTAVSNRAKNISGQSPARTSSDPGTNT\n",
      "                     TTEDHKIMASENSSAMVQVHSQGREAAVSHLTTLATISTSPQPPTTKPGPDNSTHNTP\n",
      "                     VYKLDISEATQVEQHHRRTDNDSTASDTPPATTAAGPLKAENTNTSKGTDLLDPATTT\n",
      "                     SPQNHSETAGNNNTHHQDTGEESASSGKLGLITNTIAGVAGLITGGRRARREAIVNAQ\n",
      "                     PKCNPNLHYWTTQDEGAAIGLAWIPYFGPAAEGIYIEGLMHNQDGLICGLRQLANETT\n",
      "                     QALQLFLRATTELRTFSILNRKAIDFLLQRWGGTCHILGPDCCIEPHDWTKNITDKID\n",
      "                     QIIHDFVDKTLPDQGDNDNWWTGWRQWIPAGIGVTGVIIAVIALFCICKFVF\"\n",
      "     gene            6039..7133\n",
      "                     /gene=\"sGP\"\n",
      "                     /locus_tag=\"DF49_53416gpsGP\"\n",
      "     CDS             6039..7133\n",
      "                     /gene=\"sGP\"\n",
      "                     /locus_tag=\"DF49_53416gpsGP\"\n",
      "                     /codon_start=1\n",
      "                     /product=\"small secreted glycoprotein sGP\"\n",
      "                     /protein_id=\"ALT19753.1\"\n",
      "                     /translation=\"MGVTGILQLPRDRFKRTSFFLWVIILFQRTFSIPLGVIHNSTLQ\n",
      "                     VSDVDKLVCRDKLSSTNQLRSVGLNLEGNGVATDVPSATKRWGFRSGVPPKVVNYEAG\n",
      "                     EWAENCYNLEIKKPDGSECLPAAPDGIRGFPRCRYVHKVSGTGPCAGDFAFHKEGAFF\n",
      "                     LYDRLASTVIYRGTTFAEGVVAFLILPQAKKDFFSSHPLREPVNATEDPSSGYYSTTI\n",
      "                     RYQATGFGTNETEYLFEVDNLTYVQLESRFTPQFLLQLNETIYTSGKRSNTTGKLIWK\n",
      "                     VNPEIDTTIGEWAFWETKKTSLEKFAVKSCLSQLYQTEPKTSVVRVRRELLPTQGPTQ\n",
      "                     QLKTTKSWLQKIPLQWFKCTVKEGKLQCRI\"\n",
      "     gene            6039..6933\n",
      "                     /gene=\"SsGP\"\n",
      "                     /locus_tag=\"DF49_53416gpSsGP\"\n",
      "     CDS             6039..6933\n",
      "                     /gene=\"SsGP\"\n",
      "                     /locus_tag=\"DF49_53416gpSsGP\"\n",
      "                     /exception=\"RNA editing\"\n",
      "                     /note=\"two non-templated A's inserted during transcription\n",
      "                     due to RNA editing\"\n",
      "                     /codon_start=1\n",
      "                     /product=\"super small secreted glycoprotein SsGP\"\n",
      "                     /protein_id=\"ALT19752.1\"\n",
      "                     /translation=\"MGVTGILQLPRDRFKRTSFFLWVIILFQRTFSIPLGVIHNSTLQ\n",
      "                     VSDVDKLVCRDKLSSTNQLRSVGLNLEGNGVATDVPSATKRWGFRSGVPPKVVNYEAG\n",
      "                     EWAENCYNLEIKKPDGSECLPAAPDGIRGFPRCRYVHKVSGTGPCAGDFAFHKEGAFF\n",
      "                     LYDRLASTVIYRGTTFAEGVVAFLILPQAKKDFFSSHPLREPVNATEDPSSGYYSTTI\n",
      "                     RYQATGFGTNETEYLFEVDNLTYVQLESRFTPQFLLQLNETIYTSGKRSNTTGKLIWK\n",
      "                     VNPEIDTTIGEWAFWETKKKPH\"\n",
      "     misc_feature    6917..6924\n",
      "                     /gene=\"GP\"\n",
      "                     /locus_tag=\"DF49_53416gpGP\"\n",
      "                     /note=\"location of RNA editing (TAAAAAAA => TAAAAAAAA) in\n",
      "                     envelope glycoprotein\"\n",
      "     misc_feature    6917..6924\n",
      "                     /gene=\"SsGP\"\n",
      "                     /locus_tag=\"DF49_53416gpSsGP\"\n",
      "                     /note=\"location of RNA editing (TAAAAAAA => TAAAAAAAAA) in\n",
      "                     super small secreted glycoprotein SsGP\"\n",
      "     gene            8509..9375\n",
      "                     /gene=\"VP30\"\n",
      "                     /locus_tag=\"DF49_53416gpVP30\"\n",
      "     CDS             8509..9375\n",
      "                     /gene=\"VP30\"\n",
      "                     /locus_tag=\"DF49_53416gpVP30\"\n",
      "                     /note=\"synonym: transcription activator VP30\"\n",
      "                     /codon_start=1\n",
      "                     /product=\"minor nucleoprotein VP30\"\n",
      "                     /protein_id=\"ALT19755.1\"\n",
      "                     /translation=\"MEASYERGRPRAARQHSRDGHDHHVRARSSSRENYRGEYRQSRS\n",
      "                     ASQVRVPTVFHKKRVEPLTVPPAPKDICPTLKKGFLCDSSFCKKDHQLESLTDRELLL\n",
      "                     LIARKTCGSVEQQLNITAPKDSRLANPTADDFQQEEGPKITLLTLIKTAEHWARQDIR\n",
      "                     TIEDSKLRALLTLCAVMTRKFSKSQLSLLCETHLRREGLGQDQAEPVLEVYQRLHSDK\n",
      "                     GGSFEAALWQQWDRQSLIMFITAFLNIALQLPCESSAVVVSGLRTLVPQSDNEEASTN\n",
      "                     PGTCSWSDEGTP\"\n",
      "     gene            10345..11100\n",
      "                     /gene=\"VP24\"\n",
      "                     /locus_tag=\"DF49_53416gpVP24\"\n",
      "     CDS             10345..11100\n",
      "                     /gene=\"VP24\"\n",
      "                     /locus_tag=\"DF49_53416gpVP24\"\n",
      "                     /note=\"secondary matrix protein\"\n",
      "                     /codon_start=1\n",
      "                     /product=\"membrane-associated protein VP24\"\n",
      "                     /protein_id=\"ALT19756.1\"\n",
      "                     /translation=\"MAKATGRYNLISPKKDLEKGVVLSDLCNFLVSQTIQGWKVYWAG\n",
      "                     IEFDVTHKGMALLQRLKTNDFAPAWSMTRNLFPHLFQNPNSTIESPLWALRVILAAGI\n",
      "                     QDQLIDQSLIEPLAGALGLISDWLLTTNTNHFNMRTQRVKEQLSLKMLSLIRSNILKF\n",
      "                     INKLDALHVVNYNGLLSSIEIGTQNHTIIITRTNMGFLVELQEPDKSAMNRKKPGPAK\n",
      "                     FSLLHESTLKAFTQGSSTRMQSLILEFNSSLAI\"\n",
      "     gene            11581..18219\n",
      "                     /gene=\"L\"\n",
      "                     /locus_tag=\"DF49_53416gpL\"\n",
      "     CDS             11581..18219\n",
      "                     /gene=\"L\"\n",
      "                     /locus_tag=\"DF49_53416gpL\"\n",
      "                     /codon_start=1\n",
      "                     /product=\"RNA-directed RNA polymerase L\"\n",
      "                     /protein_id=\"ALT19757.1\"\n",
      "                     /translation=\"MATQHTQYPDARLSSPIVLDQCDLVTRACGLYSSYSLNPQLRNC\n",
      "                     KLPKHIYRLKYDVTVTKFLSDVPVATLPIDFIVPILLKALSGNGFCPVEPRCQQFLDE\n",
      "                     IIKYTMQDALFLKYYLKNVGAQEDCVDDHFQEKILSSIQGNEFLHQMFFWYDLAILTR\n",
      "                     RGRLNRGNSRSTWFVHDDLIDILGYGDYVFWKIPISMLPLNTQGIPHAAMDWYQASVF\n",
      "                     KEAVQGHTHIVSVSTADVLIMCKDLITCRFNTTLISKIAEIEDPVCSDYPNFKIVSML\n",
      "                     YQSGDYLLSILGSDGYKIIKFLEPLCLAKIQLCSKYTERKGRFLTQMHLAVNHTLEEI\n",
      "                     TEMRALKPSQAQKIREFHRTLIRLEMTPQQLCELFSIQKHWGHPVLHSETAIQKVKKH\n",
      "                     ATVLKALRPIVIFETYCVFKYSIAKHYFDSQGSWYSVTSDRNLTPGLNSYIKRNQFPP\n",
      "                     LPMIKELLWEFYHLDHPPLFSTKIISDLSIFIKDRATAVERTCWDAVFEPNVLGYNPP\n",
      "                     HKFSTKRVPEQFLEQENFSIENVLSYAQKLEYLLPQYRNFSFSLKEKELNVGRTFGKL\n",
      "                     PYPTRNVQTLCEALLADGLAKAFPSNMMVVTEREQKESLLHQASWHHTSDDFGEHATV\n",
      "                     RGSSFVTDLEKYNLAFRYEFTAPFIEYCNRCYGVKNVFNWMHYTIPQCYMHVSDYYNP\n",
      "                     PHNLTLENRDNPPEGPSSYRGHMGGIEGLQQKLWTSISCAQISLVEIKTGFKLRSAVM\n",
      "                     GDNQCITVLSVFPLETDADEQEQSAEDNAARVAASLAKVTSACGIFLKPDETFVHSGF\n",
      "                     IYFGKKQYLNGVQLPQSLKTATRMAPLSDAIFDDLQGTLASIGTAFERSISETRHIFP\n",
      "                     CRITAAFHTFFSVRILQYHHLGFNKGFDLGQLTLGKPLDFGTISLALAVPQVLGGLSF\n",
      "                     LNPEKCFYRNLGDPVTSGLFQLKTYLRMIEMDDLFLPLIAKNPGNCTAIDFVLNPSGL\n",
      "                     NVPGSQDLTSFLRQIVRRTITLSAKNKLINTLFHASADFEDEMVCKWLLSSTPVMSRF\n",
      "                     AADIFSRTPSGKRLQILGYLEGTRTLLASRIINNNTETPVLDRLRKITLQRWSLWFSY\n",
      "                     LDHCDNILAEALTQITCTVDLAQILREYSWAHILEGRPLIGATLPCMIEQFKVFWLKP\n",
      "                     YEQCPQCSNAKQPGGKPFVSVAVKKHIVSAWPNASRLSWTIGDGIPYIGSRTEDKIGQ\n",
      "                     PAIKPKCPSAALREAIELASRLTWVTQGSSNSDLLIKPFLEARVNLSVQEILQMTPSH\n",
      "                     YSGNIVHRYNDQYSPHSFMANRMSNSATRLIVSTNTLGEFSGGGQSARDSNIIFQNVI\n",
      "                     NYAVALFDIKFRNTEATDIQYNRAHLHLTKCCTREVPAQYLTYTSTLDLDLTRYRENE\n",
      "                     LIYDNNPLKGGLNCNISFDNPFFQGKQLNIIEDDLIRLPHLSGWELAKTIMQSIISDS\n",
      "                     NNSSTDPISSGETRSFTTHFLTYPKIGLLYSFGAFVSYYLGNTILRTKKLTLDNFLYY\n",
      "                     LTTQIHNLPHRSLRILKPTFKHASVMSRLMSIDPHFSIYIGGAAGDRGLSDAARLFLR\n",
      "                     TSISSFLTFVKEWIINRGTIVPLWIVYPLEGQNPTPVNNFLYQIVELLVHDSSRQQAL\n",
      "                     KTTISDHVHPHDNLVYTCKSTASNFFHASLAYWRSRHRNSNRKYLARDSSTRSSTNNS\n",
      "                     DGHIERSQEQTTRDPHDGTERNLVLQMSHEIKRTTIPQENTHQGPSFQSFLSYSACGT\n",
      "                     ANPKLNFDRSRHNVKSQDHNSASKREGHQIISHRLVLPFFTLSQGTRQLTSSNESQTQ\n",
      "                     DEISKYLRQLRSVIDTTVYCRFTGIVSSMHYKLDEVLWEIESFKSAVTLAEGEGAGAL\n",
      "                     LLIQKYQVKTLFFNTLATESSIESEIVSGMTTPRMLLPVMSKFHNDQIEIILNNSASQ\n",
      "                     ITDITNPTWFKDQRARLPKQVEVITMDAETTENINRSKLYEAVYKLILHHIDPSVLKA\n",
      "                     VVLKVFLSDTEGMLWLNDNLAPFFATGYLIKPITSSARSSEWYLCLTNFLSTTRKMPH\n",
      "                     QNHLSCKQVILTALQLQIQRSPYWLSHLTQYADCDLHLSYIRLGFPSLEKVLYHRYNL\n",
      "                     VDSKRGPLVSITQHLAHLRAEIRELTNDYNQQRQSRTQTYHFIRTAKGRITKLVNDYL\n",
      "                     KFFLIVQALKHNGTWQAEFKKLPELISVCNRFYHIRDCNCEERFLVQTLYLHRMQDSE\n",
      "                     VKLIERLTGLLSLFPDGLYRFD\"\n",
      "ORIGIN      \n",
      "        1 cggacacaca aaaagaaaga agaattttta ggatcttttg tgtgcgaata actatgagga\n",
      "       61 agattaataa ttttcctctc attgaaattt atatcggaat ttaaattgaa attgttactg\n",
      "      121 taatcacacc tggtttgttt cagagccaca tcacaaagat agagaacaac ctaggtctct\n",
      "      181 gaagggagca agggcatcag tgtgctcagt tgaaaatccc ttgtcaacat ctaggtctta\n",
      "      241 tcacatcaca agtcccacct cagactctgc agggtgatcc aacaacctta atagaaacat\n",
      "      301 tattgttaaa ggacagcatt agttcacagt caaacaagca agattgagaa ttaaccttgg\n",
      "      361 ttttgaactt gaatacttag aggattggag attcaacaac cctaaagctt ggggtaaaac\n",
      "      421 attggaaata gttaaaagac aaattgctcg gaatcacaac attccgagta tggattctcg\n",
      "      481 tcctcagaaa gtctggatga cgccgagtct cactgaatct gacatggatt accacaagat\n",
      "      541 cttgacagca ggtctgtccg ttcaacaggg gattgttcgg caaagagtca tcccagtgta\n",
      "      601 tcaagtaaac aatcttgagg agatttgcca acttatcata caggcctttg aagcaggtgt\n",
      "      661 tgattttcaa gagagtgcgg acagtttcct tctcatgctt tgtcttcatc atgcgtacca\n",
      "      721 gggagattac aaacttttct tggaaagtgg cgcagtcaag tatttggaag ggcacgggtt\n",
      "      781 ccgttttgaa gtcaagaagc gtgatggagt gaagcgcctt gaggaattgc tgccagcagt\n",
      "      841 atctagtgga aaaaacatta agagaacact tgctgccatg ccggaagagg agacaactga\n",
      "      901 agctaatgcc ggtcagtttc tctcctttgc aagtctattc cttccgaaat tggtagtagg\n",
      "      961 agaaaaggct tgtcttgaga aggttcaaag gcaaattcaa gtacatgcag agcaaggact\n",
      "     1021 gatacaatat ccaacagctt ggcaatcagt aggacacatg atggtgattt tccgtttgat\n",
      "     1081 gcgaacaaat tttttgatca aatttctcct aatacaccaa gggatgcaca tggttgccgg\n",
      "     1141 gcatgatgcc aacgacgctg tgatttcaaa ttcagtggct caagctcgtt tttcaggttt\n",
      "     1201 attgattgtc aaaacagtac ttgatcatat cctacaaaag acagaacgcg gagttcgtct\n",
      "     1261 ccatcctctt gcaaggaccg ccaaggtaaa aaatgaggtg aactccttta aggctgcact\n",
      "     1321 cagctccctg gccaagcatg gagagtatgc tcctttcgcc cgacttttga acctttctgg\n",
      "     1381 agtaaataat cttgagcatg gtcttttccc tcaactatca gcaattgcac tcggagtcgc\n",
      "     1441 cacagcacac gggagtaccc tcgcaggagt aaatgttgga gaacagtatc aacaactcag\n",
      "     1501 agaggctgcc actgaggctg agaagcaact ccaacaatac gcagagtctc gcgaacttga\n",
      "     1561 ccatcttgga cttgatgatc aggaaaagaa aattcttatg aacttccatc agaaaaagaa\n",
      "     1621 cgaaatcagc ttccagcaaa caaacgctat ggtaactcta agaaaagagc gcctggccaa\n",
      "     1681 gctgacggaa gctatcactg ctgcgtcact gcccaaaaca agtggacatt acgatgatga\n",
      "     1741 tgacgacatt ccctttccag gacccatcaa tgatgacgac aatcctggcc atcaagatga\n",
      "     1801 tgatccaact gactcacaag atacgaccat tcctgatgtg gtggttgatc ccgatgatgg\n",
      "     1861 aagctacggc gaataccaga gttactcgga aaacggcatg aatgcaccag atgacttggt\n",
      "     1921 cctattcgat ctagacgagg acgatgagga cactaagcca gtgcctaata gatcaaccaa\n",
      "     1981 gggtggacaa cagaaaaaca gtcaaaaggg ccagcataca gagggcagac agacacaatc\n",
      "     2041 caggccaact caaaatgtcc caggccctca cagaacaatc caccacgcca gtgctccact\n",
      "     2101 cacggataat gacagaagaa atgaaccctc cggctcaacc agccctcgca tgctgacacc\n",
      "     2161 aatcaacgaa gaggcagacc cactggacga tgccgacgac gagacgtcta gccttccgcc\n",
      "     2221 cctggagtca gacgatgaag aacaggacag ggacggaact tccaaccgca cacccactgt\n",
      "     2281 cgccccaccg gctcccgtat acagagatca ctctgaaaag agagaactcc cgcaagatga\n",
      "     2341 gcaacaagat caggaccaca ctcaagaggc caggaaccag gacagtgaca acacccagcc\n",
      "     2401 agaacactct tttgaggaga tgtatcgcca cattctaaga tcacaggggc catttgatgc\n",
      "     2461 tgttttgtat tatcatatga tgaaggatga gcctgtagtt ttcagtacca gtgatggcaa\n",
      "     2521 agagtacacg tatccagact cccttgaaga ggaatatcca ccatggctca ctgaaaaaga\n",
      "     2581 ggctatgaat gaagagaata gatttgttac attggatggt caacaatttt attggccggt\n",
      "     2641 aatgaatcac aagaataaat tcatggcaat cctgcaacat catcagtgaa tgagcatgga\n",
      "     2701 acaatgggat gattcaaccg acaaatagct aacattaggt agtcaaggaa cgaaaacagg\n",
      "     2761 aagaattttt gatgtctaag gtgtgaatta ttatcacaat aaaagtgatt cttatttttg\n",
      "     2821 aatttaaagc tagcttatta ttactagccg tttttcaaag tccaatttga gtcttaatgc\n",
      "     2881 aaataggcgt taagccacag ttatagccat aattgtaact caatatccta actagcgatt\n",
      "     2941 tatctaaatt aaattacatt atgcttttat aacttaccta ctagcctacc caacatttac\n",
      "     3001 acgatcattt tataattaag aaaaaactaa tgatgaagat taaaaccttc atcatcctta\n",
      "     3061 cgtcaattga attctctagc actcgaagct tattgtcctc aatgtaaaag aaaagctggt\n",
      "     3121 ctaacaagat gacaaccaga acaaagggca ggggccacac tgcggccacg actcaaaacg\n",
      "     3181 acagaatgcc aggccctgag ctttcgggct ggatctctga gcagctaatg accggaagaa\n",
      "     3241 ttcctgtaag cgacatcttc tgtgatattg agaacaatcc aggattatgc tacgcatccc\n",
      "     3301 aaatgcaaca aaccaagcca aacccgaaga cgcgcaacag tcaaacccaa acggacccaa\n",
      "     3361 tttgcaatca tagttttgag gaggtagtac aaacattagc ttcattggct actgttgtgc\n",
      "     3421 aacaacaaac cattgcatca gaatcattag aacaacgcat tacgagtctt gagaatggtc\n",
      "     3481 taaagccagt ttatgatatg gctaaaacaa tctcctcatt gaacagggtt tgtgctgaga\n",
      "     3541 tggttgcaaa atatgatctt ctggtgatga caaccggtcg ggcaacagca accgctgcgg\n",
      "     3601 caactgaggc ttattgggcc gaacatggtc aaccaccacc tggaccatca ctttatgaag\n",
      "     3661 aaagtgcaat tcggggtaag attgaatcta gagatgagac cgtccctcaa agtgttaggg\n",
      "     3721 aggcattcaa caatctagac agtaccactt cactaactga ggaaaatttt gggaaacctg\n",
      "     3781 acatttcggc aaaggatttg agaaacatta tgtatgatca cttgcctggt tttggaactg\n",
      "     3841 ctttccacca attagtacaa gtgatttgta aattgggaaa agatagcaac tcattggaca\n",
      "     3901 tcattcatgc tgagttccag gccagcctgg ctgaaggaga ctctcctcaa tgtgccctaa\n",
      "     3961 ttcaaattac aaaaagagtt ccaatcttcc aagatgctgc tccacctgtc atccacatcc\n",
      "     4021 gctctcgagg tgacattccc cgagcttgcc agaaaagctt gcgtccagtc ccaccatcgc\n",
      "     4081 ccaagattga tcgaggttgg gtatgtgttt ttcagcttca agatggtaaa acacttggac\n",
      "     4141 tcaaaatttg agcccatctt ccttccctcc gaaagaggcg actaatagca gaggcttcaa\n",
      "     4201 ctgctgaact atagggtacg ttacattaat gatacacttg tgagtatcag ccctggataa\n",
      "     4261 tataagtcaa ttaaacgacc aagataaaat tgttcttatc tcgctagcag cttaaaatat\n",
      "     4321 gaatgtaata ggagctatat ctctgacagt attataatca atcgttatta agtaacccaa\n",
      "     4381 accaaaagtg atgaagatta agaaaaacct acctcgactg agagagtgtt ttttcattaa\n",
      "     4441 ccttcatctt gtaaacgttg agcaaaattg ttaaaaatat gaggcgggtt atattaccta\n",
      "     4501 ctgctcctcc tgaatatatg gaggccatat accctgtcag gtcaaattca acaattgcta\n",
      "     4561 gaggtggcaa cagcaataca ggcttcctga caccggagtc agtcaatggg gacactccat\n",
      "     4621 cgaatccact caggccaatt gccgatgaca ccatcgacca tgccagccac acaccaggca\n",
      "     4681 gtgtgtcatc agcattcatc cttgaagcta tggtgaatgt catatcgggc cccaaagtgc\n",
      "     4741 taatgaagca aattccaatt tggcttcctc taggtgtcgc tgatcaaaag acctacagct\n",
      "     4801 ttgactcaac aacggccgcc atcatgcttg cttcatatac tatcacccat ttcggcaagg\n",
      "     4861 caaccaatcc acttgtcaga gtcaatcggc tgggtcctgg aatcccggat caccccctca\n",
      "     4921 ggctcctgcg aattggaaac caggccttcc tccaggagtt cgttcttccg ccagtccaac\n",
      "     4981 taccccagta tttcaccttt gatttgacag cactcaaact gatcacccaa ccactgcctg\n",
      "     5041 ctgcaacatg gaccgatgac actccaacag gatcaaatgg agcgttgcgc ccagggattt\n",
      "     5101 catttcatcc aaaacttcgc cccattcttt tacccaacaa gagtgggaag aaggggaata\n",
      "     5161 gtgccgatct aacatctccg gagaaaatcc aagcaataat gacttcactc caggacttta\n",
      "     5221 agatcgttcc aattgatcca accaaaaata tcatgggaat cgaagtgcca gaaactctgg\n",
      "     5281 tccacaagct gaccggtaag aaggtgactt ctaaaaatgg acaaccaatc atccctgttc\n",
      "     5341 ttttgccaaa gtacattggt ttggacccgg tggctccagg agacctcacc atggtaatca\n",
      "     5401 cacaggattg tgacacgtgt cattctcctg cgagtcttcc agctgtgatt gagaagtaat\n",
      "     5461 tgcaataatt gactcagatc cagttttaca gaatcttctc agggatagtg ataacatcta\n",
      "     5521 tttagtaatc cgtccattag aggagatact tttaattgat caatatacta aaggtgcttt\n",
      "     5581 acaccattgt ctttttcctc tcctaaatgt agaacttaac aaaagactca taatatactt\n",
      "     5641 gtttttaaaa gattgattga tgaaagatca taactaataa cattacaaat aatcctacta\n",
      "     5701 taatcaatac ggtgatccaa atgttaatct ttctcattgc aaatactctt tgcccttatc\n",
      "     5761 ctcaaattgc ctacatgctt acatctgagg atagccagtg tgacttggat tggagatgtg\n",
      "     5821 gagaaaaaat cgggacccat ttctaggttg ttcaccatcc aagtacagac attgcccttc\n",
      "     5881 taattaagaa aaaatcggcg atgaagatta agccgacagt gagcgtaatc ttcatctctc\n",
      "     5941 ttagattatt tgtcctccag agtagggatc gtcaggtcct tttcaatcgt ataaccaaaa\n",
      "     6001 taaacttcac tagaaggata ttgtggggca acaacacaat gggtgttaca ggaatattgc\n",
      "     6061 agttacctcg tgatcgattc aagaggacat cattctttct ttgggtaatt atccttttcc\n",
      "     6121 aaagaacatt ttccatccca cttggagtca tccacaatag cacattacag gttagtgatg\n",
      "     6181 tcgacaaact ggtttgccgt gacaaactgt catccacaaa tcaattgaga tcagttggac\n",
      "     6241 tgaatctcga agggaatgga gtggcaactg acgtgccatc tgcaactaaa agatggggct\n",
      "     6301 tcaggtccgg tgtcccacca aaggtggtca attatgaagc tggtgaatgg gctgaaaact\n",
      "     6361 gctacaatct tgaaatcaaa aaacctgacg ggagtgagtg tctaccagca gcgccagacg\n",
      "     6421 ggattcgggg cttcccccgg tgccggtatg tgcacaaagt atcaggaacg ggaccgtgtg\n",
      "     6481 ccggagactt tgccttccac aaagagggtg ctttcttcct gtatgaccga cttgcttcca\n",
      "     6541 cagttatcta ccgaggaacg actttcgctg aaggtgtcgt tgcatttctg atactgcccc\n",
      "     6601 aagctaagaa ggacttcttc agctcacacc ccttgagaga gccggtcaat gcaacggagg\n",
      "     6661 acccgtctag tggctactat tctaccacaa ttagatatca agctaccggt tttggaacca\n",
      "     6721 atgagacaga gtatttgttc gaggttgaca atttgaccta cgtccaactt gaatcaagat\n",
      "     6781 tcacaccaca gtttctgctc cagctgaatg agacaatata tacaagtggg aaaaggagca\n",
      "     6841 ataccacggg aaaactaatt tggaaggtca accccgaaat tgatacaaca atcggggagt\n",
      "     6901 gggccttctg ggaaactaaa aaaacctcac tagaaaaatt cgcagtgaag agttgtcttt\n",
      "     6961 cacagctgta tcaaacagag ccaaaaacat cagtggtcag agtccggcgc gaacttcttc\n",
      "     7021 cgacccaggg accaacacaa caactgaaga ccacaaaatc atggcttcag aaaattcctc\n",
      "     7081 tgcaatggtt caagtgcaca gtcaaggaag ggaagctgca gtgtcgcatc tgacaaccct\n",
      "     7141 tgccacaatc tccacgagtc ctcaaccccc cacaaccaaa ccaggtccgg acaacagcac\n",
      "     7201 ccacaataca cccgtgtata aacttgacat ctctgaggca actcaagttg aacaacatca\n",
      "     7261 ccgcagaaca gacaacgaca gcacagcctc cgacactccc cccgccacga ccgcagccgg\n",
      "     7321 acccctaaaa gcagagaaca ccaacacgag caagggtacc gacctcctgg accccgccac\n",
      "     7381 cacaacaagt ccccaaaacc acagcgagac cgctggcaac aacaacactc atcaccaaga\n",
      "     7441 taccggagaa gagagtgcca gcagcgggaa gctaggctta attaccaata ctattgctgg\n",
      "     7501 agtcgcagga ctgatcacag gcgggaggag agctcgaaga gaagcaattg tcaatgctca\n",
      "     7561 acccaaatgc aaccctaatt tacattactg gactactcag gatgaaggtg ctgcaatcgg\n",
      "     7621 actggcctgg ataccatatt tcgggccagc agccgaggga atttacatag aggggctgat\n",
      "     7681 gcacaatcaa gatggtttaa tctgtgggtt gagacagctg gccaacgaga cgactcaagc\n",
      "     7741 tcttcaactg ttcctgagag ccacaaccga gctacgcacc ttttcaatcc tcaaccgtaa\n",
      "     7801 ggcaattgat ttcttgctgc agcgatgggg cggcacatgc cacattttgg gaccggactg\n",
      "     7861 ctgtatcgaa ccacatgatt ggaccaagaa cataacagac aaaattgatc agattattca\n",
      "     7921 tgattttgtt gataaaaccc ttccggacca gggggacaat gacaattggt ggacaggatg\n",
      "     7981 gagacaatgg ataccggcag gtattggagt tacaggcgtt ataattgcag ttatcgcttt\n",
      "     8041 attctgtata tgcaaatttg tcttttagtt tttcttcaga ttgcttcatg gcaaagctca\n",
      "     8101 gcctcaaatc aatgaaacca ggatttaatt atatggatta cttgaatcta agattacttg\n",
      "     8161 acaaatgata atataataca ctggagcttt aaacatagcc aatgtgattc taactctttt\n",
      "     8221 aaactcacag ttaatcataa ataaggtttg acatcaatct agttatctct ttgagaatga\n",
      "     8281 taaacttgat gaagattaag aaaaaggtaa tctttcgatt atctttaatc ttcatccttg\n",
      "     8341 attctacaat catgacagtt gtctttagtg acaagggaaa gaagcctttt tattaagttg\n",
      "     8401 taataatcag atctgcgaac cggtagagtt tagttgcaac ctaacacaca taaagcattg\n",
      "     8461 atcaaaaagt caatagaaat ttaaacagtg agtggagaca acttttaaat ggaagcttca\n",
      "     8521 tatgagagag gacgcccacg agctgccaga cagcattcaa gggatggaca cgaccaccat\n",
      "     8581 gttcgagcac gatcatcatc cagagagaat tatcgaggtg agtaccgtca atcaaggagc\n",
      "     8641 gcctcacaag tgcgcgttcc tactgtattt cataagaaga gagttgaacc attaacagtt\n",
      "     8701 cctccagcac ctaaagacat atgtccgacc ttgaaaaaag gatttttgtg tgacagtagt\n",
      "     8761 ttttgcaaaa aagatcacca gttggagagt ttaactgata gggaattact cctactaatc\n",
      "     8821 gcccgtaaga cttgtggatc agtagaacaa caattaaata taactgcacc caaggactcg\n",
      "     8881 cgcttagcaa atccaacggc tgatgatttc cagcaagagg aaggtccaaa aattaccttg\n",
      "     8941 ttgactctga tcaagacggc agaacactgg gcgagacaag acatcagaac catagaggat\n",
      "     9001 tcaaaattaa gagcattgtt gactctatgt gctgtgatga cgaggaaatt ctcaaaatcc\n",
      "     9061 cagctgagtc ttttatgtga gacacaccta aggcgtgagg ggcttgggca agatcaggca\n",
      "     9121 gaacctgttc tcgaagtata tcaacgatta cacagtgata aaggaggcag ttttgaagct\n",
      "     9181 gcactatggc aacaatggga ccgacaatcc ctaattatgt ttatcactgc attcttgaat\n",
      "     9241 atcgctctcc agttaccgtg tgaaagttct gctgtcgttg tttcagggtt aagaacattg\n",
      "     9301 gttcctcaat cagataatga ggaagcttca accaacccgg ggacatgctc atggtctgat\n",
      "     9361 gagggtaccc cttaataagg ctgactaaaa cactatataa ccttctactt gatcacaata\n",
      "     9421 ctccgtatac ctattatcat atatttaatc aagacgatat cctttaagac ttattcagta\n",
      "     9481 ctataatcac tctcgtttca aattaataag atgtgcatga ttgccctaat atatgaagag\n",
      "     9541 gtatgataca accctaacag tgaccaaaga aaatcataat ctcgtatcgc tcgtaatata\n",
      "     9601 acctgccaag catacctctt gcacaaagtg attcttgtac acaaataatg ttttacccta\n",
      "     9661 caggaggtag caacgatcca tcccatcaaa aaataagtat tttatgactt actaatgacc\n",
      "     9721 tcttaaaata ttaagaaaaa ctgacggaac ataaattctt tctgcttcaa gttgtggagg\n",
      "     9781 aggtgtttgg tattggctat tgttatatta caatcaataa caagcttgta aaaatattgt\n",
      "     9841 tcttgtttca agaggtagat tgtggctgga aatgctaaac taatgatgaa gattaatgcg\n",
      "     9901 gaggtctgat aagaataaac cttattattc agattaggcc ccaagaggca ttcttcatct\n",
      "     9961 ccttttagca aagtactatt tcagggtagt ccaattagtg acacgtctct tagctgtata\n",
      "    10021 tcagtcgccc ctgagatacg ccacaaaagt gtctctaagc taaattggtc tgtacacatc\n",
      "    10081 tcatacattg tattagggac aataatatct aattgaactt agccgtttaa aatttagtgc\n",
      "    10141 ataaatctgg gctaactcca ccaggtcaac tccattggct gaaaagaagc ctacctacaa\n",
      "    10201 cgaacatcac tttgagcgcc ctcacattta aaaaatagga acgtcgttcc aacaatcgag\n",
      "    10261 cgcaaggttt caaggttgaa ctgagagtgt ctagacaaca aagtatcgat cctccagaca\n",
      "    10321 ccaagcaaga cctgagaaaa aaccatggct aaagctacgg gacgatacaa tctaatatcg\n",
      "    10381 cccaaaaagg acctggagaa aggggttgtc ttaagcgacc tctgtaactt cttagttagc\n",
      "    10441 caaactattc aggggtggaa ggtttattgg gctggtattg agtttgatgt gactcacaaa\n",
      "    10501 ggaatggccc tattgcaaag actgaaaact aatgactttg cccctgcatg gtcaatgaca\n",
      "    10561 aggaatctct ttcctcattt atttcaaaat ccgaattcca caattgaatc accgctgtgg\n",
      "    10621 gcattgagag tcatccttgc agcagggata caagaccagc tgattgacca gtctttgatt\n",
      "    10681 gaacccttag caggagccct tggtctgatc tctgattggc tgctaacaac caacactaac\n",
      "    10741 catttcaaca tgcgaacaca acgtgtcaag gaacaattga gcctaaaaat gctgtcgttg\n",
      "    10801 attcgatcca atattctcaa gtttattaac aaattggatg ctctacatgt cgtgaactac\n",
      "    10861 aacggattgt tgagcagtat tgaaattgga actcaaaatc atacaatcat cataactcga\n",
      "    10921 actaacatgg gttttctggt ggagctccaa gaacccgaca aatcggcaat gaaccgcaag\n",
      "    10981 aagcctgggc cggcgaaatt ttccctcctt catgagtcca cactgaaagc atttacacaa\n",
      "    11041 ggatcctcaa cacgaatgca aagtttgatt cttgaattta atagctctct tgctatctaa\n",
      "    11101 ctaagatgga atacttcata ttgagctaac tcatatatgc tgactcaata gttatcttga\n",
      "    11161 catctctgct ttcataatca gatatataag cataataaat aaatacgcat atttcttgat\n",
      "    11221 aatttgttta accacagata aatcctcact gtaagccagc ttccaagttg acacccttac\n",
      "    11281 aaaaaccagg actcagaatc cctcaaataa gagattccaa gacaacatca tagaattgct\n",
      "    11341 ttatcatatg aataagcatt ttatcaccag aaatcttata tactaaatgg ttaattgtaa\n",
      "    11401 ctgaacccgc aggtcacatg tgttaggttt cacagatttt atatattact aactctatac\n",
      "    11461 tcgtaattaa cattagataa gtagattaag aaaaaagcct gaggaagatt aagaaaaact\n",
      "    11521 gcttattggg tctttccgtg ttttagatga agcagttgaa attcttcctc ttgatattaa\n",
      "    11581 atggctacac aacataccca atacccggac gctaggttat catcaccaat tgtattggac\n",
      "    11641 caatgtgacc tagtcactag agcttgcggg ttatattcat catactccct taatccgcaa\n",
      "    11701 ctacgcaact gtaaactccc gaaacatatc taccgtttga aatacgatgt aactgttacc\n",
      "    11761 aagttcttga gtgatgtacc agtggcgaca ttgcccatag atttcatagt cccaattctt\n",
      "    11821 ctcaaggcac tgtcaggcaa tgggttctgt cctgttgagc cgcggtgcca acagttctta\n",
      "    11881 gatgaaatca ttaagtacac aatgcaagat gctctcttct tgaaatatta tctcaaaaat\n",
      "    11941 gtgggtgctc aagaagactg tgttgatgac cactttcaag agaaaatctt atcttcaatt\n",
      "    12001 cagggcaatg aatttttaca tcaaatgttt ttctggtatg atctggctat tttaactcga\n",
      "    12061 aggggtagat taaatcgagg aaactctaga tcaacatggt ttgttcatga tgatttaata\n",
      "    12121 gacatcttag gttatgggga ctatgttttt tggaagatcc caatttcaat gttaccactg\n",
      "    12181 aacacacaag gaatccccca tgctgctatg gactggtatc aggcatcagt attcaaagaa\n",
      "    12241 gcggttcaag ggcatacaca cattgtttct gtttctactg ccgacgtctt gataatgtgc\n",
      "    12301 aaagatttaa ttacatgtcg attcaacaca actctaatct caaaaatagc agagattgag\n",
      "    12361 gatccagttt gttctgatta tcccaatttt aagattgtgt ctatgcttta ccagagcgga\n",
      "    12421 gattacttac tctccatatt agggtctgat gggtataaaa ttattaagtt cctcgaacca\n",
      "    12481 ttgtgcttgg ccaaaattca attatgctca aagtacaccg agaggaaggg ccgattctta\n",
      "    12541 acacaaatgc atttagctgt aaatcacacc ctagaagaaa ttacagaaat gcgtgcacta\n",
      "    12601 aagccttcac aggctcaaaa gatccgtgaa ttccatagaa cattgataag gctggagatg\n",
      "    12661 acgccacaac agctttgtga gctattttcc attcaaaaac actgggggca tcctgtgcta\n",
      "    12721 catagtgaaa cagcaatcca aaaagttaaa aaacatgcta cggtgctaaa agcattacgc\n",
      "    12781 cctatagtga ttttcgagac atactgtgtt tttaaatata gtattgccaa acattatttt\n",
      "    12841 gatagtcaag ggtcttggta cagtgttact tcagatagga atctaacacc aggtcttaat\n",
      "    12901 tcttatatca aaagaaatca attccctccg ttgccaatga ttaaagaact actatgggaa\n",
      "    12961 ttttaccacc ttgaccatcc tccacttttc tcaaccaaaa ttattagtga cttaagtatt\n",
      "    13021 tttataaaag acagagctac tgcagtagaa aggacatgct gggatgcagt attcgagcct\n",
      "    13081 aatgttctag gatataatcc acctcacaaa tttagtacta aacgtgtacc ggaacaattt\n",
      "    13141 ttagagcaag aaaacttttc tattgagaat gttctttcct acgcgcaaaa actcgagtat\n",
      "    13201 ctactaccac aatatcggaa cttttctttc tcattgaaag agaaagagtt gaatgtaggt\n",
      "    13261 agaaccttcg gaaaattgcc ttatccgact cgcaatgttc aaacactttg tgaagctctg\n",
      "    13321 ttagctgatg gtcttgctaa agcatttcct agcaatatga tggtagttac ggaacgtgag\n",
      "    13381 caaaaagaaa gcttattgca tcaagcatca tggcaccaca caagtgatga ttttggtgaa\n",
      "    13441 catgccacag ttagagggag tagctttgta actgatttag agaaatacaa tcttgcattt\n",
      "    13501 agatatgagt ttacagcacc ttttatagaa tattgcaacc gttgctatgg tgttaagaat\n",
      "    13561 gtttttaatt ggatgcatta tacaatccca cagtgttata tgcatgtcag tgattattat\n",
      "    13621 aatccaccac ataacctcac actggaaaat cgagacaacc cccccgaagg gcctagttca\n",
      "    13681 tacaggggtc atatgggagg gattgaagga ctgcaacaaa aactctggac aagtatttca\n",
      "    13741 tgtgctcaaa tttctttagt tgaaattaag actggtttta agttacgctc agctgtgatg\n",
      "    13801 ggtgacaatc agtgcattac cgttttatca gtcttcccct tagagactga cgcagacgag\n",
      "    13861 caggaacaga gcgccgaaga caatgcagcg agggtggccg ctagcctagc aaaagttaca\n",
      "    13921 agtgcctgtg gaatcttttt aaaacctgat gaaacatttg tacattcagg ttttatctat\n",
      "    13981 tttggaaaaa aacaatattt gaatggggtc caattgcctc agtcccttaa aacggctaca\n",
      "    14041 agaatggcac cattgtctga tgcaattttt gatgatcttc aagggaccct agctagtata\n",
      "    14101 ggcactgctt ttgagcgatc catctctgag acacgacata tctttccttg caggataacc\n",
      "    14161 gcagctttcc atacgttttt ttcggtgaga atcttgcaat atcatcatct cgggttcaat\n",
      "    14221 aaaggttttg accttggaca gttaacactc ggtaaacctc tggatttcgg aacaatatca\n",
      "    14281 ttggcactag cggtaccgca ggtgcttgga ggattatcct tcttgaatcc tgagaaatgt\n",
      "    14341 ttctaccgga atctaggaga tccagttacc tcaggtttat tccagttaaa aacttatctc\n",
      "    14401 cgaatgattg agatggatga tttattctta cctttaattg caaagaaccc tgggaactgc\n",
      "    14461 actgccattg actttgtgct aaatcctagc ggattaaatg tccctgggtc gcaagactta\n",
      "    14521 acttcatttc tgcgccagat tgtacgcagg accatcaccc taagtgcgaa aaacaaactt\n",
      "    14581 attaatacct tatttcatgc gtcagctgac ttcgaagacg aaatggtttg taaatggcta\n",
      "    14641 ttatcatcaa ctcctgttat gagtcgtttt gcggccgata tcttttcacg cacgccgagc\n",
      "    14701 gggaagcgat tgcaaattct aggatacctg gaaggaacac gcacattatt agcctctagg\n",
      "    14761 atcatcaaca ataatacaga gacaccggtt ttggacagac tgaggaaaat aacattgcaa\n",
      "    14821 aggtggagtc tgtggtttag ttatcttgat cattgtgata atatcctggc ggaggcttta\n",
      "    14881 acccaaataa cttgcacagt tgatttagca cagatcctga gggaatattc atgggctcat\n",
      "    14941 attttagagg gaagacctct tattggagcc acactcccat gtatgattga gcaattcaaa\n",
      "    15001 gtgttttggc tgaaacccta cgaacaatgt ccgcagtgtt caaatgcaaa gcaaccaggt\n",
      "    15061 gggaaaccat tcgtgtcagt ggcagtcaag aaacatattg ttagtgcatg gcctaacgca\n",
      "    15121 tcccgactaa gctggactat cggggatgga atcccttaca ttggatcaag gacagaagat\n",
      "    15181 aagataggac aacctgctat taaaccaaaa tgtccttccg cagccttaag agaggccatt\n",
      "    15241 gaattggcgt cccgtttaac atgggtaact caaggcagtt cgaacagtga cttgctaata\n",
      "    15301 aagccatttt tggaagcacg agtaaattta agtgttcaag aaatacttca aatgacccct\n",
      "    15361 tcacattact caggaaatat tgttcacagg tacaacgatc aatacagtcc tcattctttc\n",
      "    15421 atggccaatc gtatgagtaa ttcagcaacg cgattgattg tttctacaaa cactttaggt\n",
      "    15481 gagttttcag gaggtggcca gtctgcacgc gacagcaata ttattttcca gaatgttata\n",
      "    15541 aattatgcag ttgcactgtt cgatattaaa tttagaaaca ctgaggctac agatatccaa\n",
      "    15601 tataatcgtg ctcaccttca tctaactaag tgttgcaccc gggaagtacc agctcagtat\n",
      "    15661 ttaacataca catctacatt ggatttagat ttaacaagat accgagaaaa cgaattgatt\n",
      "    15721 tatgacaata atcctctaaa aggaggactc aattgcaata tctcattcga taacccattt\n",
      "    15781 ttccaaggta aacagctgaa cattatagaa gatgatctta ttcgactgcc tcacttatct\n",
      "    15841 ggatgggagc tagccaagac catcatgcaa tcaattattt cagatagcaa caattcatct\n",
      "    15901 acagacccaa ttagcagtgg agaaacaaga tcattcacta cccatttctt aacttatccc\n",
      "    15961 aagataggac ttctgtacag ttttggggcc tttgtaagtt attatcttgg caatacaatt\n",
      "    16021 cttcggacta agaaattgac acttgacaat tttttatatt acttaactac ccaaattcat\n",
      "    16081 aatctaccac atcgctcatt gcgaatactt aagccaacat tcaaacatgc aagcgttatg\n",
      "    16141 tcacggttaa tgagtattga tcctcatttt tctatttaca taggcggtgc tgcaggtgac\n",
      "    16201 agaggactct cagatgcggc caggttattt ttgagaacgt ctatttcatc ttttcttaca\n",
      "    16261 tttgtaaaag aatggataat taatcgcgga acaattgtcc ctttatggat agtatatccg\n",
      "    16321 ctagagggtc aaaacccaac acctgttaat aattttctct atcagatcgt agaactgctg\n",
      "    16381 gtgcatgatt catcaagaca acaggcttta aaaactacca taagtgatca tgtacatcct\n",
      "    16441 cacgacaatc ttgtttacac atgtaagagt acagccagca atttcttcca tgcatcattg\n",
      "    16501 gcgtactgga ggagcaggca cagaaacagc aaccgaaaat acttggcaag agactcttca\n",
      "    16561 actagatcaa gcacaaacaa cagtgatggt catattgaga gaagtcaaga acaaaccacc\n",
      "    16621 agagatccac atgatggcac tgaacggaat ctagttctac aaatgagcca tgaaataaaa\n",
      "    16681 agaacgacaa ttccacaaga aaacacgcac cagggtccgt cgttccagtc ctttctaagt\n",
      "    16741 tactctgctt gtggtacagc aaatccaaaa ctaaatttcg atcgatcgag acacaatgtg\n",
      "    16801 aaatctcagg atcataactc ggcatccaag agggaaggtc atcaaataat ctcacaccgt\n",
      "    16861 ctagtcctac ctttctttac attatctcaa gggacacgcc aattaacgtc atctaatgag\n",
      "    16921 tcacagaccc aagacgagat atcaaagtac ttacggcaat tgagatccgt cattgatacc\n",
      "    16981 acagtttatt gtaggtttac cggtatagtc tcgtccatgc attacaaact tgatgaggtc\n",
      "    17041 ctttgggaaa tagagagttt taagtcggct gtgacgctag cagagggaga aggtgctggt\n",
      "    17101 gccttactat tgattcagaa ataccaagtt aagaccttat ttttcaacac gctagctact\n",
      "    17161 gagtccagta tagagtcaga aatagtatca ggaatgacta ctcctaggat gcttctacct\n",
      "    17221 gttatgtcaa aattccataa tgaccaaatt gagattattc ttaacaactc agcaagccaa\n",
      "    17281 ataacagaca taacaaatcc tacttggttt aaagaccaaa gagcaaggct acctaagcaa\n",
      "    17341 gtcgaggtta taaccatgga tgcagagacg acagagaata taaacagatc gaaattgtac\n",
      "    17401 gaagctgtat ataaactgat cttacaccat attgatccca gcgtattgaa agcagtggtc\n",
      "    17461 cttaaagtct tcctaagtga tactgagggt atgttatggc taaatgataa tttagccccg\n",
      "    17521 ttttttgcca ctggttattt aattaagcca ataacgtcaa gtgctagatc tagtgagtgg\n",
      "    17581 tatctttgtc tgacgaactt cttatcaact acacgaaaga tgccacacca aaaccatctc\n",
      "    17641 agttgtaaac aggtaatact tacggcattg caactgcaaa ttcaacggag cccatactgg\n",
      "    17701 ctaagtcatt taactcagta tgctgactgt gatttacatt taagttatat ccgccttggt\n",
      "    17761 tttccatcat tagagaaagt actataccac aggtataacc tcgtcgattc aaaaagaggt\n",
      "    17821 ccactagtct ctatcactca gcacttagca catcttagag cagagattcg agaattaact\n",
      "    17881 aatgattata atcaacagcg acaaagtcga actcaaacat atcactttat tcgtactgca\n",
      "    17941 aaaggacgaa tcacaaaact agtcaatgat tatttaaaat tctttcttat tgtgcaagca\n",
      "    18001 ttaaaacata atgggacatg gcaagctgag tttaagaaat taccagagtt gattagtgtg\n",
      "    18061 tgcaataggt tctaccatat tagagattgc aattgtgaag aacgtttctt agttcaaacc\n",
      "    18121 ttatatttac atagaatgca ggattctgaa gttaagctta tcgaaaggct gacagggctt\n",
      "    18181 ctgagtttat ttccggatgg tctctacagg tttgattgaa ttaccgtgca tagtatcctg\n",
      "    18241 atacttgcaa aggttggtta tcaacataca gattataaaa aactcataaa ttgctctcat\n",
      "    18301 acatcatatt gatctgattt caataaacaa ctatttaaat aacgaaagga gtccctatat\n",
      "    18361 tatacactat atttagcctc tctccctgcg tgataatcaa aaaattcaca atgcagcatg\n",
      "    18421 tgtgacatat taatgccgca atgaatttaa cgcaacataa taaactctgc actctttata\n",
      "    18481 attaagcttt aacgaaaggt ctgggctcat attgttattg atataataat gttatatcaa\n",
      "    18541 tgtcctgtca gatggaatag tgttttggtt gataacacga cttcttaaaa caaaattgat\n",
      "    18601 ctttaagatt aagtttttta taattatcat tactttaatt tatcgatttg aaaatggtaa\n",
      "    18661 tagccttaat ctttgtgtaa aataagagat taggtgtaat aactttaaca tttttgtcta\n",
      "    18721 gtaagttact atttcataca gaatgataaa attaaaagaa aaggcaggac tgtaaaatca\n",
      "    18781 taaatacctt ctttacaata tagcagacta gataataatc tttgtgttaa tgataattaa\n",
      "    18841 gacattgacc acgctcatca gaaggctcgc caaaataaac gttgcaaaaa ggattcctgg\n",
      "    18901 aaaaatggtc gcacacaaaa atttaaaaat aaatctattt cttctttttt gtgtgtcca\n",
      "//\n",
      "\n",
      "Annotation([Feature(\"gene\", [Location(6039, 6933, strand=Location.Strand.FORWARD, defect=Location.Defect.NONE)], qual={'gene': 'SsGP', 'locus_tag': 'DF49_53416gpSsGP'}), Feature(\"gene\", [Location(470, 2689, strand=Location.Strand.FORWARD, defect=Location.Defect.NONE)], qual={'gene': 'NP', 'locus_tag': 'DF49_53416gpNP'}), Feature(\"CDS\", [Location(11581, 18219, strand=Location.Strand.FORWARD, defect=Location.Defect.NONE)], qual={'gene': 'L', 'locus_tag': 'DF49_53416gpL', 'codon_start': '1', 'product': 'RNA-directed RNA polymerase L', 'protein_id': 'ALT19757.1', 'translation': 'MATQHTQYPDARLSSPIVLDQCDLVTRACGLYSSYSLNPQLRNC KLPKHIYRLKYDVTVTKFLSDVPVATLPIDFIVPILLKALSGNGFCPVEPRCQQFLDE IIKYTMQDALFLKYYLKNVGAQEDCVDDHFQEKILSSIQGNEFLHQMFFWYDLAILTR RGRLNRGNSRSTWFVHDDLIDILGYGDYVFWKIPISMLPLNTQGIPHAAMDWYQASVF KEAVQGHTHIVSVSTADVLIMCKDLITCRFNTTLISKIAEIEDPVCSDYPNFKIVSML YQSGDYLLSILGSDGYKIIKFLEPLCLAKIQLCSKYTERKGRFLTQMHLAVNHTLEEI TEMRALKPSQAQKIREFHRTLIRLEMTPQQLCELFSIQKHWGHPVLHSETAIQKVKKH ATVLKALRPIVIFETYCVFKYSIAKHYFDSQGSWYSVTSDRNLTPGLNSYIKRNQFPP LPMIKELLWEFYHLDHPPLFSTKIISDLSIFIKDRATAVERTCWDAVFEPNVLGYNPP HKFSTKRVPEQFLEQENFSIENVLSYAQKLEYLLPQYRNFSFSLKEKELNVGRTFGKL PYPTRNVQTLCEALLADGLAKAFPSNMMVVTEREQKESLLHQASWHHTSDDFGEHATV RGSSFVTDLEKYNLAFRYEFTAPFIEYCNRCYGVKNVFNWMHYTIPQCYMHVSDYYNP PHNLTLENRDNPPEGPSSYRGHMGGIEGLQQKLWTSISCAQISLVEIKTGFKLRSAVM GDNQCITVLSVFPLETDADEQEQSAEDNAARVAASLAKVTSACGIFLKPDETFVHSGF IYFGKKQYLNGVQLPQSLKTATRMAPLSDAIFDDLQGTLASIGTAFERSISETRHIFP CRITAAFHTFFSVRILQYHHLGFNKGFDLGQLTLGKPLDFGTISLALAVPQVLGGLSF LNPEKCFYRNLGDPVTSGLFQLKTYLRMIEMDDLFLPLIAKNPGNCTAIDFVLNPSGL NVPGSQDLTSFLRQIVRRTITLSAKNKLINTLFHASADFEDEMVCKWLLSSTPVMSRF AADIFSRTPSGKRLQILGYLEGTRTLLASRIINNNTETPVLDRLRKITLQRWSLWFSY LDHCDNILAEALTQITCTVDLAQILREYSWAHILEGRPLIGATLPCMIEQFKVFWLKP YEQCPQCSNAKQPGGKPFVSVAVKKHIVSAWPNASRLSWTIGDGIPYIGSRTEDKIGQ PAIKPKCPSAALREAIELASRLTWVTQGSSNSDLLIKPFLEARVNLSVQEILQMTPSH YSGNIVHRYNDQYSPHSFMANRMSNSATRLIVSTNTLGEFSGGGQSARDSNIIFQNVI NYAVALFDIKFRNTEATDIQYNRAHLHLTKCCTREVPAQYLTYTSTLDLDLTRYRENE LIYDNNPLKGGLNCNISFDNPFFQGKQLNIIEDDLIRLPHLSGWELAKTIMQSIISDS NNSSTDPISSGETRSFTTHFLTYPKIGLLYSFGAFVSYYLGNTILRTKKLTLDNFLYY LTTQIHNLPHRSLRILKPTFKHASVMSRLMSIDPHFSIYIGGAAGDRGLSDAARLFLR TSISSFLTFVKEWIINRGTIVPLWIVYPLEGQNPTPVNNFLYQIVELLVHDSSRQQAL KTTISDHVHPHDNLVYTCKSTASNFFHASLAYWRSRHRNSNRKYLARDSSTRSSTNNS DGHIERSQEQTTRDPHDGTERNLVLQMSHEIKRTTIPQENTHQGPSFQSFLSYSACGT ANPKLNFDRSRHNVKSQDHNSASKREGHQIISHRLVLPFFTLSQGTRQLTSSNESQTQ DEISKYLRQLRSVIDTTVYCRFTGIVSSMHYKLDEVLWEIESFKSAVTLAEGEGAGAL LLIQKYQVKTLFFNTLATESSIESEIVSGMTTPRMLLPVMSKFHNDQIEIILNNSASQ ITDITNPTWFKDQRARLPKQVEVITMDAETTENINRSKLYEAVYKLILHHIDPSVLKA VVLKVFLSDTEGMLWLNDNLAPFFATGYLIKPITSSARSSEWYLCLTNFLSTTRKMPH QNHLSCKQVILTALQLQIQRSPYWLSHLTQYADCDLHLSYIRLGFPSLEKVLYHRYNL VDSKRGPLVSITQHLAHLRAEIRELTNDYNQQRQSRTQTYHFIRTAKGRITKLVNDYL KFFLIVQALKHNGTWQAEFKKLPELISVCNRFYHIRDCNCEERFLVQTLYLHRMQDSE VKLIERLTGLLSLFPDGLYRFD'}), Feature(\"CDS\", [Location(10345, 11100, strand=Location.Strand.FORWARD, defect=Location.Defect.NONE)], qual={'gene': 'VP24', 'locus_tag': 'DF49_53416gpVP24', 'note': 'secondary matrix protein', 'codon_start': '1', 'product': 'membrane-associated protein VP24', 'protein_id': 'ALT19756.1', 'translation': 'MAKATGRYNLISPKKDLEKGVVLSDLCNFLVSQTIQGWKVYWAG IEFDVTHKGMALLQRLKTNDFAPAWSMTRNLFPHLFQNPNSTIESPLWALRVILAAGI QDQLIDQSLIEPLAGALGLISDWLLTTNTNHFNMRTQRVKEQLSLKMLSLIRSNILKF INKLDALHVVNYNGLLSSIEIGTQNHTIIITRTNMGFLVELQEPDKSAMNRKKPGPAK FSLLHESTLKAFTQGSSTRMQSLILEFNSSLAI'}), Feature(\"CDS\", [Location(4479, 5459, strand=Location.Strand.FORWARD, defect=Location.Defect.NONE)], qual={'gene': 'VP40', 'locus_tag': 'DF49_53416gpVP40', 'codon_start': '1', 'product': 'matrix protein VP40', 'protein_id': 'ALT19751.1', 'translation': 'MRRVILPTAPPEYMEAIYPVRSNSTIARGGNSNTGFLTPESVNG DTPSNPLRPIADDTIDHASHTPGSVSSAFILEAMVNVISGPKVLMKQIPIWLPLGVAD QKTYSFDSTTAAIMLASYTITHFGKATNPLVRVNRLGPGIPDHPLRLLRIGNQAFLQE FVLPPVQLPQYFTFDLTALKLITQPLPAATWTDDTPTGSNGALRPGISFHPKLRPILL PNKSGKKGNSADLTSPEKIQAIMTSLQDFKIVPIDPTKNIMGIEVPETLVHKLTGKKV TSKNGQPIIPVLLPKYIGLDPVAPGDLTMVITQDCDTCHSPASLPAVIEK'}), Feature(\"gene\", [Location(3129, 4151, strand=Location.Strand.FORWARD, defect=Location.Defect.NONE)], qual={'gene': 'VP35', 'locus_tag': 'DF49_53416gpVP35'}), Feature(\"misc_feature\", [Location(6917, 6924, strand=Location.Strand.FORWARD, defect=Location.Defect.NONE)], qual={'gene': 'SsGP', 'locus_tag': 'DF49_53416gpSsGP', 'note': 'location of RNA editing (TAAAAAAA => TAAAAAAAAA) in super small secreted glycoprotein SsGP'}), Feature(\"gene\", [Location(8509, 9375, strand=Location.Strand.FORWARD, defect=Location.Defect.NONE)], qual={'gene': 'VP30', 'locus_tag': 'DF49_53416gpVP30'}), Feature(\"gene\", [Location(10345, 11100, strand=Location.Strand.FORWARD, defect=Location.Defect.NONE)], qual={'gene': 'VP24', 'locus_tag': 'DF49_53416gpVP24'}), Feature(\"gene\", [Location(6039, 7133, strand=Location.Strand.FORWARD, defect=Location.Defect.NONE)], qual={'gene': 'sGP', 'locus_tag': 'DF49_53416gpsGP'}), Feature(\"CDS\", [Location(6039, 8068, strand=Location.Strand.FORWARD, defect=Location.Defect.NONE)], qual={'gene': 'GP', 'locus_tag': 'DF49_53416gpGP', 'exception': 'RNA editing', 'note': 'synonym: preGP; non-templated A inserted during transcription due to RNA editing', 'codon_start': '1', 'product': 'envelope glycoprotein', 'protein_id': 'ALT19754.1', 'translation': 'MGVTGILQLPRDRFKRTSFFLWVIILFQRTFSIPLGVIHNSTLQ VSDVDKLVCRDKLSSTNQLRSVGLNLEGNGVATDVPSATKRWGFRSGVPPKVVNYEAG EWAENCYNLEIKKPDGSECLPAAPDGIRGFPRCRYVHKVSGTGPCAGDFAFHKEGAFF LYDRLASTVIYRGTTFAEGVVAFLILPQAKKDFFSSHPLREPVNATEDPSSGYYSTTI RYQATGFGTNETEYLFEVDNLTYVQLESRFTPQFLLQLNETIYTSGKRSNTTGKLIWK VNPEIDTTIGEWAFWETKKNLTRKIRSEELSFTAVSNRAKNISGQSPARTSSDPGTNT TTEDHKIMASENSSAMVQVHSQGREAAVSHLTTLATISTSPQPPTTKPGPDNSTHNTP VYKLDISEATQVEQHHRRTDNDSTASDTPPATTAAGPLKAENTNTSKGTDLLDPATTT SPQNHSETAGNNNTHHQDTGEESASSGKLGLITNTIAGVAGLITGGRRARREAIVNAQ PKCNPNLHYWTTQDEGAAIGLAWIPYFGPAAEGIYIEGLMHNQDGLICGLRQLANETT QALQLFLRATTELRTFSILNRKAIDFLLQRWGGTCHILGPDCCIEPHDWTKNITDKID QIIHDFVDKTLPDQGDNDNWWTGWRQWIPAGIGVTGVIIAVIALFCICKFVF'}), Feature(\"misc_feature\", [Location(6917, 6924, strand=Location.Strand.FORWARD, defect=Location.Defect.NONE)], qual={'gene': 'GP', 'locus_tag': 'DF49_53416gpGP', 'note': 'location of RNA editing (TAAAAAAA => TAAAAAAAA) in envelope glycoprotein'}), Feature(\"CDS\", [Location(6039, 6933, strand=Location.Strand.FORWARD, defect=Location.Defect.NONE)], qual={'gene': 'SsGP', 'locus_tag': 'DF49_53416gpSsGP', 'exception': 'RNA editing', 'note': \"two non-templated A's inserted during transcription due to RNA editing\", 'codon_start': '1', 'product': 'super small secreted glycoprotein SsGP', 'protein_id': 'ALT19752.1', 'translation': 'MGVTGILQLPRDRFKRTSFFLWVIILFQRTFSIPLGVIHNSTLQ VSDVDKLVCRDKLSSTNQLRSVGLNLEGNGVATDVPSATKRWGFRSGVPPKVVNYEAG EWAENCYNLEIKKPDGSECLPAAPDGIRGFPRCRYVHKVSGTGPCAGDFAFHKEGAFF LYDRLASTVIYRGTTFAEGVVAFLILPQAKKDFFSSHPLREPVNATEDPSSGYYSTTI RYQATGFGTNETEYLFEVDNLTYVQLESRFTPQFLLQLNETIYTSGKRSNTTGKLIWK VNPEIDTTIGEWAFWETKKKPH'}), Feature(\"CDS\", [Location(3129, 4151, strand=Location.Strand.FORWARD, defect=Location.Defect.NONE)], qual={'gene': 'VP35', 'locus_tag': 'DF49_53416gpVP35', 'codon_start': '1', 'product': 'polymerase cofactor VP35', 'protein_id': 'ALT19750.1', 'translation': 'MTTRTKGRGHTAATTQNDRMPGPELSGWISEQLMTGRIPVSDIF CDIENNPGLCYASQMQQTKPNPKTRNSQTQTDPICNHSFEEVVQTLASLATVVQQQTI ASESLEQRITSLENGLKPVYDMAKTISSLNRVCAEMVAKYDLLVMTTGRATATAAATE AYWAEHGQPPPGPSLYEESAIRGKIESRDETVPQSVREAFNNLDSTTSLTEENFGKPD ISAKDLRNIMYDHLPGFGTAFHQLVQVICKLGKDSNSLDIIHAEFQASLAEGDSPQCA LIQITKRVPIFQDAAPPVIHIRSRGDIPRACQKSLRPVPPSPKIDRGWVCVFQLQDGK TLGLKI'}), Feature(\"misc_feature\", [Location(1, 18959, strand=Location.Strand.FORWARD, defect=Location.Defect.NONE)], qual={'db_xref': 'ViPR:FLVSP_ABEBO2_00005.MAIN'}), Feature(\"source\", [Location(1, 18959, strand=Location.Strand.FORWARD, defect=Location.Defect.NONE)], qual={'organism': 'Ebola virus', 'mol_type': 'viral cRNA', 'isolate': 'Ebola virus/H. sapiens-tc/COD/1995/Kikwit-9510622', 'host': 'Homo sapiens; female; 65 years', 'db_xref': 'taxon:1570291', 'lab_host': 'Vero E6 cells', 'country': 'Zaire', 'collection_date': '04-May-1995', 'PCR_primers': 'fwd_name: 3RACE_Forward, fwd_seq: ccaaaacgccatttccaccttctcttc, rev_name: 3RACE_Reverse, rev_seq: ccgcactctcttgaaaatcaacacc\\nfwd_name: 5RACE_Forward, fwd_seq: ccttctttacaatatagcagactagataataat, rev_name: 5RACE_Reverse, rev_seq: ccaaaacgccatttccaccttctcttc\\nfwd_name: 3_5_self-ligation_Forward, fwd_seq: gtgttaatgataattaagrcattgacc, rev_name: 3_5_self-ligation_Reverse, rev_seq: aagggattttcaactgagca\\nfwd_name: Zaire_1a_Forward, fwd_seq: ttgtgtgcgaataactatgagg, rev_name: Zaire_1a_Reverse, rev_seq: caagtgttttaccatcttgaagc\\nfwd_name: Zaire_1b_Forward, fwd_seq: tcctctcattgaaatttatatcgg, rev_name: Zaire_1b_Reverse, rev_seq: ttcagcagttgaagcctctgc\\nfwd_name: Zaire_2a_Forward, fwd_seq: gagttccaatcttccaagatgc, rev_name: Zaire_2a_Reverse, rev_seq: catttgtcaagtaatcttagattcaag\\nfwd_name: Zaire_2b_Forward, fwd_seq: gacattccccgagcttgcc, rev_name: Zaire_2b_Reverse, rev_seq: ttaaargagttagaatcacattggc\\nfwd_name: Zaire_3a_Forward, fwd_seq: tggataccggcaggtattgg, rev_name: Zaire_3a_Reverse, rev_seq: gtaacattgaaatygggatcttcc\\nfwd_name: Zaire_3b_Forward, fwd_seq: agttatygctttattctgtatatgc, rev_name: Zaire_3b_Reverse, rev_seq: ttgaatactgatgcctgatacc\\nfwd_name: Zaire_4a_Forward, fwd_seq: tagattaaatcgaggaaactctag, rev_name: Zaire_4a_Reverse, rev_seq: caactgcataatttataacattctgg\\nfwd_name: Zaire_4b_Forward, fwd_seq: catgatgatttaatagacatcttagg, rev_name: Zaire_4b_Reverse, rev_seq: ttagttagatgaaggtgagcacg\\nfwd_name: Zaire_5a_Forward, fwd_seq: gaaaccattcgtgtcagtggc, rev_name: Zaire_5a_Reverse, rev_seq: atttttaaatttttgtgtgcgacc\\nfwd_name: Zaire_5b_Forward, fwd_seq: gataggacaacctgctattaaacc, rev_name: Zaire_5b_Reverse, rev_seq: tccaggaatcctttttgcaacg\\nfwd_name: Zaire_2aV2_Forward, fwd_seq: actgctttccaccarttagtaca, rev_name: Zaire_2aV2_Reverse, rev_seq: argagttagaatcacattggct\\nfwd_name: Zaire_2bV2_Forward, fwd_seq: atyattcatgctgagttccagg, rev_name: Zaire_2bV2_Reverse, rev_seq: caatgctttrtgtgtgttaggtt\\nfwd_name: Zaire_AdapLig_II_Forward, fwd_seq: gtgttaatgataattaagrcattgacc, rev_name: Zaire_AdapLig_II_Reverse, rev_seq: gaagagaaggtggaaatggcgttttgg\\nfwd_name: Zaire_AdapLig_III_Forward, fwd_seq: ccaaaacgccatttccaccttctcttc, rev_name: Zaire_AdapLig_III_Reverse, rev_seq: tgattacagtaacaatttcaatttaaattc', 'note': 'Replicate 2: Virus propagated in confluent Vero E6 cell monolayer at 0.01 MOI using P2 material in 850 cm2 roller bottle (n=1) without antibiotics. Supernatant collected upon reaching 70% CPE. Harvested supernatant was clarified via centrifugation (10,000 x g for 10 min). Clarified supernatant was supplemented with 16.5% HI-FBS and 0.01M Tris pH 8.5.'}), Feature(\"CDS\", [Location(6039, 7133, strand=Location.Strand.FORWARD, defect=Location.Defect.NONE)], qual={'gene': 'sGP', 'locus_tag': 'DF49_53416gpsGP', 'codon_start': '1', 'product': 'small secreted glycoprotein sGP', 'protein_id': 'ALT19753.1', 'translation': 'MGVTGILQLPRDRFKRTSFFLWVIILFQRTFSIPLGVIHNSTLQ VSDVDKLVCRDKLSSTNQLRSVGLNLEGNGVATDVPSATKRWGFRSGVPPKVVNYEAG EWAENCYNLEIKKPDGSECLPAAPDGIRGFPRCRYVHKVSGTGPCAGDFAFHKEGAFF LYDRLASTVIYRGTTFAEGVVAFLILPQAKKDFFSSHPLREPVNATEDPSSGYYSTTI RYQATGFGTNETEYLFEVDNLTYVQLESRFTPQFLLQLNETIYTSGKRSNTTGKLIWK VNPEIDTTIGEWAFWETKKTSLEKFAVKSCLSQLYQTEPKTSVVRVRRELLPTQGPTQ QLKTTKSWLQKIPLQWFKCTVKEGKLQCRI'}), Feature(\"CDS\", [Location(470, 2689, strand=Location.Strand.FORWARD, defect=Location.Defect.NONE)], qual={'gene': 'NP', 'locus_tag': 'DF49_53416gpNP', 'codon_start': '1', 'product': 'nucleoprotein', 'protein_id': 'ALT19749.1', 'translation': 'MDSRPQKVWMTPSLTESDMDYHKILTAGLSVQQGIVRQRVIPVY QVNNLEEICQLIIQAFEAGVDFQESADSFLLMLCLHHAYQGDYKLFLESGAVKYLEGH GFRFEVKKRDGVKRLEELLPAVSSGKNIKRTLAAMPEEETTEANAGQFLSFASLFLPK LVVGEKACLEKVQRQIQVHAEQGLIQYPTAWQSVGHMMVIFRLMRTNFLIKFLLIHQG MHMVAGHDANDAVISNSVAQARFSGLLIVKTVLDHILQKTERGVRLHPLARTAKVKNE VNSFKAALSSLAKHGEYAPFARLLNLSGVNNLEHGLFPQLSAIALGVATAHGSTLAGV NVGEQYQQLREAATEAEKQLQQYAESRELDHLGLDDQEKKILMNFHQKKNEISFQQTN AMVTLRKERLAKLTEAITAASLPKTSGHYDDDDDIPFPGPINDDDNPGHQDDDPTDSQ DTTIPDVVVDPDDGSYGEYQSYSENGMNAPDDLVLFDLDEDDEDTKPVPNRSTKGGQQ KNSQKGQHTEGRQTQSRPTQNVPGPHRTIHHASAPLTDNDRRNEPSGSTSPRMLTPIN EEADPLDDADDETSSLPPLESDDEEQDRDGTSNRTPTVAPPAPVYRDHSEKRELPQDE QQDQDHTQEARNQDSDNTQPEHSFEEMYRHILRSQGPFDAVLYYHMMKDEPVVFSTSD GKEYTYPDSLEEEYPPWLTEKEAMNEENRFVTLDGQQFYWPVMNHKNKFMAILQHHQ'}), Feature(\"gene\", [Location(4479, 5459, strand=Location.Strand.FORWARD, defect=Location.Defect.NONE)], qual={'gene': 'VP40', 'locus_tag': 'DF49_53416gpVP40'}), Feature(\"gene\", [Location(11581, 18219, strand=Location.Strand.FORWARD, defect=Location.Defect.NONE)], qual={'gene': 'L', 'locus_tag': 'DF49_53416gpL'}), Feature(\"gene\", [Location(6039, 8068, strand=Location.Strand.FORWARD, defect=Location.Defect.NONE)], qual={'gene': 'GP', 'locus_tag': 'DF49_53416gpGP'}), Feature(\"CDS\", [Location(8509, 9375, strand=Location.Strand.FORWARD, defect=Location.Defect.NONE)], qual={'gene': 'VP30', 'locus_tag': 'DF49_53416gpVP30', 'note': 'synonym: transcription activator VP30', 'codon_start': '1', 'product': 'minor nucleoprotein VP30', 'protein_id': 'ALT19755.1', 'translation': 'MEASYERGRPRAARQHSRDGHDHHVRARSSSRENYRGEYRQSRS ASQVRVPTVFHKKRVEPLTVPPAPKDICPTLKKGFLCDSSFCKKDHQLESLTDRELLL LIARKTCGSVEQQLNITAPKDSRLANPTADDFQQEEGPKITLLTLIKTAEHWARQDIR TIEDSKLRALLTLCAVMTRKFSKSQLSLLCETHLRREGLGQDQAEPVLEVYQRLHSDK GGSFEAALWQQWDRQSLIMFITAFLNIALQLPCESSAVVVSGLRTLVPQSDNEEASTN PGTCSWSDEGTP'})])\n"
     ]
    }
   ],
   "source": [
    "import biotite.sequence.io.fasta as fasta\n",
    "\n",
    "dir(fasta)\n",
    "\n",
    "# read fasta file including fasta header\n",
    "file = fasta.FastaFile()\n",
    "file.read(\"DNAToolkit/Examples/ebola_sequence.fasta\")\n",
    "\n",
    "# only sequence without fasta header\n",
    "ebola_sequence = fasta.get_sequence(file)\n",
    "# print(ebola_sequence)\n",
    "\n",
    "# reading genbank file\n",
    "import biotite.sequence.io.genbank as gb\n",
    "\n",
    "dir(gb)\n",
    "\n",
    "gfile = gb.GenBankFile()\n",
    "gfile.read(\"DNAToolkit/Examples/ebola_sequence.gb\")\n",
    "print(gfile)\n",
    "\n",
    "ebola_seq = gb.get_annotation(gfile)\n",
    "print(ebola_seq)\n",
    "\n",
    "# read/downloading using the ID"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['2123541081', '2123540882', '2123540881', '2123540880', '2123540879', '2123540878', '2123540877', '2123540876', '2123540875', '2123540874']\n",
      "KU182909.1 Ebola virus isolate Ebola virus/H. sapiens-tc/COD/1995/Kikwit-9510622, complete genome\n",
      "CGGACACACAAAAAGAAAGAAGAATTTTTAGGATCTTTTGTGTGCGAATAACTATGAGGAAGATTAATAATTTTCCTCTCATTGAAATTTATATCGGAATTTAAATTGAAATTGTTACTGTAATCACACCTGGTTTGTTTCAGAGCCACATCACAAAGATAGAGAACAACCTAGGTCTCTGAAGGGAGCAAGGGCATCAGTGTGCTCAGTTGAAAATCCCTTGTCAACATCTAGGTCTTATCACATCACAAGTCCCACCTCAGACTCTGCAGGGTGATCCAACAACCTTAATAGAAACATTATTGTTAAAGGACAGCATTAGTTCACAGTCAAACAAGCAAGATTGAGAATTAACCTTGGTTTTGAACTTGAATACTTAGAGGATTGGAGATTCAACAACCCTAAAGCTTGGGGTAAAACATTGGAAATAGTTAAAAGACAAATTGCTCGGAATCACAACATTCCGAGTATGGATTCTCGTCCTCAGAAAGTCTGGATGACGCCGAGTCTCACTGAATCTGACATGGATTACCACAAGATCTTGACAGCAGGTCTGTCCGTTCAACAGGGGATTGTTCGGCAAAGAGTCATCCCAGTGTATCAAGTAAACAATCTTGAGGAGATTTGCCAACTTATCATACAGGCCTTTGAAGCAGGTGTTGATTTTCAAGAGAGTGCGGACAGTTTCCTTCTCATGCTTTGTCTTCATCATGCGTACCAGGGAGATTACAAACTTTTCTTGGAAAGTGGCGCAGTCAAGTATTTGGAAGGGCACGGGTTCCGTTTTGAAGTCAAGAAGCGTGATGGAGTGAAGCGCCTTGAGGAATTGCTGCCAGCAGTATCTAGTGGAAAAAACATTAAGAGAACACTTGCTGCCATGCCGGAAGAGGAGACAACTGAAGCTAATGCCGGTCAGTTTCTCTCCTTTGCAAGTCTATTCCTTCCGAAATTGGTAGTAGGAGAAAAGGCTTGTCTTGAGAAGGTTCAAAGGCAAATTCAAGTACATGCAGAGCAAGGACTGATACAATATCCAACAGCTTGGCAATCAGTAGGACACATGATGGTGATTTTCCGTTTGATGCGAACAAATTTTTTGATCAAATTTCTCCTAATACACCAAGGGATGCACATGGTTGCCGGGCATGATGCCAACGACGCTGTGATTTCAAATTCAGTGGCTCAAGCTCGTTTTTCAGGTTTATTGATTGTCAAAACAGTACTTGATCATATCCTACAAAAGACAGAACGCGGAGTTCGTCTCCATCCTCTTGCAAGGACCGCCAAGGTAAAAAATGAGGTGAACTCCTTTAAGGCTGCACTCAGCTCCCTGGCCAAGCATGGAGAGTATGCTCCTTTCGCCCGACTTTTGAACCTTTCTGGAGTAAATAATCTTGAGCATGGTCTTTTCCCTCAACTATCAGCAATTGCACTCGGAGTCGCCACAGCACACGGGAGTACCCTCGCAGGAGTAAATGTTGGAGAACAGTATCAACAACTCAGAGAGGCTGCCACTGAGGCTGAGAAGCAACTCCAACAATACGCAGAGTCTCGCGAACTTGACCATCTTGGACTTGATGATCAGGAAAAGAAAATTCTTATGAACTTCCATCAGAAAAAGAACGAAATCAGCTTCCAGCAAACAAACGCTATGGTAACTCTAAGAAAAGAGCGCCTGGCCAAGCTGACGGAAGCTATCACTGCTGCGTCACTGCCCAAAACAAGTGGACATTACGATGATGATGACGACATTCCCTTTCCAGGACCCATCAATGATGACGACAATCCTGGCCATCAAGATGATGATCCAACTGACTCACAAGATACGACCATTCCTGATGTGGTGGTTGATCCCGATGATGGAAGCTACGGCGAATACCAGAGTTACTCGGAAAACGGCATGAATGCACCAGATGACTTGGTCCTATTCGATCTAGACGAGGACGATGAGGACACTAAGCCAGTGCCTAATAGATCAACCAAGGGTGGACAACAGAAAAACAGTCAAAAGGGCCAGCATACAGAGGGCAGACAGACACAATCCAGGCCAACTCAAAATGTCCCAGGCCCTCACAGAACAATCCACCACGCCAGTGCTCCACTCACGGATAATGACAGAAGAAATGAACCCTCCGGCTCAACCAGCCCTCGCATGCTGACACCAATCAACGAAGAGGCAGACCCACTGGACGATGCCGACGACGAGACGTCTAGCCTTCCGCCCCTGGAGTCAGACGATGAAGAACAGGACAGGGACGGAACTTCCAACCGCACACCCACTGTCGCCCCACCGGCTCCCGTATACAGAGATCACTCTGAAAAGAGAGAACTCCCGCAAGATGAGCAACAAGATCAGGACCACACTCAAGAGGCCAGGAACCAGGACAGTGACAACACCCAGCCAGAACACTCTTTTGAGGAGATGTATCGCCACATTCTAAGATCACAGGGGCCATTTGATGCTGTTTTGTATTATCATATGATGAAGGATGAGCCTGTAGTTTTCAGTACCAGTGATGGCAAAGAGTACACGTATCCAGACTCCCTTGAAGAGGAATATCCACCATGGCTCACTGAAAAAGAGGCTATGAATGAAGAGAATAGATTTGTTACATTGGATGGTCAACAATTTTATTGGCCGGTAATGAATCACAAGAATAAATTCATGGCAATCCTGCAACATCATCAGTGAATGAGCATGGAACAATGGGATGATTCAACCGACAAATAGCTAACATTAGGTAGTCAAGGAACGAAAACAGGAAGAATTTTTGATGTCTAAGGTGTGAATTATTATCACAATAAAAGTGATTCTTATTTTTGAATTTAAAGCTAGCTTATTATTACTAGCCGTTTTTCAAAGTCCAATTTGAGTCTTAATGCAAATAGGCGTTAAGCCACAGTTATAGCCATAATTGTAACTCAATATCCTAACTAGCGATTTATCTAAATTAAATTACATTATGCTTTTATAACTTACCTACTAGCCTACCCAACATTTACACGATCATTTTATAATTAAGAAAAAACTAATGATGAAGATTAAAACCTTCATCATCCTTACGTCAATTGAATTCTCTAGCACTCGAAGCTTATTGTCCTCAATGTAAAAGAAAAGCTGGTCTAACAAGATGACAACCAGAACAAAGGGCAGGGGCCACACTGCGGCCACGACTCAAAACGACAGAATGCCAGGCCCTGAGCTTTCGGGCTGGATCTCTGAGCAGCTAATGACCGGAAGAATTCCTGTAAGCGACATCTTCTGTGATATTGAGAACAATCCAGGATTATGCTACGCATCCCAAATGCAACAAACCAAGCCAAACCCGAAGACGCGCAACAGTCAAACCCAAACGGACCCAATTTGCAATCATAGTTTTGAGGAGGTAGTACAAACATTAGCTTCATTGGCTACTGTTGTGCAACAACAAACCATTGCATCAGAATCATTAGAACAACGCATTACGAGTCTTGAGAATGGTCTAAAGCCAGTTTATGATATGGCTAAAACAATCTCCTCATTGAACAGGGTTTGTGCTGAGATGGTTGCAAAATATGATCTTCTGGTGATGACAACCGGTCGGGCAACAGCAACCGCTGCGGCAACTGAGGCTTATTGGGCCGAACATGGTCAACCACCACCTGGACCATCACTTTATGAAGAAAGTGCAATTCGGGGTAAGATTGAATCTAGAGATGAGACCGTCCCTCAAAGTGTTAGGGAGGCATTCAACAATCTAGACAGTACCACTTCACTAACTGAGGAAAATTTTGGGAAACCTGACATTTCGGCAAAGGATTTGAGAAACATTATGTATGATCACTTGCCTGGTTTTGGAACTGCTTTCCACCAATTAGTACAAGTGATTTGTAAATTGGGAAAAGATAGCAACTCATTGGACATCATTCATGCTGAGTTCCAGGCCAGCCTGGCTGAAGGAGACTCTCCTCAATGTGCCCTAATTCAAATTACAAAAAGAGTTCCAATCTTCCAAGATGCTGCTCCACCTGTCATCCACATCCGCTCTCGAGGTGACATTCCCCGAGCTTGCCAGAAAAGCTTGCGTCCAGTCCCACCATCGCCCAAGATTGATCGAGGTTGGGTATGTGTTTTTCAGCTTCAAGATGGTAAAACACTTGGACTCAAAATTTGAGCCCATCTTCCTTCCCTCCGAAAGAGGCGACTAATAGCAGAGGCTTCAACTGCTGAACTATAGGGTACGTTACATTAATGATACACTTGTGAGTATCAGCCCTGGATAATATAAGTCAATTAAACGACCAAGATAAAATTGTTCTTATCTCGCTAGCAGCTTAAAATATGAATGTAATAGGAGCTATATCTCTGACAGTATTATAATCAATCGTTATTAAGTAACCCAAACCAAAAGTGATGAAGATTAAGAAAAACCTACCTCGACTGAGAGAGTGTTTTTTCATTAACCTTCATCTTGTAAACGTTGAGCAAAATTGTTAAAAATATGAGGCGGGTTATATTACCTACTGCTCCTCCTGAATATATGGAGGCCATATACCCTGTCAGGTCAAATTCAACAATTGCTAGAGGTGGCAACAGCAATACAGGCTTCCTGACACCGGAGTCAGTCAATGGGGACACTCCATCGAATCCACTCAGGCCAATTGCCGATGACACCATCGACCATGCCAGCCACACACCAGGCAGTGTGTCATCAGCATTCATCCTTGAAGCTATGGTGAATGTCATATCGGGCCCCAAAGTGCTAATGAAGCAAATTCCAATTTGGCTTCCTCTAGGTGTCGCTGATCAAAAGACCTACAGCTTTGACTCAACAACGGCCGCCATCATGCTTGCTTCATATACTATCACCCATTTCGGCAAGGCAACCAATCCACTTGTCAGAGTCAATCGGCTGGGTCCTGGAATCCCGGATCACCCCCTCAGGCTCCTGCGAATTGGAAACCAGGCCTTCCTCCAGGAGTTCGTTCTTCCGCCAGTCCAACTACCCCAGTATTTCACCTTTGATTTGACAGCACTCAAACTGATCACCCAACCACTGCCTGCTGCAACATGGACCGATGACACTCCAACAGGATCAAATGGAGCGTTGCGCCCAGGGATTTCATTTCATCCAAAACTTCGCCCCATTCTTTTACCCAACAAGAGTGGGAAGAAGGGGAATAGTGCCGATCTAACATCTCCGGAGAAAATCCAAGCAATAATGACTTCACTCCAGGACTTTAAGATCGTTCCAATTGATCCAACCAAAAATATCATGGGAATCGAAGTGCCAGAAACTCTGGTCCACAAGCTGACCGGTAAGAAGGTGACTTCTAAAAATGGACAACCAATCATCCCTGTTCTTTTGCCAAAGTACATTGGTTTGGACCCGGTGGCTCCAGGAGACCTCACCATGGTAATCACACAGGATTGTGACACGTGTCATTCTCCTGCGAGTCTTCCAGCTGTGATTGAGAAGTAATTGCAATAATTGACTCAGATCCAGTTTTACAGAATCTTCTCAGGGATAGTGATAACATCTATTTAGTAATCCGTCCATTAGAGGAGATACTTTTAATTGATCAATATACTAAAGGTGCTTTACACCATTGTCTTTTTCCTCTCCTAAATGTAGAACTTAACAAAAGACTCATAATATACTTGTTTTTAAAAGATTGATTGATGAAAGATCATAACTAATAACATTACAAATAATCCTACTATAATCAATACGGTGATCCAAATGTTAATCTTTCTCATTGCAAATACTCTTTGCCCTTATCCTCAAATTGCCTACATGCTTACATCTGAGGATAGCCAGTGTGACTTGGATTGGAGATGTGGAGAAAAAATCGGGACCCATTTCTAGGTTGTTCACCATCCAAGTACAGACATTGCCCTTCTAATTAAGAAAAAATCGGCGATGAAGATTAAGCCGACAGTGAGCGTAATCTTCATCTCTCTTAGATTATTTGTCCTCCAGAGTAGGGATCGTCAGGTCCTTTTCAATCGTATAACCAAAATAAACTTCACTAGAAGGATATTGTGGGGCAACAACACAATGGGTGTTACAGGAATATTGCAGTTACCTCGTGATCGATTCAAGAGGACATCATTCTTTCTTTGGGTAATTATCCTTTTCCAAAGAACATTTTCCATCCCACTTGGAGTCATCCACAATAGCACATTACAGGTTAGTGATGTCGACAAACTGGTTTGCCGTGACAAACTGTCATCCACAAATCAATTGAGATCAGTTGGACTGAATCTCGAAGGGAATGGAGTGGCAACTGACGTGCCATCTGCAACTAAAAGATGGGGCTTCAGGTCCGGTGTCCCACCAAAGGTGGTCAATTATGAAGCTGGTGAATGGGCTGAAAACTGCTACAATCTTGAAATCAAAAAACCTGACGGGAGTGAGTGTCTACCAGCAGCGCCAGACGGGATTCGGGGCTTCCCCCGGTGCCGGTATGTGCACAAAGTATCAGGAACGGGACCGTGTGCCGGAGACTTTGCCTTCCACAAAGAGGGTGCTTTCTTCCTGTATGACCGACTTGCTTCCACAGTTATCTACCGAGGAACGACTTTCGCTGAAGGTGTCGTTGCATTTCTGATACTGCCCCAAGCTAAGAAGGACTTCTTCAGCTCACACCCCTTGAGAGAGCCGGTCAATGCAACGGAGGACCCGTCTAGTGGCTACTATTCTACCACAATTAGATATCAAGCTACCGGTTTTGGAACCAATGAGACAGAGTATTTGTTCGAGGTTGACAATTTGACCTACGTCCAACTTGAATCAAGATTCACACCACAGTTTCTGCTCCAGCTGAATGAGACAATATATACAAGTGGGAAAAGGAGCAATACCACGGGAAAACTAATTTGGAAGGTCAACCCCGAAATTGATACAACAATCGGGGAGTGGGCCTTCTGGGAAACTAAAAAAACCTCACTAGAAAAATTCGCAGTGAAGAGTTGTCTTTCACAGCTGTATCAAACAGAGCCAAAAACATCAGTGGTCAGAGTCCGGCGCGAACTTCTTCCGACCCAGGGACCAACACAACAACTGAAGACCACAAAATCATGGCTTCAGAAAATTCCTCTGCAATGGTTCAAGTGCACAGTCAAGGAAGGGAAGCTGCAGTGTCGCATCTGACAACCCTTGCCACAATCTCCACGAGTCCTCAACCCCCCACAACCAAACCAGGTCCGGACAACAGCACCCACAATACACCCGTGTATAAACTTGACATCTCTGAGGCAACTCAAGTTGAACAACATCACCGCAGAACAGACAACGACAGCACAGCCTCCGACACTCCCCCCGCCACGACCGCAGCCGGACCCCTAAAAGCAGAGAACACCAACACGAGCAAGGGTACCGACCTCCTGGACCCCGCCACCACAACAAGTCCCCAAAACCACAGCGAGACCGCTGGCAACAACAACACTCATCACCAAGATACCGGAGAAGAGAGTGCCAGCAGCGGGAAGCTAGGCTTAATTACCAATACTATTGCTGGAGTCGCAGGACTGATCACAGGCGGGAGGAGAGCTCGAAGAGAAGCAATTGTCAATGCTCAACCCAAATGCAACCCTAATTTACATTACTGGACTACTCAGGATGAAGGTGCTGCAATCGGACTGGCCTGGATACCATATTTCGGGCCAGCAGCCGAGGGAATTTACATAGAGGGGCTGATGCACAATCAAGATGGTTTAATCTGTGGGTTGAGACAGCTGGCCAACGAGACGACTCAAGCTCTTCAACTGTTCCTGAGAGCCACAACCGAGCTACGCACCTTTTCAATCCTCAACCGTAAGGCAATTGATTTCTTGCTGCAGCGATGGGGCGGCACATGCCACATTTTGGGACCGGACTGCTGTATCGAACCACATGATTGGACCAAGAACATAACAGACAAAATTGATCAGATTATTCATGATTTTGTTGATAAAACCCTTCCGGACCAGGGGGACAATGACAATTGGTGGACAGGATGGAGACAATGGATACCGGCAGGTATTGGAGTTACAGGCGTTATAATTGCAGTTATCGCTTTATTCTGTATATGCAAATTTGTCTTTTAGTTTTTCTTCAGATTGCTTCATGGCAAAGCTCAGCCTCAAATCAATGAAACCAGGATTTAATTATATGGATTACTTGAATCTAAGATTACTTGACAAATGATAATATAATACACTGGAGCTTTAAACATAGCCAATGTGATTCTAACTCTTTTAAACTCACAGTTAATCATAAATAAGGTTTGACATCAATCTAGTTATCTCTTTGAGAATGATAAACTTGATGAAGATTAAGAAAAAGGTAATCTTTCGATTATCTTTAATCTTCATCCTTGATTCTACAATCATGACAGTTGTCTTTAGTGACAAGGGAAAGAAGCCTTTTTATTAAGTTGTAATAATCAGATCTGCGAACCGGTAGAGTTTAGTTGCAACCTAACACACATAAAGCATTGATCAAAAAGTCAATAGAAATTTAAACAGTGAGTGGAGACAACTTTTAAATGGAAGCTTCATATGAGAGAGGACGCCCACGAGCTGCCAGACAGCATTCAAGGGATGGACACGACCACCATGTTCGAGCACGATCATCATCCAGAGAGAATTATCGAGGTGAGTACCGTCAATCAAGGAGCGCCTCACAAGTGCGCGTTCCTACTGTATTTCATAAGAAGAGAGTTGAACCATTAACAGTTCCTCCAGCACCTAAAGACATATGTCCGACCTTGAAAAAAGGATTTTTGTGTGACAGTAGTTTTTGCAAAAAAGATCACCAGTTGGAGAGTTTAACTGATAGGGAATTACTCCTACTAATCGCCCGTAAGACTTGTGGATCAGTAGAACAACAATTAAATATAACTGCACCCAAGGACTCGCGCTTAGCAAATCCAACGGCTGATGATTTCCAGCAAGAGGAAGGTCCAAAAATTACCTTGTTGACTCTGATCAAGACGGCAGAACACTGGGCGAGACAAGACATCAGAACCATAGAGGATTCAAAATTAAGAGCATTGTTGACTCTATGTGCTGTGATGACGAGGAAATTCTCAAAATCCCAGCTGAGTCTTTTATGTGAGACACACCTAAGGCGTGAGGGGCTTGGGCAAGATCAGGCAGAACCTGTTCTCGAAGTATATCAACGATTACACAGTGATAAAGGAGGCAGTTTTGAAGCTGCACTATGGCAACAATGGGACCGACAATCCCTAATTATGTTTATCACTGCATTCTTGAATATCGCTCTCCAGTTACCGTGTGAAAGTTCTGCTGTCGTTGTTTCAGGGTTAAGAACATTGGTTCCTCAATCAGATAATGAGGAAGCTTCAACCAACCCGGGGACATGCTCATGGTCTGATGAGGGTACCCCTTAATAAGGCTGACTAAAACACTATATAACCTTCTACTTGATCACAATACTCCGTATACCTATTATCATATATTTAATCAAGACGATATCCTTTAAGACTTATTCAGTACTATAATCACTCTCGTTTCAAATTAATAAGATGTGCATGATTGCCCTAATATATGAAGAGGTATGATACAACCCTAACAGTGACCAAAGAAAATCATAATCTCGTATCGCTCGTAATATAACCTGCCAAGCATACCTCTTGCACAAAGTGATTCTTGTACACAAATAATGTTTTACCCTACAGGAGGTAGCAACGATCCATCCCATCAAAAAATAAGTATTTTATGACTTACTAATGACCTCTTAAAATATTAAGAAAAACTGACGGAACATAAATTCTTTCTGCTTCAAGTTGTGGAGGAGGTGTTTGGTATTGGCTATTGTTATATTACAATCAATAACAAGCTTGTAAAAATATTGTTCTTGTTTCAAGAGGTAGATTGTGGCTGGAAATGCTAAACTAATGATGAAGATTAATGCGGAGGTCTGATAAGAATAAACCTTATTATTCAGATTAGGCCCCAAGAGGCATTCTTCATCTCCTTTTAGCAAAGTACTATTTCAGGGTAGTCCAATTAGTGACACGTCTCTTAGCTGTATATCAGTCGCCCCTGAGATACGCCACAAAAGTGTCTCTAAGCTAAATTGGTCTGTACACATCTCATACATTGTATTAGGGACAATAATATCTAATTGAACTTAGCCGTTTAAAATTTAGTGCATAAATCTGGGCTAACTCCACCAGGTCAACTCCATTGGCTGAAAAGAAGCCTACCTACAACGAACATCACTTTGAGCGCCCTCACATTTAAAAAATAGGAACGTCGTTCCAACAATCGAGCGCAAGGTTTCAAGGTTGAACTGAGAGTGTCTAGACAACAAAGTATCGATCCTCCAGACACCAAGCAAGACCTGAGAAAAAACCATGGCTAAAGCTACGGGACGATACAATCTAATATCGCCCAAAAAGGACCTGGAGAAAGGGGTTGTCTTAAGCGACCTCTGTAACTTCTTAGTTAGCCAAACTATTCAGGGGTGGAAGGTTTATTGGGCTGGTATTGAGTTTGATGTGACTCACAAAGGAATGGCCCTATTGCAAAGACTGAAAACTAATGACTTTGCCCCTGCATGGTCAATGACAAGGAATCTCTTTCCTCATTTATTTCAAAATCCGAATTCCACAATTGAATCACCGCTGTGGGCATTGAGAGTCATCCTTGCAGCAGGGATACAAGACCAGCTGATTGACCAGTCTTTGATTGAACCCTTAGCAGGAGCCCTTGGTCTGATCTCTGATTGGCTGCTAACAACCAACACTAACCATTTCAACATGCGAACACAACGTGTCAAGGAACAATTGAGCCTAAAAATGCTGTCGTTGATTCGATCCAATATTCTCAAGTTTATTAACAAATTGGATGCTCTACATGTCGTGAACTACAACGGATTGTTGAGCAGTATTGAAATTGGAACTCAAAATCATACAATCATCATAACTCGAACTAACATGGGTTTTCTGGTGGAGCTCCAAGAACCCGACAAATCGGCAATGAACCGCAAGAAGCCTGGGCCGGCGAAATTTTCCCTCCTTCATGAGTCCACACTGAAAGCATTTACACAAGGATCCTCAACACGAATGCAAAGTTTGATTCTTGAATTTAATAGCTCTCTTGCTATCTAACTAAGATGGAATACTTCATATTGAGCTAACTCATATATGCTGACTCAATAGTTATCTTGACATCTCTGCTTTCATAATCAGATATATAAGCATAATAAATAAATACGCATATTTCTTGATAATTTGTTTAACCACAGATAAATCCTCACTGTAAGCCAGCTTCCAAGTTGACACCCTTACAAAAACCAGGACTCAGAATCCCTCAAATAAGAGATTCCAAGACAACATCATAGAATTGCTTTATCATATGAATAAGCATTTTATCACCAGAAATCTTATATACTAAATGGTTAATTGTAACTGAACCCGCAGGTCACATGTGTTAGGTTTCACAGATTTTATATATTACTAACTCTATACTCGTAATTAACATTAGATAAGTAGATTAAGAAAAAAGCCTGAGGAAGATTAAGAAAAACTGCTTATTGGGTCTTTCCGTGTTTTAGATGAAGCAGTTGAAATTCTTCCTCTTGATATTAAATGGCTACACAACATACCCAATACCCGGACGCTAGGTTATCATCACCAATTGTATTGGACCAATGTGACCTAGTCACTAGAGCTTGCGGGTTATATTCATCATACTCCCTTAATCCGCAACTACGCAACTGTAAACTCCCGAAACATATCTACCGTTTGAAATACGATGTAACTGTTACCAAGTTCTTGAGTGATGTACCAGTGGCGACATTGCCCATAGATTTCATAGTCCCAATTCTTCTCAAGGCACTGTCAGGCAATGGGTTCTGTCCTGTTGAGCCGCGGTGCCAACAGTTCTTAGATGAAATCATTAAGTACACAATGCAAGATGCTCTCTTCTTGAAATATTATCTCAAAAATGTGGGTGCTCAAGAAGACTGTGTTGATGACCACTTTCAAGAGAAAATCTTATCTTCAATTCAGGGCAATGAATTTTTACATCAAATGTTTTTCTGGTATGATCTGGCTATTTTAACTCGAAGGGGTAGATTAAATCGAGGAAACTCTAGATCAACATGGTTTGTTCATGATGATTTAATAGACATCTTAGGTTATGGGGACTATGTTTTTTGGAAGATCCCAATTTCAATGTTACCACTGAACACACAAGGAATCCCCCATGCTGCTATGGACTGGTATCAGGCATCAGTATTCAAAGAAGCGGTTCAAGGGCATACACACATTGTTTCTGTTTCTACTGCCGACGTCTTGATAATGTGCAAAGATTTAATTACATGTCGATTCAACACAACTCTAATCTCAAAAATAGCAGAGATTGAGGATCCAGTTTGTTCTGATTATCCCAATTTTAAGATTGTGTCTATGCTTTACCAGAGCGGAGATTACTTACTCTCCATATTAGGGTCTGATGGGTATAAAATTATTAAGTTCCTCGAACCATTGTGCTTGGCCAAAATTCAATTATGCTCAAAGTACACCGAGAGGAAGGGCCGATTCTTAACACAAATGCATTTAGCTGTAAATCACACCCTAGAAGAAATTACAGAAATGCGTGCACTAAAGCCTTCACAGGCTCAAAAGATCCGTGAATTCCATAGAACATTGATAAGGCTGGAGATGACGCCACAACAGCTTTGTGAGCTATTTTCCATTCAAAAACACTGGGGGCATCCTGTGCTACATAGTGAAACAGCAATCCAAAAAGTTAAAAAACATGCTACGGTGCTAAAAGCATTACGCCCTATAGTGATTTTCGAGACATACTGTGTTTTTAAATATAGTATTGCCAAACATTATTTTGATAGTCAAGGGTCTTGGTACAGTGTTACTTCAGATAGGAATCTAACACCAGGTCTTAATTCTTATATCAAAAGAAATCAATTCCCTCCGTTGCCAATGATTAAAGAACTACTATGGGAATTTTACCACCTTGACCATCCTCCACTTTTCTCAACCAAAATTATTAGTGACTTAAGTATTTTTATAAAAGACAGAGCTACTGCAGTAGAAAGGACATGCTGGGATGCAGTATTCGAGCCTAATGTTCTAGGATATAATCCACCTCACAAATTTAGTACTAAACGTGTACCGGAACAATTTTTAGAGCAAGAAAACTTTTCTATTGAGAATGTTCTTTCCTACGCGCAAAAACTCGAGTATCTACTACCACAATATCGGAACTTTTCTTTCTCATTGAAAGAGAAAGAGTTGAATGTAGGTAGAACCTTCGGAAAATTGCCTTATCCGACTCGCAATGTTCAAACACTTTGTGAAGCTCTGTTAGCTGATGGTCTTGCTAAAGCATTTCCTAGCAATATGATGGTAGTTACGGAACGTGAGCAAAAAGAAAGCTTATTGCATCAAGCATCATGGCACCACACAAGTGATGATTTTGGTGAACATGCCACAGTTAGAGGGAGTAGCTTTGTAACTGATTTAGAGAAATACAATCTTGCATTTAGATATGAGTTTACAGCACCTTTTATAGAATATTGCAACCGTTGCTATGGTGTTAAGAATGTTTTTAATTGGATGCATTATACAATCCCACAGTGTTATATGCATGTCAGTGATTATTATAATCCACCACATAACCTCACACTGGAAAATCGAGACAACCCCCCCGAAGGGCCTAGTTCATACAGGGGTCATATGGGAGGGATTGAAGGACTGCAACAAAAACTCTGGACAAGTATTTCATGTGCTCAAATTTCTTTAGTTGAAATTAAGACTGGTTTTAAGTTACGCTCAGCTGTGATGGGTGACAATCAGTGCATTACCGTTTTATCAGTCTTCCCCTTAGAGACTGACGCAGACGAGCAGGAACAGAGCGCCGAAGACAATGCAGCGAGGGTGGCCGCTAGCCTAGCAAAAGTTACAAGTGCCTGTGGAATCTTTTTAAAACCTGATGAAACATTTGTACATTCAGGTTTTATCTATTTTGGAAAAAAACAATATTTGAATGGGGTCCAATTGCCTCAGTCCCTTAAAACGGCTACAAGAATGGCACCATTGTCTGATGCAATTTTTGATGATCTTCAAGGGACCCTAGCTAGTATAGGCACTGCTTTTGAGCGATCCATCTCTGAGACACGACATATCTTTCCTTGCAGGATAACCGCAGCTTTCCATACGTTTTTTTCGGTGAGAATCTTGCAATATCATCATCTCGGGTTCAATAAAGGTTTTGACCTTGGACAGTTAACACTCGGTAAACCTCTGGATTTCGGAACAATATCATTGGCACTAGCGGTACCGCAGGTGCTTGGAGGATTATCCTTCTTGAATCCTGAGAAATGTTTCTACCGGAATCTAGGAGATCCAGTTACCTCAGGTTTATTCCAGTTAAAAACTTATCTCCGAATGATTGAGATGGATGATTTATTCTTACCTTTAATTGCAAAGAACCCTGGGAACTGCACTGCCATTGACTTTGTGCTAAATCCTAGCGGATTAAATGTCCCTGGGTCGCAAGACTTAACTTCATTTCTGCGCCAGATTGTACGCAGGACCATCACCCTAAGTGCGAAAAACAAACTTATTAATACCTTATTTCATGCGTCAGCTGACTTCGAAGACGAAATGGTTTGTAAATGGCTATTATCATCAACTCCTGTTATGAGTCGTTTTGCGGCCGATATCTTTTCACGCACGCCGAGCGGGAAGCGATTGCAAATTCTAGGATACCTGGAAGGAACACGCACATTATTAGCCTCTAGGATCATCAACAATAATACAGAGACACCGGTTTTGGACAGACTGAGGAAAATAACATTGCAAAGGTGGAGTCTGTGGTTTAGTTATCTTGATCATTGTGATAATATCCTGGCGGAGGCTTTAACCCAAATAACTTGCACAGTTGATTTAGCACAGATCCTGAGGGAATATTCATGGGCTCATATTTTAGAGGGAAGACCTCTTATTGGAGCCACACTCCCATGTATGATTGAGCAATTCAAAGTGTTTTGGCTGAAACCCTACGAACAATGTCCGCAGTGTTCAAATGCAAAGCAACCAGGTGGGAAACCATTCGTGTCAGTGGCAGTCAAGAAACATATTGTTAGTGCATGGCCTAACGCATCCCGACTAAGCTGGACTATCGGGGATGGAATCCCTTACATTGGATCAAGGACAGAAGATAAGATAGGACAACCTGCTATTAAACCAAAATGTCCTTCCGCAGCCTTAAGAGAGGCCATTGAATTGGCGTCCCGTTTAACATGGGTAACTCAAGGCAGTTCGAACAGTGACTTGCTAATAAAGCCATTTTTGGAAGCACGAGTAAATTTAAGTGTTCAAGAAATACTTCAAATGACCCCTTCACATTACTCAGGAAATATTGTTCACAGGTACAACGATCAATACAGTCCTCATTCTTTCATGGCCAATCGTATGAGTAATTCAGCAACGCGATTGATTGTTTCTACAAACACTTTAGGTGAGTTTTCAGGAGGTGGCCAGTCTGCACGCGACAGCAATATTATTTTCCAGAATGTTATAAATTATGCAGTTGCACTGTTCGATATTAAATTTAGAAACACTGAGGCTACAGATATCCAATATAATCGTGCTCACCTTCATCTAACTAAGTGTTGCACCCGGGAAGTACCAGCTCAGTATTTAACATACACATCTACATTGGATTTAGATTTAACAAGATACCGAGAAAACGAATTGATTTATGACAATAATCCTCTAAAAGGAGGACTCAATTGCAATATCTCATTCGATAACCCATTTTTCCAAGGTAAACAGCTGAACATTATAGAAGATGATCTTATTCGACTGCCTCACTTATCTGGATGGGAGCTAGCCAAGACCATCATGCAATCAATTATTTCAGATAGCAACAATTCATCTACAGACCCAATTAGCAGTGGAGAAACAAGATCATTCACTACCCATTTCTTAACTTATCCCAAGATAGGACTTCTGTACAGTTTTGGGGCCTTTGTAAGTTATTATCTTGGCAATACAATTCTTCGGACTAAGAAATTGACACTTGACAATTTTTTATATTACTTAACTACCCAAATTCATAATCTACCACATCGCTCATTGCGAATACTTAAGCCAACATTCAAACATGCAAGCGTTATGTCACGGTTAATGAGTATTGATCCTCATTTTTCTATTTACATAGGCGGTGCTGCAGGTGACAGAGGACTCTCAGATGCGGCCAGGTTATTTTTGAGAACGTCTATTTCATCTTTTCTTACATTTGTAAAAGAATGGATAATTAATCGCGGAACAATTGTCCCTTTATGGATAGTATATCCGCTAGAGGGTCAAAACCCAACACCTGTTAATAATTTTCTCTATCAGATCGTAGAACTGCTGGTGCATGATTCATCAAGACAACAGGCTTTAAAAACTACCATAAGTGATCATGTACATCCTCACGACAATCTTGTTTACACATGTAAGAGTACAGCCAGCAATTTCTTCCATGCATCATTGGCGTACTGGAGGAGCAGGCACAGAAACAGCAACCGAAAATACTTGGCAAGAGACTCTTCAACTAGATCAAGCACAAACAACAGTGATGGTCATATTGAGAGAAGTCAAGAACAAACCACCAGAGATCCACATGATGGCACTGAACGGAATCTAGTTCTACAAATGAGCCATGAAATAAAAAGAACGACAATTCCACAAGAAAACACGCACCAGGGTCCGTCGTTCCAGTCCTTTCTAAGTTACTCTGCTTGTGGTACAGCAAATCCAAAACTAAATTTCGATCGATCGAGACACAATGTGAAATCTCAGGATCATAACTCGGCATCCAAGAGGGAAGGTCATCAAATAATCTCACACCGTCTAGTCCTACCTTTCTTTACATTATCTCAAGGGACACGCCAATTAACGTCATCTAATGAGTCACAGACCCAAGACGAGATATCAAAGTACTTACGGCAATTGAGATCCGTCATTGATACCACAGTTTATTGTAGGTTTACCGGTATAGTCTCGTCCATGCATTACAAACTTGATGAGGTCCTTTGGGAAATAGAGAGTTTTAAGTCGGCTGTGACGCTAGCAGAGGGAGAAGGTGCTGGTGCCTTACTATTGATTCAGAAATACCAAGTTAAGACCTTATTTTTCAACACGCTAGCTACTGAGTCCAGTATAGAGTCAGAAATAGTATCAGGAATGACTACTCCTAGGATGCTTCTACCTGTTATGTCAAAATTCCATAATGACCAAATTGAGATTATTCTTAACAACTCAGCAAGCCAAATAACAGACATAACAAATCCTACTTGGTTTAAAGACCAAAGAGCAAGGCTACCTAAGCAAGTCGAGGTTATAACCATGGATGCAGAGACGACAGAGAATATAAACAGATCGAAATTGTACGAAGCTGTATATAAACTGATCTTACACCATATTGATCCCAGCGTATTGAAAGCAGTGGTCCTTAAAGTCTTCCTAAGTGATACTGAGGGTATGTTATGGCTAAATGATAATTTAGCCCCGTTTTTTGCCACTGGTTATTTAATTAAGCCAATAACGTCAAGTGCTAGATCTAGTGAGTGGTATCTTTGTCTGACGAACTTCTTATCAACTACACGAAAGATGCCACACCAAAACCATCTCAGTTGTAAACAGGTAATACTTACGGCATTGCAACTGCAAATTCAACGGAGCCCATACTGGCTAAGTCATTTAACTCAGTATGCTGACTGTGATTTACATTTAAGTTATATCCGCCTTGGTTTTCCATCATTAGAGAAAGTACTATACCACAGGTATAACCTCGTCGATTCAAAAAGAGGTCCACTAGTCTCTATCACTCAGCACTTAGCACATCTTAGAGCAGAGATTCGAGAATTAACTAATGATTATAATCAACAGCGACAAAGTCGAACTCAAACATATCACTTTATTCGTACTGCAAAAGGACGAATCACAAAACTAGTCAATGATTATTTAAAATTCTTTCTTATTGTGCAAGCATTAAAACATAATGGGACATGGCAAGCTGAGTTTAAGAAATTACCAGAGTTGATTAGTGTGTGCAATAGGTTCTACCATATTAGAGATTGCAATTGTGAAGAACGTTTCTTAGTTCAAACCTTATATTTACATAGAATGCAGGATTCTGAAGTTAAGCTTATCGAAAGGCTGACAGGGCTTCTGAGTTTATTTCCGGATGGTCTCTACAGGTTTGATTGAATTACCGTGCATAGTATCCTGATACTTGCAAAGGTTGGTTATCAACATACAGATTATAAAAAACTCATAAATTGCTCTCATACATCATATTGATCTGATTTCAATAAACAACTATTTAAATAACGAAAGGAGTCCCTATATTATACACTATATTTAGCCTCTCTCCCTGCGTGATAATCAAAAAATTCACAATGCAGCATGTGTGACATATTAATGCCGCAATGAATTTAACGCAACATAATAAACTCTGCACTCTTTATAATTAAGCTTTAACGAAAGGTCTGGGCTCATATTGTTATTGATATAATAATGTTATATCAATGTCCTGTCAGATGGAATAGTGTTTTGGTTGATAACACGACTTCTTAAAACAAAATTGATCTTTAAGATTAAGTTTTTTATAATTATCATTACTTTAATTTATCGATTTGAAAATGGTAATAGCCTTAATCTTTGTGTAAAATAAGAGATTAGGTGTAATAACTTTAACATTTTTGTCTAGTAAGTTACTATTTCATACAGAATGATAAAATTAAAAGAAAAGGCAGGACTGTAAAATCATAAATACCTTCTTTACAATATAGCAGACTAGATAATAATCTTTGTGTTAATGATAATTAAGACATTGACCACGCTCATCAGAAGGCTCGCCAAAATAAACGTTGCAAAAAGGATTCCTGGAAAAATGGTCGCACACAAAAATTTAAAAATAAATCTATTTCTTCTTTTTTGTGTGTCCA\n"
     ]
    }
   ],
   "source": [
    "# working with DB\n",
    "\n",
    "import biotite.database.entrez as entrez\n",
    "\n",
    "dir(entrez)\n",
    "\n",
    "# get our database name\n",
    "entrez.get_database_name(\"Nucleotide\")\n",
    "entrez.get_database_name(\"Protein\")\n",
    "\n",
    "# search\n",
    "# database => query => search => ID => fetch\n",
    "\n",
    "# making a simple query\n",
    "myquery = entrez.SimpleQuery(\"Covid19\")\n",
    "\n",
    "# searching with query\n",
    "ans_id = entrez.search(myquery, \"nucleotide\", number=10)\n",
    "print(ans_id)\n",
    "\n",
    "# downloading our file\n",
    "file_path = entrez.fetch(\"2105187405\", \"myresult_dir\", suffix ='fa', db_name =\"nuccore\", ret_type=\"fasta\")\n",
    "\n",
    "file_path\n",
    "\n",
    "# Reading and downloading \n",
    "# id KU182909.1\n",
    "\n",
    "file_path2 = entrez.fetch(\"KU182909.1\", \"myresult_dir\", suffix ='fa', db_name =\"nuccore\", ret_type=\"fasta\")\n",
    "\n",
    "import biotite.sequence.io.fasta as fasta\n",
    "\n",
    "file2 = fasta.FastaFile()\n",
    "file2.read(file_path2)\n",
    "\n",
    "for h, st in file2.items():\n",
    "\tprint(h)\n",
    "\tprint(st)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "mypdb_dir\\4ZS6.pdb\n"
     ]
    }
   ],
   "source": [
    "# working with 3D structures\n",
    "\n",
    "# fetching our 3D structure file\n",
    "\n",
    "import biotite.database.rcsb as rcsb \n",
    "\n",
    "dir(rcsb)\n",
    "\n",
    "pdb_file_path = rcsb.fetch(\"4ZS6\", \"pdb\", \"mypdb_dir\")\n",
    "\n",
    "print(pdb_file_path)\n",
    "\n",
    "# multiple files\n",
    "pdb_file_path = rcsb.fetch([\"4ZS6\", \"6LU7\"], \"pdb\", \"mypdb_dir\")\n",
    "\n",
    "\n",
    "# reading 3D pdb files\n",
    "# pdb, cif, npz, xtc, mmt\n",
    "import biotite.structure.io.pdb as pdb\n",
    "\n",
    "dir(pdb)\n",
    "\n",
    "file_reader = pdb.PDBFile()\n",
    "file_reader.read(\"mypdb_dir/4ZS6.pdb\")\n",
    "# print(file_reader)\n",
    "\n",
    "protein_struc = file_reader.get_structure()\n",
    "# print(protein_struc)\n",
    "\n",
    "\n",
    "# shape and coord\n",
    "# print(\"Shape\", protein_struc.shape)\n",
    "# print(\"Coordinates\", protein_struc.coord)\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Tree: ((Apple:2.0,Pear:2.0):2.0,((Orange:1.0,Lemon:1.0):2.0,Banana:3.0):1.0):0.0;\n",
      "Tree w/o distances: ((Apple,Pear),((Orange,Lemon),Banana));\n",
      "Distance Apple-Banana: 8.0\n"
     ]
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAagAAAGoCAYAAAATsnHAAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjQuMywgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy/MnkTPAAAACXBIWXMAAAsTAAALEwEAmpwYAAAUIUlEQVR4nO3dfbBtB1nf8d9DEt5yEahkkOEt1EAZSSUvgBBauNDBojAESxyhFEiLWHVKbR3HQbSlpeIUZ3RqodYXwKggYpGSQAVKC7EwFQKBG0N4jYS20EhSFMKF8Hbz9I+9rhwO9+Xce0/Ofi7n85k5c9ZZe+21n7Nyz/6etfaenOruAMA0t1n3AABwKAIFwEgCBcBIAgXASAIFwEgCBcBIAgXASAIFwEgCBcBIAgXASAIFwEgCBcBIAgXASAIFwEgCBcBIAgXASAIFwEgCBcBIAgXASAIFwEgCBcBIAgXASAIFwEgCBcBIAgXASAIFwEgCBcBIAgXASAIFwEgCBcBIAgXASAIFwEgCBcBIAgXASAIFwEgCBcBIAgXASAIFwEgCBcBIAgXASAIFwEgCBcBIAgXASAIFwEgCBcBIAgXASAIFwEgCBcBIAgXASAIFwEgCBcBIAgXASAIFwEgCBcBIAgXASAIFwEgCBcBIAgXASAIFwEgCBcBIAgXASAIFwEgCBcBIAgXASAIFwEgCBcBIAgXASAIFwEgCBbBLVNWBqtpXVVdV1fuq6oJ1z3Qk1d3rngGAHVBV+7t7z7L8d5M8v7sfveaxDuvUdQ/A4VXV+5OckeTadc8CzNfde49h829L8pdJUlV7klya5K5JTkvyc919aVWdmeRNSd6Z5IIkn0pyYXffXFXPSfIjSW6b1XPUM7r7i1V1SZKbkjwkyXck+enufu3hHuNIAzqDGqyqPplkT5J9ax4FOAkcLVBVdSDJ1Ulun+QeSR7b3VdW1alJ7tjdN1XV3ZK8K8n9k9w3q/g8pLv3VdUfJLmsu19ZVd/e3Z9Z9vvzST7d3S9ZAnV6kh9K8sBl+7MO9xh9hAg5g5rt2uSYfysCOJybu/ucJKmqRyT5nao6O0kl+YWqelSSW5LcM8ndl/tc1937luUrk5y5LJ+9hOkuWf0i/ZYNj/P67r4lyQer6uB+DvcYf364YQUKYBfq7j9ZzmTOSPL9y+fzu/urVfWJrM6ykuTLG+52IMkdluVLkjy5u6+qqouT7N2w3cb71PL56Ud4jEPyLj6AXaiqHpjklCSfSXLnJDcs4XhMVpf2juZOSa6vqtOyis/RHPNjOIMC2D3uUFX7luVK8qzuPlBVr0ryhqq6Osl7k3x4C/v6F0neneTG5fOdjrL9MT+GN0kMVlWXJ16DAnYnl/gAGEmgABhJoAAYSaAAGEmgABhJoAAYSaAAGEmgABhJoAAYSaAAGEmgABhJoAAYSaAAGEmgABhJoAAYSaAAGEmgABhJoAAYSaAAGEmgABhJoAAYSaAAGEmgABhJoAAYSaAAGEmgABhJoAAYSaAAGEmgABhJoAAYSaAAGEmgABhJoAAYSaAAGEmgABhJoAAYSaAAGEmgABhJoAAYSaAAGEmgABhJoAAY6dTjvWNVvT/JGUmu3b5x2OScJPvXPQTAOpzIGdQZSfZs1yAAsNFxn0FlOXPq7r3bMwqbVdXl654BYF28BgXASAIFwEgCBcBIAgXASAIFwEgCBcBIAgXASAIFwEgCBcBIAgXASAIFwEgCBbBLVNVJ9dcRBAqAkQQKYBerqu+sqjdX1ZVV9Y6qeuCy/pKq+o9V9a6q+nhV7a2qV1TVh6rqkg33f1pVXV1VH6iqF29Yv7+qXlRVVy37uPuxznYif24DYEv8gdOdcZx//ug3kvxod3+sqr4nya8meexy212TPCLJk5JcluSRSX44yXuq6pwkNyR5cZLzk/xlkv9aVU/u7tcnOT3Ju7r7Z6vqF5M8J8nPH8tgAgXsBH/gdKCq2pPkgiT/qaoOrr7dhk3e0N1dVVcn+XR3X73c75okZya5b5LLu/vGZf2rkjwqyeuTfCXJG5f9XJnkccc6n0ABO8EfOJ3pNkk+293nHOb2Ly+fb9mwfPDrU5N89Qj7/mp397J8IMfRG69BAexS3X1Tkuuq6geTpFYefAy7uCLJo6vqblV1SpKnJfnj7ZpPoAB2jztW1Sc3fPxkkqcneXZVXZXkmiQXbnVn3X19kucleXuSq5Jc2d2Xbtew9fUzsGO8Y9Xly4B7t2sYvpFjzLcK/5Y5Hs6gABhJoAAYSaAAGEmgABhJoAAYSaAAGEmgABhJoAAYSaAAGEmgABhJoAAYSaAAGEmgABhJoAAYSaAAGEmgABhJoAAYSaAAGEmgABhJoAAYSaAAGEmgABhJoAAYSaAAGEmgABhJoAAY6dR1D8ARnZVkT1Vdvu5B4ASdk2T/uofg5CJQs9247gEA1kWgBuvuc9c9A2wHVwE4Hl6DAmAkgQJgJIECYCSBAmAkgQJgJIECYCSBAmAkgQJgJIECYCSBAmAkgQJgJIECYCSBAmAkgQJgJIECYCSBAmAkgQJgJIECYCSBAmAkgQJgJIECYCSBAmAkgQJgJIECYCSBAmAkgQJgJIECYCSBAmAkgQLYRarqXlV1aVV9rKr+rKp+papuu+65DkWgAHaJqqokr0vy+u6+f5IHJNmT5EWbtjt1DeN9kxFDAN/yzkqyp6ouX/cg38q6e+9RNnlski91928t2x+oqn+e5Lqqui7J47MK1ilV9YQklya5a5LTkvxcd19aVWcmeVOSdya5IMmnklzY3TdX1UOTvDzJLUnemuT7uvvsqjolyb9NsjfJ7ZL8h+7+9aN9P86ggJ1wY5L96x6CPCjJlRtXdPdNSf53Vics5yW5qLsfneRLSX6gu89L8pgkv7ScgSXJ/bOKzIOSfDbJU5b1v5XkH3f3OUkObHiYZyf5XHc/NMlDkzynqu53tGGdQQG3uu4+d90zsCVv7e6/WJYryS9U1aOyOiO6Z5K7L7dd1937luUrk5xZVXdJcqfu/pNl/e8leeKy/L1JvruqLlq+vnNWkbvuSMMIFMDu8cEkF21cUVXfluQ+Sb6W5Asbbnp6kjOSnN/dX62qTyS5/XLblzdsdyDJHY7yuJXkud39lmMZ1iU+gN3jvye5Y1U9M0mW14Z+KcklSb64ads7J7lhidNjktz3SDvu7s8m+XxVfc+y6qkbbn5Lkh+rqtOWx31AVZ1+tGEFCmCX6O5O8gNJfrCqPpbko1m91vT8Q2z+qiQPqaqrkzwzyYe38BDPTvKbVbUvyelJPresf1lWZ2/vq6oPJPn1bOEKXq3mPXYH342zhXeNALALVNWe7t6/LD8vyT26+yeOd39egwJguzyhqn4mq7b8ryQXn8jOnEEBMJLXoAAYSaAAGEmgABhJoAAYSaAAGEmgABhJoAAYSaAAGEmgABhJoAAYSaAAGEmgABhJoAAYSaAAGEmgABhJoAAYSaAAGEmgABhJoAAYSaAAGEmgABhJoAAYSaAAGEmgABhJoAAYSaAAGEmgABhJoAAYSaAAGEmgABhJoAAYSaAAGEmgABhJoAAYSaAAGEmgABhJoAAYSaAAGEmgABhJoAAYSaAAGEmgABhJoAAYSaAAGEmgABhJoAAYSaAAGEmgABhJoAAYSaAAGEmgABhJoAAYSaAAGEmgABjp1HUPAMDOqKoDSa7O6rn/Q0me1d1fXO9Uh+cMCmD3uLm7z+nus5N8JcmPnsjOqupWPckRKIDd6R1Jzqqq06vqFVV1RVW9v6ouTJKqOrOq3lFV71s+LljW713WX5bkg7fmgC7xsetV1fuTnJHk2nXPAieiu/duZbvlzOf7krw5yc8meVt3/6OqukuSK6rqvyW5IcnjuvtLVXX/JK9O8pBlF+clObu7r9vmb+EbCBSs4rRn3UPADrhDVe1blt+R5OVJ/meSJ1XVTy3rb5/kPkn+b5KXVtU5SQ4kecCG/Vxxa8cpEShIljOnrf72CSexm7v7nI0rqqqSPKW7P7Jp/b9K8ukkD87q5aAvbbj5C7fumCtegwLY3d6S5LlLqFJV5y7r75zk+u6+Jckzkpyy04MJFMDu9m+SnJbkT6vqmuXrJPnVJM+qqquSPDA7dNa0UXX38d2x6vLEZRFOfv4tw0zOoAAYSaAAGEmgABhJoAAYSaAAGEmgABhJoAAYSaAAGEmgABhJoAAYSaAAGEmgABhJoAAYSaAAGEmgABhJoAAYSaAAGEmgABhJoAAYSaAAGEmgABhJoAAYSaAAGEmgABhJoAAYSaAAGOnUE7jvWUn2VNXl2zQLrMs5SfavewjgG51IoG7ctikAYJPjDlR3n7udg8C6uAoAM3kNCoCRBAqAkQQKgJEECoCRBAqAkQQKgJEECoCRBAqAkQQKgJEECoCRBAqAkQQKgJEECoCRBAqAkQQKgJEECoCRBAqAkQQKgJEECoCRBAqAkQQKgJEECoCRBAqAkQQKgJEECoCRBAqAkQQKgJEECoCRBApgF6mqJ1dVV9UDT2Afl1TVRds516EIFMDu8rQk71w+j3bqugeAAc5KsqeqLl/3IHAiunvvkW6vqj1J/laSxyR5Q5IXVNXeJC9M8vmsfhbenuTHu/uWqtqf5DeTfG+SP0/y1O6+cdM+z0/yy0n2JPl/SS7u7uu34/txBgXJjUn2r3sI2AEXJnlzd380yWeWuCTJw5I8N8l3JfnOJH9vWX96kvd294OS/HGSF2zcWVWdluQlSS7q7vOTvCLJi7ZrWGdQ7Hrdfe66Z4Ad8rQkv7Is//7y9RuTXNHdH0+Sqnp1VmdZr01yS5LXLNu/MsnrNu3vbyQ5O8lbqypJTkmyLWdPiUAB7ApV9deSPDbJ36yqziomneS/LJ832vz14dZXkmu6+xHbOetBLvEB7A4XJfnd7r5vd5/Z3fdOcl2Sv53kYVV1v6q6TZIfyupNFMmqEQffrff3N6w/6CNJzqiqRySrS35V9aDtGligAHaHpyX5z5vW/eGy/j1JXprkQ1lF6+B2X8gqXh/I6uzrhRvv3N1fySpgL66qq5LsS3LBdg1c3Yc7kwPgW93yLr6f6u4nHuK2/d29Z8eHWjiDAmAkZ1AAjOQMCoCRBAqAkQQKgJEECoCRBAqAkQQKgJEECoCRBAqAkQQKgJEECoCRBAqAkQQKgJEECoCRBAqAkQQKgJEECoCRBAqAkQQKgJEECoCRBAqAkQQKgJEECoCRBAqAkQQKgJEECoCRBAqAkQQKgJEECoCRBAqAkQQKgJEECoCRBAqAkQQKgJEECoCRBAqAkQQKgJEECoCRBAqAkQQKgJEECoCRBAqAkQQKgJEECoCRBAqAkQQKgJEECoCRBAqAkQQKgJEECoCRBAqAkQQKgJEECoCRBAqAkQQKgJEECoCRBAqAkQQKgJEECoCRTihQVfUj2zXITjnZZj7Z5k3MvBNOtnkTM++Ek23eoznRM6iT8WCcbDOfbPMmZt4JJ9u8iZl3wsk27xG5xAfASAIFwEgnGqjf2JYpdtbJNvPJNm9i5p1wss2bmHknnGzzHlF197pnAIBv4hIfACMJFAAjbSlQVfX4qvpIVV1bVc87xO23q6rXLLe/u6rO3PZJj8EW5r24qm6sqn3Lxw+vY85NM72iqm6oqg8c5vaqqn+/fE9/WlXn7fSMm+Y52rx7q+pzG47xv9zpGQ8x072r6u1V9cGquqaqfuIQ24w5zlucd9RxrqrbV9UVVXXVMvO/PsQ2Y54vtjjvuOeLJKmqU6rq/VX1xkPcNuYYn5DuPuJHklOS/FmSv57ktkmuSvJdm7b58SS/tiw/NclrjrbfW+tji/NenOSl65rxMHM/Ksl5ST5wmNu/P8mbklSShyd59/B59yZ547qP66aZ7pHkvGX5Tkk+eoh/G2OO8xbnHXWcl+O2Z1k+Lcm7kzx80zaTni+2Mu+454tlrp9M8nuH+u8/6RifyMdWzqAeluTa7v54d38lye8nuXDTNhcm+e1l+bVJ/k5V1Rb2fWvYyrzjdPf/SPIXR9jkwiS/0yvvSnKXqrrHzkz3zbYw7zjdfX13v29Z/nySDyW556bNxhznLc47ynLc9i9fnrZ8bH4n1pjniy3OO05V3SvJE5K87DCbjDnGJ2Irgbpnkv+z4etP5pt/SP5qm+7+WpLPJfn27RjwOGxl3iR5ynIJ57VVde+dGe2EbPX7muQRy6WTN1XVg9Y9zEbLJY9zs/qNeaORx/kI8ybDjvNy6WlfkhuSvLW7D3uMBzxfbGXeZN7zxb9L8tNJbjnM7aOO8fHarW+SeEOSM7v7u5O8NV//TYPt874k9+3uByd5SZLXr3ecr6uqPUn+MMk/6+6b1j3P0Rxl3nHHubsPdPc5Se6V5GFVdfaaRzqiLcw76vmiqp6Y5IbuvnKdc+yErQTqU0k2/sZwr2XdIbepqlOT3DnJZ7ZjwONw1Hm7+zPd/eXly5clOX+HZjsRW/nvMEZ333Tw0kl3/1GS06rqbmseK1V1WlZP9q/q7tcdYpNRx/lo8049zknS3Z9N8vYkj99006Tni79yuHkHPl88MsmTquoTWb2E8diqeuWmbUYe42O1lUC9J8n9q+p+VXXbrF5wu2zTNpcledayfFGSt/Xy6twaHHXeTa8pPCmra/vTXZbkmcu7zB6e5HPdff26hzqcqvqOg9e8q+phWf1bW+sPyDLPy5N8qLt/+TCbjTnOW5l32nGuqjOq6i7L8h2SPC7JhzdtNub5YivzTnu+6O6f6e57dfeZWT2/va27/8GmzcYc4xNx6tE26O6vVdU/SfKWrN4h94ruvqaqXpjkvd19WVY/RL9bVddm9cL5U2/Nobdh3n9aVU9K8rVl3ovXNe9BVfXqrN6Rdbeq+mSSF2T1gm26+9eS/FFW7zC7NskXk/zD9Uy6soV5L0ryY1X1tSQ3J3nqgB+QRyZ5RpKrl9cckuT5Se6TjDzOW5l32nG+R5LfrqpTsorlH3T3G6c+X2Rr8457vjiUwcf4uPlfHQEw0m59kwQAwwkUACMJFAAjCRQAIwkUACMJFAAjCRQAI/1/HlokKXljD/kAAAAASUVORK5CYII=",
      "text/plain": [
       "<Figure size 432x432 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Phylogenetics with biotite\n",
    "import biotite.sequence.phylo as phylo\n",
    "\n",
    "dir(phylo)\n",
    "\n",
    "\n",
    "# the reference objects\n",
    "fruits = [\"Apple\", \"Pear\", \"Orange\", \"Lemon\", \"Banana\"]\n",
    "# create nodes\n",
    "apple  = phylo.TreeNode(index = fruits.index(\"Apple\"))\n",
    "pear   = phylo.TreeNode(index = fruits.index(\"Pear\"))\n",
    "orange = phylo.TreeNode(index = fruits.index(\"Orange\"))\n",
    "lemon  = phylo.TreeNode(index = fruits.index(\"Lemon\"))\n",
    "banana = phylo.TreeNode(index = fruits.index(\"Banana\"))\n",
    "intermediate1 = phylo.TreeNode(\n",
    "\tchildren = (apple, pear), distances = (2.0, 2.0)\n",
    ")\n",
    "intermediate2 = phylo.TreeNode((orange, lemon), (1.0, 1.0))\n",
    "intermediate3 = phylo.TreeNode((intermediate2, banana), (2.0, 3.0))\n",
    "root = phylo.TreeNode((intermediate1, intermediate3), (2.0, 1.0))\n",
    "\n",
    "# create tree from root node\n",
    "tree = phylo.Tree(root = root)\n",
    "# trees can be converted into Newick notation\n",
    "print(\"Tree:\", tree.to_newick(labels = fruits))\n",
    "# distances can be omitted\n",
    "print(\n",
    "\t\"Tree w/o distances:\", \n",
    "\ttree.to_newick(labels = fruits, include_distance = False)\n",
    ")\n",
    "# distances can be measured\n",
    "distance = tree.get_distance(fruits.index(\"Apple\"), fruits.index(\"Banana\"))\n",
    "print(\"Distance Apple-Banana:\", distance)\n",
    "\n",
    "# plot a phylo tree\n",
    "import matplotlib.pyplot as plt\n",
    "import biotite.sequence.graphics as graphics\n",
    "\n",
    "fig, ax = plt.subplots(figsize = (6.0, 6.0))\n",
    "graphics.plot_dendrogram(ax, tree, labels=fruits)\n",
    "fig.tight_layout()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Score 101\n"
     ]
    },
    {
     "ename": "NameError",
     "evalue": "name 'best' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[1;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[1;32m~\\AppData\\Local\\Temp/ipykernel_4868/3175416242.py\u001b[0m in \u001b[0;36m<module>\u001b[1;34m\u001b[0m\n\u001b[0;32m     20\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m     21\u001b[0m \u001b[0mprint\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;34m\"Score\"\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mbest_alignments\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mscore\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m---> 22\u001b[1;33m \u001b[0mprint\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;34m\"E-value\"\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mbest\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0malignments\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0me_value\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m     23\u001b[0m \u001b[0mprint\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;34m\"HIT_ID\"\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mbest_alignments\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mhit_id\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m     24\u001b[0m \u001b[0mprint\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;34m\"HIT NAME\"\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mbest_alignments\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mhit_definition\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;31mNameError\u001b[0m: name 'best' is not defined"
     ]
    }
   ],
   "source": [
    "# simple app\n",
    "\n",
    "import biotite.application\n",
    "\n",
    "dir(biotite.application)\n",
    "\n",
    "# BLAST\n",
    "import biotite.application.blast as blast\n",
    "\n",
    "import biotite.sequence as seq\n",
    "mini_protein = seq.ProteinSequence(\"NLYIQWLKDGGPSSGRPPPS\")\n",
    "\n",
    "# initialise the app and created state\n",
    "app = blast.BlastWebApp(\"blastp\" , mini_protein) \n",
    "app.start() # task\n",
    "app.join() # make accessible\n",
    "\n",
    "alignments = app.get_alignments()\n",
    "best_alignments = alignments[0] # print best alignments\n",
    "\n",
    "print(\"Score\", best_alignments.score)\n",
    "print(\"E-value\", best.alignments.e_value)\n",
    "print(\"HIT_ID\", best_alignments.hit_id)\n",
    "print(\"HIT NAME\", best_alignments.hit_definition)\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# final project in biotite\n",
    "\n",
    "import biotite\n",
    "import biotite.sequence as seq\n",
    "\n",
    "dir(biotite)\n",
    "\n",
    "# create DNA sequence\n",
    "dna = seq.NucleotideSequence(\"ATCTAGGAC\")\n",
    "\n",
    "# check for the alphabet in the nucleotide sequence\n",
    "dna.get_alphabet()\n",
    "\n",
    "# nucleotide frequencies in the string\n",
    "dna_freq = dna.get_symbol_frequency()\n",
    "print(f\"Nucleotide frequencies: {dna_freq}\")\n",
    "\n",
    "# plot it\n",
    "import matplotlib.pyplot as plt\n",
    "import biotite.sequence.graphics as graphics\n",
    "plt.bar(dna_freq.keys(), dna_freq.values())\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "# complement\n",
    "print(f\"DNA complement: {dna.complement()}\")\n",
    "# translation\n",
    "protein = dna.translate(complete=True)\n",
    "print(protein)\n",
    "\n",
    "# creating protein sequence\n",
    "protein_seq = seq.ProteinSequence(\"MIT\")\n",
    "print(protein_seq)\n",
    "\n",
    "# protein: AA 1letter -> 3letter code\n",
    "for sym in protein_seq:\n",
    "\tprint(seq.ProteinSequence.convert_letter_1to3(sym))\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "# sequence alignments\n",
    "seq1 = seq.ProteinSequence(\"MITITET\")\n",
    "seq2 = seq.ProteinSequence(\"ITITET\")\n",
    "\n",
    "import biotite.sequence.align as align\n",
    "dir(align)\n",
    "\n",
    "# sub matrix\n",
    "p_matrix = align.SubstitutionMatrix.std_protein_matrix()\n",
    "n_matrix = align.SubstitutionMatrix.std_nucleotide_matrix()\n",
    "\n",
    "loc_alignments = align.align_optimal(seq1, seq2, p_matrix, local=True)\n",
    "\n",
    "glb_alignments = align.align_optimal(seq1, seq2, p_matrix, local=False)\n",
    "\n",
    "import biotite.sequence.graphics as graphics\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "fig, ax = plt.subplots(figsize = (2.0, 0.8))\n",
    "graphics.plot_alignment_similarity_based(ax, glb_alignments[0], matrix=p_matrix, symbols_per_line=len(glb_alignments[0]))\n",
    "fig.tight_layout()\n",
    "\n",
    "# alignment score\n",
    "# seq identity\n",
    "print(\"Score\", glb_alignments[0].score)\n",
    "print(\"Seq identity\", align.get_sequence_identity(glb_alignments[0]))\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# working with DB\n",
    "import biotite.database.entrez as entrez\n",
    "dir(entrez)\n",
    "# get DB name\n",
    "entrez.get_database_name(\"Nucleotide\")\n",
    "# get DB for protein\n",
    "entrez.get_database_name(\"Protein\")\n",
    "\n",
    "# download and parse protein sequences of covid and Mers\n",
    "covid_file_path = entrez.fetch(\"MN908947\", \"myresult_dir\", suffix=\"fa\", db_name=\"nuccore\", ret_type=\"fasta\")\n",
    "mers_file_path = entrez.fetch(\"NC_019843.3\", \"myresult_dir\", suffix=\"fa\", db_name=\"nuccore\", ret_type=\"fasta\")\n",
    "\n",
    "# reading FASTA file\n",
    "import biotite.sequence.io.fasta as fasta\n",
    "\n",
    "# methods/attributes\n",
    "dir(fasta)\n",
    "\n",
    "c_file = fasta.FastaFile()\n",
    "c_file.read(covid_file_path)\n",
    "\n",
    "m_file = fasta.FastaFile()\n",
    "m_file.read(mers_file_path)\n",
    "\n",
    "for header, s in c_file.items():\n",
    "\tprint(header)\n",
    "\t# print(s)\n",
    "\tcovid_seq = seq.NucleotideSequence(s)\n",
    "\n",
    "for header, s in m_file.items():\n",
    "\tmers_seq = seq.NucleotideSequence(s)\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "# len of seqs\n",
    "print(\"Len of Covid\", len(covid_seq))\n",
    "print(\"Len of MERS\", len(mers_seq))\n",
    "\n",
    "print(\"Freq of N in Covid: \", covid_seq.get_symbol_frequency())\n",
    "print(\"Freq of N in MERS: \", mers_seq.get_symbol_frequency())\n",
    "\n",
    "\n",
    "# gc content of DNA\n",
    "def gc_content(sequence):\n",
    "\tresult = float(str(sequence).count('G') + str(sequence).count('C'))/len(sequence) * 100\n",
    "\treturn result\n",
    "\n",
    "print(f\"GC content Covid: {gc_content(covid_seq)}; GC content MERS: {gc_content(covid_seq)}\")\n",
    "\n",
    "# plot\n",
    "freq_covid = covid_seq.get_symbol_frequency()\n",
    "freq_mers = mers_seq.get_symbol_frequency()\n",
    "plt.bar(freq_covid.keys(), freq_covid.values())\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "# Sequence alignment btw Covid and MERS\n",
    "mini_covid_seq = covid_seq[0:200]\n",
    "mini_mers_seq = mers_seq[0:200]\n",
    "\n",
    "# print(n_matrix)\n",
    "\n",
    "c_alignments = align.align_optimal(mini_covid_seq, mini_mers_seq, n_matrix, gap_penalty = (-10, -1), terminal_penalty=False)\n",
    "\n",
    "# for i in c_alignments:\n",
    "# \tprint(i)\n",
    "\n",
    "fig = plt.figure(figsize=(8.0, 2.5))\n",
    "ax = fig.add_subplot(111)\n",
    "graphics.plot_alignment_similarity_based(\n",
    "\tax, c_alignments[0], matrix=n_matrix, labels=['Covid', 'Mers'], \n",
    "\tshow_numbers = True, show_line_position = True\n",
    ")\n",
    "fig.tight_layout()\n",
    "plt.show()\n",
    "\n",
    "\n",
    "\n",
    "# how identical\n",
    "c_alignments[0].score\n",
    "# seq_identity\n",
    "align.get_sequence_identity(c_alignments[0])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# NeatBio"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import neatbio as nt\n",
    "\n",
    "# methods / attributes\n",
    "dir(nt)\n",
    "\n",
    "# create a DNA sequence\n",
    "seq1 = nt.Sequence(\"ATCATGCT\")\n",
    "type(seq1)\n",
    "\n",
    "# complement\n",
    "seq1.complement()\n",
    "\n",
    "# reverse complement\n",
    "seq1.reverse_complement()\n",
    "\n",
    "\n",
    "\n",
    "# protein synthesis\n",
    "seq1.transcribe()\n",
    "mrna = seq1.transcribe()\n",
    "\n",
    "\n",
    "# translation\n",
    "mrna.translate()\n",
    "seq1.translate()\n",
    "\n",
    "# back transcribe\n",
    "mrna.back_transcribe()\n",
    "\n",
    "# back translation\n",
    "from neatbio.sequtils import back_translate\n",
    "back_translate(\"IM\")\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "# dna composition\n",
    "\n",
    "# GC content\n",
    "seq1.gc\n",
    "\n",
    "seq1.at\n",
    "\n",
    "# GC frequency / count\n",
    "seq1.gc_frequency()\n",
    "\n",
    "seq1.at_frequency()\n",
    "\n",
    "# full GC content\n",
    "from neatbio.sequtils import gc_content, at_content\n",
    "\n",
    "gc_content(seq1)\n",
    "at_content(seq1)\n",
    "\n",
    "# freq of nucleotides\n",
    "seq_freq = seq1.get_symbol_frequency()\n",
    "\n",
    "# visualisation\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "plt.bar(seq_freq.keys(), seq_freq.values())\n",
    "\n",
    "\n",
    "# count of N\n",
    "seq1.count(\"G\")\n",
    "\n",
    "# find position of nucleotide\n",
    "seq1.index(\"G\")\n",
    "\n",
    "# randomise\n",
    "seq1.randomize()\n",
    "seq1\n",
    "\n",
    "# randomly generate N length seq\n",
    "seq2 = nt.Sequence(\"ATGCATCG\")\n",
    "seq2.random_gen(100)\n",
    "\n",
    "p1 = nt.ProteinSeq(\"MITIVF\")\n",
    "\n",
    "dir(p1)\n",
    "\n",
    "# back translate\n",
    "p1.back_translate()\n",
    "\n",
    "seq4 = nt.Sequence(\"ATGATAACTATAGTTTTT\")\n",
    "seq4.translate()\n",
    "\n",
    "# get the aromacity\n",
    "# p1.aromaticity()\n",
    "\n",
    "# # molar extinction coefficient\n",
    "# p1.molar_coeff()\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "# amino acid percentage\n",
    "# p1.get_amino_acid_percentage()\n",
    "\n",
    "\n",
    "# convert 1 letter to 3 letter\n",
    "from neatbio.sequtils import convert_3to1, convert_1to3, get_acid_name\n",
    "\n",
    "convert_1to3(\"MITIVF\")\n",
    "convert_1to3('MI*TIVF', allow_stop = True)\n",
    "\n",
    "convert_3to1('MetIleThrIleValPhe')\n",
    "\n",
    "get_acid_name(\"Met\")\n",
    "get_acid_name(\"MetIleThrIleValPhe\")\n",
    "\n",
    "# dotplot\n",
    "import neatbio.sequtils as utils\n",
    "seq1 = nt.Sequence(\"ATGCTG\")\n",
    "seq2 = nt.Sequence(\"ATGCAG\")\n",
    "\n",
    "utils.dotplot(seq1, seq2)\n",
    "\n",
    "# checking similarity with distance\n",
    "dir(utils)\n",
    "# hamming distance\n",
    "utils.hamming_distance(seq1, seq2)\n",
    "# lev distance\n",
    "utils.levenshtein_distance(seq1, seq2)\n",
    "# minimum_edit distance\n",
    "utils.minimum_edit_distance(seq1, seq2)\n",
    "\n",
    "\n",
    "# Kmers\n",
    "# get Kmers of k2\n",
    "utils.count_kmers(seq1, k=2)\n",
    "# get kmers of k3\n",
    "utils.count_kmers(seq1, k=3)\n",
    "\n",
    "utils.kmer_distance(seq1, seq2)\n",
    "\n",
    "\n",
    "# reading a FASTA file\n",
    "import neatbio as nt\n",
    "my_seq_file = nt.read_fasta(\"sequence.fasta\")\n",
    "my_seq_file['header']\n",
    "my_seq = nt.Sequence(my_seq_file['seqRecord'])\n",
    "my_seq_translate = my_seq.translate()"
   ]
  }
 ],
 "metadata": {
  "interpreter": {
   "hash": "1bc81a7310222af0af7ed6be5dba3d911a1c9b8aee41598a695375ef1a531c1e"
  },
  "kernelspec": {
   "display_name": "Python 3.9.6 64-bit",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.6"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
